---
title: "malepoolmpile"
author: "Casey Flay"
date: "30/09/2020"
output:
  html_document: default
  word_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(RLinuxModules)
library(data.table)
library(glue)
library(here)
library(tidyverse)
library(magrittr)
library(GenomicRanges)
library(tools)
library(future)
library(future.batchtools)
library(here)
plan(batchtools_openlava, resources = list(threads=1))
setDTthreads(1)
module("load openlava asub/2.2 bwa/0.7.12 samtools/1.9 bcftools/1.10.2 popoolation2 perlbrew/0.76")#old samtools module loaded for popoolation2
setwd("/powerplant/workspace/hrtcdf/github/FSTs/PSApopoolation")
here()
```

**parent files pulled from database, merged and piled with samtools/0.1.18 below**
```{r parents NCBI aligned bamfiles from male pool and testpools}
Rualn_bamDir <- "/output/genomic/plant/Actinidia/chinensis/Resequencing/Alignments/Russell_V2a.chromosomes.and.unassigned"
Rualn_bams  <- as.data.table(grep(paste('.bam$', collapse = "|"), list.files(path=Rualn_bamDir), invert=FALSE, value=TRUE))
Rualn_bams  <- Rualn_bams[, lapply(.SD, gsub, pattern = ("_Russell_V2a"), replacement = "&_Russell_V2a&"), .SDcols = colnames(Rualn_bams)][, c("er","reference",  "ext") := tstrsplit(V1, "&", fixed=TRUE)]
ER  <- Rualn_bams[, lapply(.SD, gsub, pattern = ("ER\\d{5}_"), replacement = ""), .SDcols = colnames(Rualn_bams)]
ER  <- ER[, lapply(.SD, gsub, pattern = ("_ER\\d{5}"), replacement = ""), .SDcols = colnames(Rualn_bams)]
ER  <- ER[, lapply(.SD, gsub, pattern = ("SRA_"), replacement = ""), .SDcols = colnames(Rualn_bams)]
ER  <- ER[, lapply(.SD, gsub, pattern = ("UNK_"), replacement = ""), .SDcols = colnames(Rualn_bams)]
setnames(ER, c("er"), c("PedigreeItem"))
Rualn_bams  <-cbind(Rualn_bams$er, ER)
ER  <- Rualn_bams[, c(1,3,4,5)]

ER  <- ER[, "er1" := str_extract(Rualn_bams$V1, "ER\\d{5}_")][, "er2" := str_extract(Rualn_bams$V1, "_ER\\d{5}")][, "er3" := str_extract(Rualn_bams$V1, "SRA_")][, "er4" := as.data.table(str_extract(Rualn_bams$V1, "UNK_"))]
ER  <- ER[, "er5" := paste0(er1,er2,er3,er4)]
ER5  <- ER[, lapply(.SD, gsub, pattern = ("NA"), replacement = ""), .SDcols = "er5"]
Rualn_bams  <-cbind(ER, ER5)
Rualn_bams  <- Rualn_bams[, c(1,2,3,4,10)]
Rualn_bams <- Rualn_bams[, "full" := paste0(er5,PedigreeItem,reference,ext)]
Rualn_bams <- Rualn_bams[, "dir" := Rualn_bamDir]
Rualn_bams <- Rualn_bams[, "pathandfile" := paste(dir, full, sep="/")]
t<- as.data.table(Rualn_bams$full)

Rualn_bams <- Rualn_bams[, "row" := seq_along(full)]
Rualn_bams <- Rualn_bams[PedigreeItem=="Red19", PedigreeItem := ("Red 19")]
Rualn_bams <- Rualn_bams[, "isfile" := (file.exists(pathandfile))]
Rualn_bams <- Rualn_bams[isfile != FALSE,]
rm(ER,ER5)
```

```{r}
#Rualn_bams <- as.data.table(unique(Rualn_bams, by = ("PedigreeItem")))

#male pool parents are all the same for each pool e.g A1, B2 etc. unless adding multiple copies of each.
Russell_tp_bams <- Rualn_bams[PedigreeItem == 'CK02_04' | PedigreeItem == 'CK06_01' | PedigreeItem == 'CK09_06'| PedigreeItem == 'CK13_02'| PedigreeItem =='CK20_01'| PedigreeItem =='CK51_06'| PedigreeItem =='Russell'][, `:=` ("pool_parent"= 'Russell', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]# only found "CK06_01","CK20_01","CK51_06". missing 'CK02_04', 'CK13_02','CK09_06'.

#Full pool parents these tables are also needed in the merge step but cant be used in a single table.
CK02_01_tp_bams <- Rualn_bams[PedigreeItem=='CK02_01'| PedigreeItem=='CK15_03'| PedigreeItem=='CK17_02'| PedigreeItem=='CK20_01'| PedigreeItem=='Bejing #36'| PedigreeItem == 'CK51_01'][, `:=` ("pool_parent"= 'CK02_01', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK02_01
CK10_02_tp_bams <- Rualn_bams[PedigreeItem=='CK10_02'| PedigreeItem=='CK17_03'| PedigreeItem=='CK18_01'| PedigreeItem=='CK19_03'| PedigreeItem=='CK51_09'][, `:=` ("pool_parent"= 'CK10_02', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK10_02
CK19_03_tp_bams <- Rualn_bams[PedigreeItem=='CK19_03'| PedigreeItem=='CK10_02'| PedigreeItem=='CK10_05'| PedigreeItem=='CK13_02'| PedigreeItem=='CK15_04'| PedigreeItem=='Bejing #36'| PedigreeItem=='CK51_01'][, `:=` ("pool_parent"= 'CK19_03', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK19_03
CK20_01_tp_bams <- Rualn_bams[PedigreeItem=='CK20_01'| PedigreeItem=='CK02_01'| PedigreeItem=='CK09_11'| PedigreeItem=='CK10_03'][, `:=` ("pool_parent"= 'CK20_01', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK20_01
CK22_03_tp_bams <- Rualn_bams[PedigreeItem=='CK22_03'| PedigreeItem=='CK23_08'| PedigreeItem=='CK51_09'][, `:=` ("pool_parent"= 'CK22_03', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK22_03
CK23_08_tp_bams <- Rualn_bams[PedigreeItem=='CK23_08'| PedigreeItem=='CK01_01_01_01'| PedigreeItem=='CK15_03'| PedigreeItem=='CK16_01'| PedigreeItem=='CK22_03'][, `:=` ("pool_parent"= 'CK23_08', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK23_08
CK51_01_tp_bams <- Rualn_bams[PedigreeItem=='CK51_01'| PedigreeItem=='CK02_01'| PedigreeItem=='CK17_03'| PedigreeItem=='CK19_03'][, `:=` ("pool_parent"= 'CK51_01', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK51_01
CK51_09_tp_bams <- Rualn_bams[PedigreeItem=='CK51_09'| PedigreeItem=='CK02_04'| PedigreeItem=='CK06_01'| PedigreeItem=='CK10_02'| PedigreeItem=='CK15_04'| PedigreeItem=='CK22_03'][, `:=` ("pool_parent"= 'CK51_09', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK51_09
parent_bams <- rbindlist(list(Russell_tp_bams, CK02_01_tp_bams, CK10_02_tp_bams, CK19_03_tp_bams, CK20_01_tp_bams, CK22_03_tp_bams, CK23_08_tp_bams, CK51_01_tp_bams, CK51_09_tp_bams))

parent_bams[, "psymlink_dir" := "/powerplant/workspace/hrtcdf/github/FSTs/parentpools"]
parent_bams <- parent_bams[, "pbams" := paste0(er5,PedigreeItem,reference, ext, sep="")]
#file.symlink(unique(parent_bams$pathandfile), "/powerplant/workspace/hrtcdf/github/FSTs/PSApopoolation")
```



```{r index}
bamin          <- paste0(unique(parent_bams$pbams))
file.exists(bamin)
for (i in bamin){
vcf_cmd        <- glue::glue("samtools index {i}")
logDir         <- "/powerplant/workspace/hrtcdf/github/FSTs/PSApopoolation/logDir"
bsub_out_file  <- file.path(logDir, "job.out"); bsub_err_file <- file.path(logDir, "job.err")
bsub_cmd       <- glue("bsub -n 8 -o {bsub_out_file} -e {bsub_err_file} {vcf_cmd}")
system(bsub_cmd)
}
```

```{r  merge vcfs into parent pools}
module("unload samtools")
module ("load samtools/1.9")

for (i in unique(parent_bams$pool_parent)){ 
x <- parent_bams[pool_parent=={i}, pbams]
xt <- paste0(x, sep = "")

file.exists(xt)
vcfcmd         <- glue("samtools merge -f -r {i}_mrgdparents.bam {xt}")
OutEr          <- "stmrg5";  logD<- here("logD");  bsO<- file.path(logD, paste0(OutEr,".out")); bsE<- file.path(logD, paste0(OutEr,".err"))
bsub_cmd       <- glue("bsub -n 8 -o {bsO} -e {bsE} {vcfcmd}")
system(bsub_cmd)
}
```

```{r update malepooldt_bam to malepooldt1}
merged_parent_bams <- unique(parent_bams, by = c("pool_parent"))
merged_parent_bams <- merged_parent_bams[, `:=` ("mrgedparents"= paste0(merged_parent_bams$pool_parent, "_mrgdparents.bam", sep=""))]
merged_parent_bams <- merged_parent_bams[, "mergedDir" := "/powerplant/workspace/hrtcdf/github/FSTs/PSApopoolation/"]
```

```{r  replace read groups not used yet as merge altered}
module("unload samtools")
module ("load samtools/1.9")
x <- merged_parent_bams[, mrgedparents]

for (i in x){ 
f <- file_path_sans_ext({i})
file.exists(i)
vcfcmd         <- glue("samtools addreplacerg -R {f} {f}_rg.bam {i}")
OutEr          <- "rg2";  logD<- here("logD");  bsO<- file.path(logD, paste0(OutEr,".out")); bsE<- file.path(logD, paste0(OutEr,".err"))
bsub_cmd       <- glue("bsub -n 8 -o {bsO} -e {bsE} {vcfcmd}")
#system(bsub_cmd)
}
```


```{r index}
bamin          <- paste0(merged_parent_bams$mrgedparents)
file.exists(bamin)
for (i in bamin){
vcf_cmd        <- glue::glue("samtools index {i}")
OutEr          <- "index133";  logD<- here("logD");  bsO<- file.path(logD, paste0(OutEr,".out")); bsE<- file.path(logD, paste0(OutEr,".err"))
bsub_cmd       <- glue("bsub -n 8 -o {bsub_out_file} -e {bsub_err_file} {vcf_cmd}")
system(bsub_cmd)
}
```
##havent run this chunk as the male test needs to be run first.
```{r run samtools mpileup for chr25 using samtools/0.1.18. since popoolation2 requires that version }
module("unload samtools")
module ("load samtools/0.1.18")

submit_mpileup <- function(reference = "*.fa.gz", output = c(), bamfiles = c(), job.init = "asub", job.opts = "-q normal") {
  cmd          <- "samtools mpileup -r 'chr25' -f {reference} {bamfiles} > {out}" 
  tmpf         <- tempfile(pattern = "submit_mileup.", fileext = ".sh")
  writeLines(text = c(glue(cmd)), con = tmpf)
  submit       <- "{job.init} {job.opts} {tmpf}"
  system(glue(submit), intern = TRUE)
}

PedigreeItem <- paste0(merged_parent_bams$pool_parent)
bamfiles     <- paste0(merged_parent_bams$mrgedparents)
file.exists(bamfiles)

out         <- paste("sampile_parents", PedigreeItem, "chr25.vcf", sep = "_")
reference   <- dir("/workspace/hrards/projects/hrtcdf/genome/fasta", pattern = ".*\\.gz$", full.names = TRUE)
#file.exists(reference)
submit_mpileup(reference = reference, output = out, bamfiles = bamfiles)
```





***male pools from hrtcdf/github/FSTs/malepool/malepool.Rmd***
```{r symlink male pool bam files}
mainDir <- "/powerplant/workspace/hrtcdf/github/FSTs/"
subDir <- "PSApopoolation"
royDir <- "/workspace/hrards/projects/hrtcdf/10.alignments/bwa"
outDir <- str_c(mainDir, subDir, sep = "", collapse = TRUE)

if (dir.exists(mainDir)){
    setwd(file.path(mainDir, subDir))
} else {
    dir.create(file.path(mainDir, subDir))
    setwd(file.path(mainDir, subDir))
}

mspoolfiles       <- dir(royDir, pattern = ".bam", full.names = TRUE, include.dirs = TRUE)

file.symlink(mspoolfiles, outDir)
```

```{r transform filenames to sample, lane and suffix for data.table from symlink}
mpooldt_bam <- as.data.table(grep(paste("^HW5YYDSXX"), list.files(path="."), invert=FALSE, value=TRUE))
mpooldt_bam <- mpooldt_bam[, "pooltype" := "variablemale"]
spooldt_bam <- as.data.table(grep(paste("^HL55WDRXX"), list.files(path="."), invert=FALSE, value=TRUE))
spooldt_bam <- spooldt_bam[, "pooltype" := "lowPSA"]
pooldt_bam  <- rbind(mpooldt_bam, spooldt_bam)

pooldt_bam <- pooldt_bam[, V2 := str_replace_all(V1, "_L" , "&_L")
               ][, V2 := str_replace_all(V2, "XX_" , "XX_&")
               ][, V2 := str_replace_all(V2, ".bam" , "&.bam") 
                 
               ][, c("H", "sample", "lane", "ftype") := tstrsplit(V2, "&", fixed=TRUE)
               ][, V2:=NULL]

#subset malepool from parent pool files
fwrite(pooldt_bam, 'pooldt_bam.tsv', append=FALSE)
```

```{r sort}
pooldt_bam <- pooldt_bam[, "sort" := "s_"] # for next step
bamin          <- pooldt_bam[ftype==".bam", V1]
for (i in bamin){
vcfcmd         <- glue::glue("samtools sort -o s_{i} {i}")
OutEr          <- "si206";  logD<- here("logD");  bsO<- file.path(logD, paste0(OutEr,".out")); bsE<- file.path(logD, paste0(OutEr,".err"))
bsub_cmd       <- glue("bsub -n 8 -o {bsO} -e {bsE} {vcfcmd}")
system(bsub_cmd)
}
```

```{r index}
bamin <- paste(pooldt_bam[ftype==".bam", sort], pooldt_bam[ftype==".bam", V1], sep="")

for (i in bamin){
vcfcmd         <- glue::glue("samtools index {i}")
OutEr          <- "index218";  logD<- here("logD");  bsO<- file.path(logD, paste0(OutEr,".out")); bsE<- file.path(logD, paste0(OutEr,".err"))
bsub_cmd       <- glue("bsub -n 8 -o {bsO} -e {bsE} {vcfcmd}")
system(bsub_cmd)
}
```

```{r merge lanes for sample lowPSA bam files}
module("unload samtools")
module ("load samtools/1.9")

for (i in unique(pooldt_bam$sample)){ 
x <- pooldt_bam[sample=={i} & ftype==".bam", V1]
xt <- paste0(x, sep = "")

file.exists(xt)

cmd       <- glue("samtools merge -f -r {i}_mrgdlowPSA.bam {xt}")
OutEr     <- "merge216";  logD<- here("logD");  bsO<- file.path(logD, paste0(OutEr,".out")); bsE<- file.path(logD, paste0(OutEr,".err"))
bsub_cmd  <- glue("bsub -n 8 -o {bsO} -e {bsE} {cmd}")
system(bsub_cmd)
}
```
```{r re- add readgroups?}

```

```{r}
lowPSApooldt_bam <- as.data.table(paste0(unique(mpooldt_bam$sample), "_mrgdlowPSA.bam"))
lowPSApooldt_bam <- lowPSApooldt_bam[, sample := str_replace_all(V1, "_mrgdlowPSA.bam" , "")]
variablemale     <- pooldt_bam[pooltype=="variablemale", .(V1, sample)]
mpilepoolbams    <- rbind(lowPSApooldt_bam, variablemale)
```

```{r index}
bamin          <- paste(mpilepoolbams$V1)
for (i in bamin){
vcfcmd         <- glue::glue("samtools index {i}")
logDir         <- "/powerplant/workspace/hrtcdf/github/FSTs/PSApopoolation/logDir"
outer          <- "a3"; bs_out<- file.path(logDir, paste0(outer,".out")); bs_err<- file.path(logDir, paste0(outer,".err"))
bsub_cmd       <- glue("bsub -n 8 -o {bsub_out_file} -e {bsub_err_file} {vcfcmd}")
system(bsub_cmd)
}
```


***run samtools mpileup for each pool using samtools/0.1.18. in "PSApopoolation.Rmd" since popoolation2 requires samtools/0.1.18 run mpileup to make files which contain reads from both pools to be analysed.***


```{r remove unneeded files}
r1     <- paste0(merged_parent_bams$mrgedparents)
bamin1          <- paste(mpilepoolbams$V1)
bamin2          <- paste(pooldt_bam$V1)
bai <- grep(paste("*.bai"), list.files(path="."), invert=FALSE, value=TRUE)

#file.remove(r1, bamin1, bamin2, bai)
```

