---
title: "PSApopoolationChr25"
author: "Casey Flay"
date: "30/09/2020"
output:
  html_document: default
  word_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(RLinuxModules)
library(glue)
library(tidyverse)
library(here)
plan(batchtools_openlava, resources = list(threads=1))
setDTthreads(1)
module("load openlava asub/2.2 bwa/0.7.12 samtools/1.9 bcftools/1.10.2 popoolation2 perlbrew/0.76")#old samtools module loaded for popoolation2
setwd("/powerplant/workspace/hrtcdf/github/FSTs/PSApopoolation")
here()
```

```{r run samtools mpileup for indexed reference mpilechr25 using samtools/0.1.18.}
module("unload samtools")
module ("load samtools/0.1.18")
submit_mpileup <- function(reference = "*.fa.gz", output = c(), bamfiles = c(), job.init = "asub", job.opts = "-q normal") {
  cmd          <- "samtools mpileup -r 'chr25' -f {reference} {bamfiles} > {out}" 
  tmpf         <- tempfile(pattern = "submit_mileup.", fileext = ".sh")
  writeLines(text = c(glue(cmd)), con = tmpf)
  submit       <- "{job.init} {job.opts} {tmpf}"
  system(glue(submit), intern = TRUE)
}

bamfiles    <- paste(mpilepoolbams$V1)
file.exists(bamfiles)
out         <- paste(mpilepoolbams$sample, "sampile_chr25.vcf", sep = "_")
reference   <- dir("/workspace/hrards/projects/hrtcdf/genome/fasta", pattern = ".*\\.gz$", full.names = TRUE)
file.exists(reference)
#submit_mpileup(reference = reference, output = out, bamfiles = bamfiles)
```

```{r files dt}
pofiles <- as.data.table(grep(".vcf$", list.files(path="."), invert=FALSE, value=TRUE))
pofiles <- pofiles[, V2 := str_replace_all(V1, "_sampile_" , "&_sampile_")]
pofiles <- pofiles[, V2 := str_replace_all(V2, "_chr25.vcf" , "&_chr25.vcf")]
pofiles <- pofiles[, V2 := str_replace_all(V2, "sampile_parents_" , "sampile_parents_&")
                 ][, c("V3", "V4", "V5") := tstrsplit(V2, "&", fixed=TRUE)
                 ][, V2:=NULL]
pofiles <- pofiles[V3=="A1", "malepercentage" := "m90"
                  ][V3=="B2", "malepercentage" := "m75" 
                  ][V3=="C3", "malepercentage" := "m60"                    
                  ][V3=="D6", "malepercentage" := "m44"                    
                  ][V3=="E11", "malepercentage" := "m80"                    
                  ][V3=="F12", "malepercentage" := "m50"                    
                  ][V3=="G13", "malepercentage" := "m45"
                  ][V3=="H14", "malepercentage" := "m68"  
                  ][V3=="I15", "malepercentage" := "m21"                    
                  ][V3=="J16", "malepercentage" := "m11"]
pofiles <- pofiles[, "sync" := ".sync"]# for next step
```

```{r make sync files with mpileup2sync}
parents     <- pofiles[V4=="Russell", V1]
m           <- pofiles[!is.na(malepercentage), V1]
mp <- append(parents, m)

#file.exists(mp)

cmd          <- glue::glue("java -ea -Xmx16g -jar  /software/bioinformatics/popoolation2-1201/mpileup2sync.jar --input {mp} --output {mp}.sync --fastq-type sanger --min-qual 20 --threads 8")
outer        <- "a14";  logD<- here("logD");  bsO<- file.path(logD, paste0(outer,".out")); bsE<- file.path(logD, paste0(outer,".err"))
bsub_cmd     <- glue("bsub -n 8 -o {bsO} -e {bsE} {cmd}")
system(bsub_cmd)
```


```{r run CMH test - not running yet as i dont know what to insert into input if pools are compared in population}
m     <- pofiles[malepercentage=="m90", V1]
parents <- pofiles[V4=="Russell", V1]

cmd          <- glue::glue("perl /software/bioinformatics/popoolation2-1201/cmh-test.pl --min-count 6 --mincoverage 4 --max-coverage 250 --population {m},{parents} --input smallcmh.sync --output cmhtest.txt")
outer        <- "atest";  logD<- here("logD");  bsO<- file.path(logD, paste0(outer,".out")); bsE<- file.path(logD, paste0(outer,".err"))
bsub_cmd     <- glue("bsub -n 8 -o {bsO} -e {bsE} {cmd}")
system(bsub_cmd)
```

```{r}
m     <- pofiles[, V1]
perl <popoolation2-path>/ --input p1_p2.sync --output-prefix p1_p2 --min-count 6 --min-coverage 50 --max-coverage 200

cmd          <- glue::glue("perl /software/bioinformatics/popoolation2-1201/snp-frequency-diff.pl --input {m}.sync --output-prefix p1_p2 --min-count 6 --min-coverage 50 --max-coverage 200")
outer        <- "atest";  logD<- here("logD");  bsO<- file.path(logD, paste0(outer,".out")); bsE<- file.path(logD, paste0(outer,".err"))
bsub_cmd     <- glue("bsub -n 8 -o {bsO} -e {bsE} {cmd}")
system(bsub_cmd)
```

```{r harsh resample of sync files - mainly needed once subsampling from all chromosomes}
bamin        <- paste(pofiles$V1, pofiles$sync, sep="")
file.exists(bamin)
for (i in bamin){
cmd          <- glue::glue("perl /software/bioinformatics/popoolation2-1201/subsample-synchronized.pl --input {i} --output {i}rs --target-coverage 20 --max-coverage 250 --method withreplace")
outer        <- "a11";  logD<- here("logD");  bsO<- file.path(logD, paste0(outer,".out")); bsE<- file.path(logD, paste0(outer,".err"))
bsub_cmd     <- glue("bsub -n 8 -o {bsO} -e {bsE} {cmd}")
system(bsub_cmd)
}
```
```{r update dt}
pofiles <- pofiles[, "syncrs" := ".syncrs"]
```

```{r Make a genebased sync for chr25 and resample}
chr25gff3 <- "/output/genomic/plant/Actinidia/chinensis/Russell/Genome/Assembly/V2a/Gene_models/Archive/Russell_V2a.chromosomes.and.unassigned.renamed.has.stops.gff3"

filein        <- paste0(pofiles$V1, pofiles$syncrs)
file.exists(filein)
for (i in filein){
cmd          <- glue::glue("/software/bioinformatics/popoolation2_1201/create-genewise-sync.pl --input {i} --output {i} \
                            --input ./popoolation/CHR26.resample.sync \
                            --gtf {chr25gff3} \
                            --output ./popoolation/CHR26.resample.genes.sync")
outer        <- "a12";  logD<- here("logD");  bsO<- file.path(logD, paste0(outer,".out")); bsE<- file.path(logD, paste0(outer,".err"))
bsub_cmd     <- glue("bsub -n 8 -o {bsO} -e {bsE} {cmd}")
system(bsub_cmd)
}
```