---
title: "Male allele frequency"
author: "Casey"
date: "03/08/2020"
output: html_document
---
##git cannot access .gitconfig move to documents or move gitconfig to c:workspace/casey-s-repository.
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{bash}
mkdir TestPool/ FullPool/
ls
```

## make links to bam files from Roys alignment

```{bash}
inputDir=/powerplant/workspace/hrards/projects/hrtcdf/10.alignments/bwa
outputDir1=/powerplant/workspace/hrtcdf/FSTs/TestPool
outputDir2=/powerplant/workspace/hrtcdf/FSTs/FullPool
cd $inputDir

## Add symbolic links
for TP in $(ls HL55*); do
  ln -s /powerplant/workspace/hrards/projects/hrtcdf/10.alignments/bwa/$TP $outputDir1/$TP
done


for TP in $(ls HW5*); do
  ln -s /powerplant/workspace/hrards/projects/hrtcdf/10.alignments/bwa/$TP $outputDir2/$TP
done
```

```{bash}
pwd
module load bwa/0.7.12
module load samtools/1.2
module load "load openlava"
```
```{bash}

```


## Load reference genome with index files

```{bash}
pwd
#ln -s /output/genomic/plant/Actinidia/chinensis/Russell/Genome/Assembly/Russell_V2a
ls -l
```

```{bash}
module load openlava
bsub samtools faidx Russell_V2a.fasta
```

```{bash}
module load openlava
bsub "java -jar /software/bioinformatics/picard-tools-2.2.4/picard.jar CreateSequenceDictionary \
    R=Russell_V2a.fasta O=Russell_V2a.dict"

```
```{bash}
index=/Russell_V2a.fasta.dict
```

Skipped "Try to work out loop for submitting alignment" part becasue Roy has aligned fasta's and made Bam files already. went straight toIndex Bams for IGV.

##Index Bams for IGV
```{bash}
module load samtools/1.2
samtools index --help
```

```{bash}
inputDir=/powerplant/workspace/hrards/projects/hrtcdf/10.alignments/bwa/
outputDir1=/powerplant/workspace/hrtcdf/FSTs/TestPool
outputDir2=/powerplant/workspace/hrtcdf/FSTs/FullPool

# cd $outputDir1/

module load samtools/1.2
module load openlava/3.2


for BAM in $(find $outputDir1 -name "*.bam"); do
  bsub -o log_stdout1.out -e log_stderr1.err "samtools index -b ${BAM}"
done

for BAM in $(find $outputDir2 -name "*.bam"); do
  bsub -o log_stdout2.out -e log_stderr2.err "samtools index -b ${BAM}"
done
```
#Run mpileupÂ¶
#from popoolation howto

#samtools mpileup -B -Q 0 -f wg/2R.chr pop1.bam pop2.
#bam > p1-2.mpileup

```{bash}
mkdir -p -v mpileup/log
```
##Make an indexed reference
```{bash}
module load samtools/1.2
module load openlava
bsub "samtools faidx Russell_V2a.fasta"
```

```{bash}
for BAM in $(ls *.bam) 
    do bsub "samtools index -b  ${BAM}"
    done
```

Submit to Cluster
NB would be best to parallel by use of -r option to call by chromsome

could probably tune better for indel realignment
see http://samtools.sourceforge.net/mpileup.shtml
and also pipe direct into Varscan
Adjust -m and -F to control when to initiate indel realignment (requiring r877+). Samtools only finds INDELs where there are sufficient reads containing the INDEL at the same position. It does this to avoid excessive realignment that is computationally demanding. The default works well for many low-coverage samples but not for, say, 500 exomes. In the latter case, using -m 3 -F 0.0002 (3 supporting reads at minimum 0.02% frequency) is necessary to find singletons.

#Submit One Big Job
```{bash}
bsub  -o ./mpileup/mpileup.out -e ./mpileup/mpileup.err -n 8  -J POOLmpileup  \
    "samtools mpileup -B -Q 0 -f ./LGs_1_to_29.fasta -o ./mpileup/P1-5.mpileup \
    ./align/pool1.merged.bam   ./align/pool2.merged.bam  ./align/pool3.merged.bam  ./align/pool4.merged.bam"
```

```{r run-mileup}
submit_mpileup <- function(reference = "*.fa.gz", output = c(), bamfiles = c(), job.init = "asub", job.opts = "-q normal") {
  cmd <- "samtools mpileup -B -Q 0 -f {reference} -o {output} {bamfiles}"
  tmpf <- tempfile(pattern = "submit_mileup.", fileext = ".sh")
  writeLines(text = c(glue(cmd)), con = tmpf)
  submit <- "{job.init} {job.opts} {tmpf}"
  system(glue(submit), intern = TRUE)
}

data.table::fread(file = "/workspace/hrards/projects/hrtcdf/targetsfile")

bamfiles <- dir("/workspace/hrards/projects/hrtcdf/10.alignments/bwa", full.names = TRUE)
bamfiles
reference <- dir("/workspace/hrards/projects/hrtcdf/genome/fasta", pattern = ".*\\.gz$", full.names = TRUE)

reference

submit_mpileup(reference = reference, output = "test.txt", bamfiles = paste0(bamfiles, collapse = " "))

```


## Get a test file of Bam files to be run for asub
```{bash} 
Dir1=TestPool
Dir2=FullPool

cd /powerplant/workspace/hrtcdf/FSTs/$Dir1
  for bam in *.bam; do 
  echo "/powerplant/workspace/hrtcdf/$Dir1/$bam" >>$Dir1"Bam.txt";
  done
  
cd /powerplant/workspace/hrtcdf/FSTs/$Dir2
  for bam in *.bam; do 
  echo "/powerplant/workspace/hrtcdf/$Dir2/$bam" >>$Dir2"Bam.txt";
  done

```

```{bash}
module load openlava
bjobs
```
```{bash}
outputDir1=/powerplant/workspace/hrtcdf/FSTs/TestPool
ls $outputDir1
```


```{bash}
pwd
```
```{bash}
cd ~
ls
```

```{bash}
inputDir=/powerplant/workspace/hrards/projects/hrtcdf/10.alignments/bwa
outputDir1=/powerplant/workspace/hrtcdf/FSTs/TestPool
outputDir2=/powerplant/workspace/hrtcdf/FSTs/FullPool
cd $outputDir1
rm core*
ls
```
```{bash}
echo "# Allele-Frequency" >> README.md
```
git init
git add your files and README.md
git commit -m "first commit"
git remote add origin https://github.com/caseyflay/Allele-Frequency.git
https://github.com/caseyflay/AlleleFrequency/tree/TestPool
git push -u origin master

```{bash}
module load openlava
bjobs
```
