---
title: "qtlseqr"
author: "Casey Flay"
date: "12/08/2021"
output:
  html_document: default
  word_document: default
---

```{r setup, include=FALSE}
library(knitr)
library(RLinuxModules)
library(data.table)
library(glue) 
library(tidyverse)
library(here)
library(QTLseqr)
setDTthreads(1)
module("load openlava asub/2.2 bwa/0.7.12 samtools/1.9 bcftools/1.10.2 perlbrew/0.76")
knitr::opts_chunk$set(echo = FALSE, include=FALSE, warning = TRUE)
```

```{r transform filenames to sample, lane and suffix for data.table from symlink}
mpool  <- as.data.table(grep(paste("_samplepoolSNPs_dt.csv"), list.files(path="."), invert=FALSE, value=TRUE))
mpool  <- mpool[, V2 := str_replace_all(V1, "_samplepoolSNPs_dt.csv", "&samplepoolSNPs_dt.csv")
              ][, V2 := str_replace_all(V2, ".csv" , "&.csv")
              ][, c("V2", "V3", "V4") := tstrsplit(V2, "&", fixed=TRUE)
              ][V2!="Russell",]
#for males
# mpool2  <- mpool[, V2 := str_replace_all(V1, "_malepoolSNPS_dt.csv", "&malepoolSNPS_dt.csv")
#              ][, V2 := str_replace_all(V2, ".csv" , "&.csv")
#              ][, c("V2", "V3") := tstrsplit(V2, "&", fixed=TRUE)]#, "V4"
# mpool <- rbind(mpool, mpool2)
#rm(mpool2)
```

##code to make these files is in qtlseqr_files.Rmd 
```{r group samples against parent pools and rename for QTLseqR and assign columns for QTLseqR}
obj_names  <- mpool[,.(V2)]
obj_rename <- obj_names[V2=='CK02_01_samplepoolSNPs_dt', 'rename' :="P1sp_bsa"][V2=='CK10_02_samplepoolSNPs_dt', rename :="P9sp_bsa"][V2=='CK19_03_samplepoolSNPs_dt', rename :="P2sp_bsa"][V2=='CK20_01_samplepoolSNPs_dt', rename :="P10sp_bsa"][V2=='CK22_03_samplepoolSNPs_dt', rename :="P11sp_bsa"][V2=='CK23_08_samplepoolSNPs_dt', rename :="P3sp_bsa"][V2=='CK51_01_samplepoolSNPs_dt', rename :="P12sp_bsa"][V2=='CK51_09_samplepoolSNPs_dt', rename :="P4sp_bsa"]
obj_names  <- obj_rename[,rename]
my_files   <- mpool[,V1]

#70gb run below, only run when reduction filters are inplace
for(i in seq(along=my_files)) {
  assign(obj_names[i], fread(my_files[i]))
}

#relabel samples for qtlseqR s=sample r=reference from qtlseqr notation which is the pool of parents here
samplesold <- c('s1CHROM','s1POS','s1REF','s1ALT','r1AD0','r1AD1','r1DP','r1GQ','r1PL','r1SNPi','s1AD0','s1AD1','s1DP','s1GQ','s1PL','s1SNPi','REF_FRQ','deltaSNP')
samplesnew <-c('CHROM','POS','REF','ALT','AD_REF.LOW','AD_ALT.LOW','DP.LOW','GQ.LOW','PL.LOW','SNPindex.LOW','AD_REF.HIGH','AD_ALT.HIGH','DP.HIGH','GQ.HIGH','PL.HIGH','SNPindex.HIGH','REF_FRQ','deltaSNP')

for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  assign(obj_names[i], setnames(dt_name, glue("{samplesold}"), glue("{samplesnew}")))
}
```

```{r clean columns and change to chr01 for ggplot to assemble in the right order}

for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  assign(obj_names[i], dt_name[, `:=` ('p1CHROM'=NULL, 'p1POS'=NULL, 'p1REF'=NULL, 'p1ALT'=NULL, 'p1AD0'=NULL, 'p1AD1'=NULL, 'p1DP'=NULL, 'p1GQ'=NULL, 'p1PL'=NULL, 'V19'=NULL)])
}

#Russell <- Russell[, `:=` ('p1CHROM'=NULL,'p1POS'=NULL,'p1REF'=NULL,'p1ALT'=NULL,'p1AD0'=NULL,'p1AD1'=NULL,'p1DP'=NULL,'p1GQ'=NULL,'p1PL'=NULL,'V19'=NULL)]

for(i in seq(along=obj_names)) {
  assign(obj_names[i], get(obj_names[i], envir = .GlobalEnv)[, "CHROM" := as.factor(CHROM)])
}
#sapply(F12G13, class)

chrs <- c("chr1", "chr2", "chr3",  "chr4",  "chr5",  "chr6",  "chr7",  "chr8",  "chr9", "chr10", "chr11", "chr12", "chr13", "chr14", "chr15", "chr16", "chr17", "chr18", "chr19",  "chr20", "chr21", "chr22", "chr23", "chr24", "chr25", "chr26", "chr27", "chr28", "chr29")

for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  assign(obj_names[i], dt_name[, CHROM := fct_relevel(dt_name[,CHROM], chrs, after = Inf)])
}
#ck0201<-setnames(ck0201, glue("{samplesold}"),glue("{samplesnew}"))
```

```{r filter, eval=TRUE}
# P1sp_bsa <- filterSNPs(SNPset = P1sp_bsa, refAlleleFreq = 0.05, minTotalDepth = 40, maxTotalDepth = 85, depthDifference = 50, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
# P9sp_bsa <- filterSNPs(SNPset = P9sp_bsa, refAlleleFreq = 0.05, minTotalDepth = 40, maxTotalDepth = 130, depthDifference = 50, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
# P2sp_bsa <- filterSNPs(SNPset = P2sp_bsa, refAlleleFreq = 0.05, minTotalDepth = 35, maxTotalDepth = 85, depthDifference = 50, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
# P10sp_bsa <- filterSNPs(SNPset = P10sp_bsa, refAlleleFreq = 0.05, minTotalDepth = 40, maxTotalDepth = 90, depthDifference = 50, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
# P11sp_bsa <- filterSNPs(SNPset = P11sp_bsa, refAlleleFreq = 0.05, minTotalDepth = 60, maxTotalDepth = 170, depthDifference = 50, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
# P3sp_bsa <- filterSNPs(SNPset = P3sp_bsa, refAlleleFreq = 0.05, minTotalDepth = 20, maxTotalDepth = 85, depthDifference = 50, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
# P12sp_bsa <- filterSNPs(SNPset = P12sp_bsa, refAlleleFreq = 0.05, minTotalDepth = 30, maxTotalDepth = 100, depthDifference = 50, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
# P4sp_bsa <- filterSNPs(SNPset = P4sp_bsa, refAlleleFreq = 0.05, minTotalDepth = 40, maxTotalDepth = 150, depthDifference = 50, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
Russell <- filterSNPs(SNPset = Russell, refAlleleFreq = 0.05, minTotalDepth = 10, maxTotalDepth = 90, depthDifference = 50, minSampleDepth = 10, minGQ = 50, verbose = TRUE)
a_df <- filterSNPs(SNPset = a_df, refAlleleFreq = 0.05, minTotalDepth = 60, maxTotalDepth = 120, depthDifference = 50, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
b_df <- filterSNPs(SNPset = b_df, refAlleleFreq = 0.05, minTotalDepth = 50, maxTotalDepth = 120, depthDifference = 50, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
c_df <- filterSNPs(SNPset = c_df, refAlleleFreq = 0.05, minTotalDepth = 50, maxTotalDepth = 110, depthDifference = 50, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
d_df <- filterSNPs(SNPset = d_df, refAlleleFreq = 0.05, minTotalDepth = 60, maxTotalDepth = 140, depthDifference = 50, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
e_df <- filterSNPs(SNPset = e_df, refAlleleFreq = 0.05, minTotalDepth = 50, maxTotalDepth = 110, depthDifference = 50, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
f_df <- filterSNPs(SNPset = f_df, refAlleleFreq = 0.05, minTotalDepth = 60, maxTotalDepth = 130, depthDifference = 50, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
g_df <- filterSNPs(SNPset = g_df, refAlleleFreq = 0.05, minTotalDepth = 50, maxTotalDepth = 110, depthDifference = 50, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
h_df <- filterSNPs(SNPset = h_df, refAlleleFreq = 0.05, minTotalDepth = 20, maxTotalDepth = 90, depthDifference = 50, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
i_df <- filterSNPs(SNPset = i_df, refAlleleFreq = 0.05, minTotalDepth = 60, maxTotalDepth = 140, depthDifference = 50, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
j_df <- filterSNPs(SNPset = j_df, refAlleleFreq = 0.05, minTotalDepth = 40, maxTotalDepth = 100, depthDifference = 50, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
```

```{r depth plot, include=TRUE, fig.show="hold", fig.width=3, fig.height=3, eval=FALSE}
for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  print(ggplot(data = dt_name) +geom_histogram(aes(x = DP.HIGH + DP.LOW)))+ggtitle(obj_names[i])+ylim(0,120000)
}
#ggplot(data = a_df) +geom_histogram(aes(x = DP.HIGH + DP.LOW)) +ggtitle("a90%m")+ylim(0,120000)
```

```{r ref frequency, include=TRUE, fig.show="hold", fig.width= 3, fig.height=3, eval=FALSE}
for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  print(ggplot(data = dt_name) +geom_histogram(aes(x = REF_FRQ))+ggtitle(obj_names[i]))
}
#ggplot(data = j_df) +geom_histogram(aes(x = REF_FRQ))+ggtitle("j11%m")
```

###QTLseqAnalysis 
```{r QTLseqAnalysis, eval=FALSE}
P1sp_bsa <- runQTLseqAnalysis(P1sp_bsa,windowSize=1e6, popStruc="F1", bulkSize=c(5, 38), replications=10000, intervals=c(95, 99))
P9sp_bsa <- runQTLseqAnalysis(P9sp_bsa,windowSize=1e6, popStruc="F1", bulkSize=c(4, 44), replications=10000, intervals=c(95, 99))
P2sp_bsa <- runQTLseqAnalysis(P2sp_bsa,windowSize=1e6, popStruc="F1", bulkSize=c(5, 38), replications=10000, intervals=c(95, 99))
P10sp_bsa <- runQTLseqAnalysis(P10sp_bsa,windowSize=1e6, popStruc="F1", bulkSize=c(3, 18), replications=10000, intervals=c(95, 99))
P11sp_bsa <- runQTLseqAnalysis(P11sp_bsa,windowSize=1e6, popStruc="F1", bulkSize=c(2, 34), replications=10000, intervals=c(95, 99))
P3sp_bsa <- runQTLseqAnalysis(P3sp_bsa,windowSize=1e6, popStruc="F1", bulkSize=c(1, 37), replications=10000, intervals=c(95, 99))
P12sp_bsa <- runQTLseqAnalysis(P12sp_bsa,windowSize=1e6, popStruc="F1", bulkSize=c(3, 63), replications=10000, intervals=c(95, 99))
P4sp_bsa <- runQTLseqAnalysis(P4sp_bsa,windowSize=1e6, popStruc="F1", bulkSize=c(3, 37), replications=10000, intervals=c(95, 99))
#Russell_qtl <- runQTLseqAnalysis(Russell,windowSize = 3e6, popStruc = "F1", bulkSize = c(3,50), replications = 10000, intervals = c(95, 99))
#a_df_qtl <- runQTLseqAnalysis(a_df,windowSize = 3e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
#b_df_qtl <- runQTLseqAnalysis(b_df,windowSize = 3e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
#c_df_qtl <- runQTLseqAnalysis(c_df,windowSize = 3e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
#d_df_qtl <- runQTLseqAnalysis(d_df,windowSize = 3e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
#e_df_qtl <- runQTLseqAnalysis(e_df,windowSize = 3e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
#f_df_qtl <- runQTLseqAnalysis(f_df,windowSize = 3e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
#g_df_qtl <- runQTLseqAnalysis(g_df,windowSize = 3e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
#h_df_qtl <- runQTLseqAnalysis(h_df,windowSize = 3e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
#i_df_qtl <- runQTLseqAnalysis(i_df,windowSize = 3e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
#j_df_qtl <- runQTLseqAnalysis(j_df,windowSize = 3e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
```

```{r GprimeAnalysis}
for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  assign(obj_names[i], runGprimeAnalysis(dt_name, windowSize = 1e6, outlierFilter = "deltaSNP", filterThreshold = 0.4))
}
#ck0201_qtl <- runGprimeAnalysis(ck0201_qtl, windowSize = 2e6, outlierFilter = "deltaSNP", filterThreshold = 0.4)
```

```{r plotGprimeDist, eval=FALSE, fig.show="hold", fig.width= 5, fig.height=3}
for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  print(plotGprimeDist(SNPset = dt_name, outlierFilter = "Hampel")+ggtitle(obj_names[i]))
}
#plotGprimeDist(SNPset = ck0201_qtl, outlierFilter = "Hampel")+ggtitle("ck02_01")
```

```{r plotGprimeDist deltaSNP outlier, eval=FALSE, fig.show="hold", fig.width= 5, fig.height=3}
for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  print(plotGprimeDist(SNPset = dt_name, outlierFilter = "deltaSNP", filterThreshold = 0.4) +ggtitle(obj_names[i]))
}
#plotGprimeDist(SNPset = ck0201_qtl, outlierFilter = "deltaSNP", filterThreshold = 0.4) +ggtitle("ck02_01")
```

```{r plot QTLstats nSNPs, eval=FALSE, include=TRUE, fig.width= 20}
for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  print(plotQTLStats(SNPset = dt_name, var = "nSNPs") +ggtitle(obj_names[i]))+ylim(0,25000)
}

#plotQTLStats(SNPset = j_df_qtl, var = "nSNPs")+ggtitle("j11%m")+ylim(0,25000)
```

```{r plot QTLstats deltaSNP, eval=FALSE, include=TRUE, fig.width=20}
for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  print(plotQTLStats(SNPset = dt_name, var = "deltaSNP", plotIntervals = TRUE) +ggtitle(obj_names[i]))
}
#plotQTLStats(SNPset = ck0201_qtl, var = "deltaSNP", plotIntervals = TRUE)+ggtitle("CK02_01")
```

```{r males plotQTLstats Gprime, eval=FALSE, warnings=TRUE, include=TRUE, fig.show="hold", fig.width=20}
#trying var= "negLog10Pval" instead of "Gprime"
for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  print(plotQTLStats(SNPset = dt_name, var = "negLog10Pval", plotThreshold = TRUE, q = 0.01) + labs(title = obj_names[i]) + theme(plot.title = element_text(size = 18, hjust = 0.5)))
ggsave(paste0(obj_names[i],"_plotgp.png"), plot = last_plot(), width = 35, height = 12, units = c("cm"), dpi = 320)
}

#ck0201_plotgp <- plotQTLStats(SNPset = ck0201_qtl, var = "negLog10Pval", plotThreshold = TRUE, q = 0.01)
#ck0201_plotgp <- ck0201_plotgp + labs(title = "CK02_01") + theme(plot.title = element_text(size = 18, hjust = 0.5))
```

```{r Gprime ggplot qtls, eval=TRUE, warnings=FALSE, include=TRUE, fig.show="hold", fig.width= 20}
for(i in seq(along=obj_names)) {
dt_name   <- as.data.table(get(obj_names[i], envir = .GlobalEnv))

dt_name   <- assign(obj_names[i], dt_name[,"POSmb" := POS/1000000])
quant     <- as.numeric(quantile(dt_name[,Gprime], probs = 0.995, na.rm=T))#probs0.99 gives top 0.01% of SNPs
dt_name   <- assign(obj_names[i], dt_name[Gprime>quant, "top":="Top 0.5 % SNPs"][Gprime<quant, top:="Remaining SNPs"])

thresholdnumber <- min(dt_name[qvalue<0.05, Gprime])
thresholdnumber[is.finite(thresholdnumber)] <- round(thresholdnumber, digits = 3)
thresholdnumber[!is.finite(thresholdnumber)] <- 0
Threshold <- data.frame(Threshold = factor(thresholdnumber) )

if  (thresholdnumber==0){
    #if there is a significant threshold to plot apply hline threshold to graph
p <- ggplot(data=dt_name, aes(x=POSmb, y=Gprime))+
  labs(title = obj_names[i]) +
  geom_point(aes(color=factor(dt_name[,top])),size=0.01, na.rm=T)+
  scale_color_manual(name="Top SNPs",breaks = c("Top 0.5 % SNPs","Remaining SNPs"),values=c("chartreuse3","black"))+
  scale_x_continuous(breaks=c(0, 10,20))+
        theme(plot.title =element_text(color = "grey20", size = 10, angle = 0, hjust = .5, vjust = 0, face = "plain"),
        legend.text=element_text(color = "grey20", size = 7, angle = 0, hjust = 0, vjust = .5, face = "plain"),
        legend.title =element_text(color = "grey20", size = 8, angle = 0, hjust = 0, vjust = 0, face = "plain"),
        axis.text.x = element_text(color = "grey20", size = 7, angle = 0, hjust = .5, vjust = 0, face = "plain"),
        axis.text.y = element_text(color = "grey20", size = 7, angle = 0, hjust = 0, vjust = 0, face = "plain"),  
        axis.title.x = element_text(color = "grey20", size = 8, angle = 0, hjust = .5, vjust = 0, face = "plain"),
        axis.title.y = element_text(color = "grey20", size = 8, angle = 0, hjust = 0, vjust = .5, face = "plain"))+
  xlab("Nucleotide Position")+  
  ylab("Gprime")+
  facet_wrap(~CHROM, nrow=3, ncol=10)
  print(p)

} else {
  #else apply graph with no threshold line
  p <- ggplot(data=dt_name, aes(x=POSmb, y=Gprime))+
  labs(title = obj_names[i]) +
  geom_point(aes(color=factor(dt_name[,top])),size=0.01, na.rm=T)+
  scale_color_manual(name="Top SNPs",breaks = c("Top 0.5 % SNPs","Remaining SNPs"),values=c("chartreuse3","black"))+
  scale_x_continuous(breaks=c(0, 10, 20))+
      theme(plot.title =element_text(color = "grey20", size = 10, angle = 0, hjust = .5, vjust = 0, face = "plain"),
        legend.text=element_text(color = "grey20", size = 7, angle = 0, hjust = 0, vjust = .5, face = "plain"),
        legend.title =element_text(color = "grey20", size = 8, angle = 0, hjust = 0, vjust = 0, face = "plain"),
        axis.text.x = element_text(color = "grey20", size = 7, angle = 0, hjust = .5, vjust = 0, face = "plain"),
        axis.text.y = element_text(color = "grey20", size = 7, angle = 0, hjust = 0, vjust = 0, face = "plain"),  
        axis.title.x = element_text(color = "grey20", size = 8, angle = 0, hjust = .5, vjust = 0, face = "plain"),
        axis.title.y = element_text(color = "grey20", size = 8, angle = 0, hjust = 0, vjust = .5, face = "plain"))+
  xlab("Nucleotide Position")+  
  ylab("Gprime")+
    geom_hline(aes(yintercept=thresholdnumber, linetype = Threshold ), Threshold)+
    scale_linetype_manual(name = "Threshold (q=.05)", values = "dashed", labels = Threshold) +
  facet_wrap(~CHROM, nrow=3, ncol=10)
  print(p)
}

ggsave(paste0(obj_names[i],"_top0.5pc_Gprime_plotgp.png"), plot = last_plot(), width = 20, height = 12, units = c("cm"), dpi = 320)
}
```

```{r write top 1percent snps and sig snps so seperate plots can be made of each, eval=TRUE}
  #Delete file if it exists
if (file.exists(paste0(i,"top0.5percandsig_CK_snps.csv"))) {
file.remove(paste0(i,"top0.5percandsig_CK_snps.csv"))
} 

for(i in seq(along=obj_names)) {
dt_name <- as.data.table(get(obj_names[i], envir = .GlobalEnv))
dt_name   <- dt_name[,"parent" := obj_names[i]]
#dt_name <- dt_name[, `:=` ("AD_REF.HIGH"=NULL, "AD_ALT.HIGH"=NULL, "DP.HIGH"=NULL, "GQ.HIGH"=NULL, "PL.HIGH"=NULL, "Cpos"=NULL,  "r1CHROM"=NULL,   "r1POS"=NULL, "r1REF"=NULL, "r1ALT"=NULL, "AD_REF.LOW"=NULL, "AD_ALT.LOW"=NULL, "DP.LOW"=NULL,  "GQ.LOW"=NULL, "PL.LOW"=NULL, "SNPindex.HIGH"=NULL, "SNPindex.LOW"=NULL, "REF_FRQ"=NULL, "deltaSNP"=NULL, "nSNPs"=NULL, "tricubeDeltaSNP"=NULL, "minDP"=NULL, "tricubeDP"=NULL, "CI_95"=NULL, "CI_99"=NULL, "G"=NULL)]

dt_name   <- dt_name[qvalue>0.05, "sig" := "nonsignificant"][qvalue<0.05, sig := "significant"]

dt_name   <- dt_name[top=="Top 0.5 % SNPs" | sig=="significant",] #filter just top and significant
fwrite(dt_name, file="top0.5percandsig_CK_snps.csv", append=T) #rename this chunk for A,BC,DE,FGHI runs
}
```

```{r plot region overlap from snp data, eval=TRUE, warnings=FALSE, include=TRUE, fig.show="hold", fig.width= 15}
BSAsnps <- fread("top0.5percandsig_CK_snps.csv")
BSAsnps <- BSAsnps[top=="Top 0.5 % SNPs",]
BSAsnps <- BSAsnps[, "Chromosome" := str_replace_all(CHROM,"chr","")]
BSAsnps <- BSAsnps[, "Chromosome" := as.factor(Chromosome)]
chrs <- c("1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29")
BSAsnps <- BSAsnps[, Chromosome := fct_relevel(BSAsnps[,Chromosome], chrs, after = Inf)]
BSAsnps <- BSAsnps[parent =="P1_bsa" | parent =="P2_bsa" | parent == "P3_bsa" | parent == "P4_bsa" | parent == "P9_bsa" | parent == "P10_bsa" | parent == "P11_bsa" | parent == "P12_bsa"]

BSAsnps <- BSAsnps[, "overlap" := .N, by=c("POSmb","Chromosome")]

#calculate percentage of position along chromosomes so it can be fit as 1.x on the x axis.
BSAsnps <- BSAsnps[, "CHR_POSpercent" := paste0(Chromosome,".",POS/max(POS)), by=Chromosome]

#check for incorrect duplicates
#BSAsnpstest <- BSAsnps[overlap==4 & Chromosome==11, ]

#fwrite(BSAsnps, file="BSA0.5snps.csv")
#t<- as.character(seq(1,29,1))

#Plot region overlap
#colours1 <- c("1" = "gold", "2" = "gold3", "3" = "darkolivegreen4", "4" = "darkgreen")
#shape1 <- c("1" = 15, "2" = 20, "3" = 18, "4" = 8)
shape1 <- c("1" = 21, "2" = 22, "3" = 23, "4" = 24)
shape2 <- c("P1_bsa" = 0, "P2_bsa" = 1,"P3_bsa" =2,"P4_bsa"=3, "P9_bsa"= 4, "P10_bsa" = 5,  "P11_bsa" = 6 ,"P12_bsa"=8)
labels1 <- c("P1_bsa", "P2_bsa", "P3_bsa","P4_bsa", "P9_bsa", "P10_bsa", "P11_bsa" ,"P12_bsa" )
#BSAsnps_sorted <- BSAsnps[order(-overlap)] #sorts into order so the most frequent sites are plotted last and thus ontop.
BSAsnps_peakpoint <- BSAsnps[top=="Top 0.5 % SNPs",]
BSAsnps_peakpoint <- BSAsnps_peakpoint[, "split_chr11" := parent]#chr11 split
BSAsnps_peakpoint <- BSAsnps_peakpoint[parent=="P3_bsa" & CHROM=="chr11" & POS==17004370, split_chr11 := "P3sp_bsa_17mb"]#chr11 split 
BSAsnps_peakpoint <- BSAsnps_peakpoint[, "Gp_max" := max(Gprime), by=.(split_chr11,CHROM)]
BSAsnps_peakpoint <- BSAsnps_peakpoint[CHROM=="chr11" & POS==17004370 | POS==17004424| POS==17004333, overlap := 4][CHROM=="chr4" & POS==13028990 | POS==13028701| POS==13028743, overlap := 3][CHROM=="chr10" & POS==12925425 | POS==12923146, overlap := 2][CHROM=="chr18" & POS==17013259 | POS==17013345, overlap := 2][CHROM=="chr22" & POS==8086169 | POS==8070518 | POS==8070212, overlap := 3][CHROM=="chr2" & POS==5720903 | POS==6296002 | POS==5714634, overlap := 3][CHROM=="chr14" & POS==10825283 | POS==11436270, overlap := 2]
BSAsnps_peakpoint <- BSAsnps_peakpoint[Gp_max == Gprime,]#, by=.(split_chr11,CHROM)
BSAsnps_peakpoint <- BSAsnps_peakpoint[,parent := factor(BSAsnps_peakpoint$parent, levels=labels1)]
#BSAsnps_peakpoint <- BSAsnps_peakpoint[,"parent_number" := str_replace_all(parent, "P", "")][,parent_number := str_replace_all(parent_number, "_bsa", "")]

p <- ggplot(data=BSAsnps_peakpoint, aes(x=Chromosome, y=POSmb))+
geom_point(aes(colour=factor(parent), shape=factor(parent)), na.rm=T, size = 6, stroke = 1, position =position_jitter(seed = 45, width=0.3))+# ,size=BSAsnps_peakpoint[,overlap], , fill=factor(BSAsnps_peakpoint[,parent]), alpha=0.8
scale_shape_manual(values=shape2, labels=labels1) +
scale_colour_viridis_d(labels=labels1) +
  #scale_fill_viridis_d(guide="none") +
#guides(size=guide_legend(order=1), shape=guide_legend(order=1), color= guide_legend(order=2))+
#theme_bw()+
theme(plot.title =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  legend.text    =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = .5, face = "plain"),
  legend.title   =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = 0, face = "plain"),
  axis.text.x    =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  axis.text.y    =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = 0, face = "plain"),  
  axis.title.x   =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  axis.title.y   =element_text(color = "grey20", size = 12, angle = 90, hjust = .5, vjust = .5, face = "plain"))+
  guides(colour=guide_legend(keywidth=1, keyheight=1, default.unit="cm"), shape=guide_legend(keywidth=1, keyheight=1, default.unit="cm"))+
  #panel.grid = element_blank())+
  #labs(size=str_wrap("Number of QTL overlapping between bulk analyses",17))+
  #labs(shape=str_wrap("Number of QTL overlapping between bulk analyses",17))+  
  labs(colour=str_wrap("BSA name",17))+
  labs(shape=str_wrap("BSA name",17))+ 
  #labs(fill=str_wrap("BSA name",17))+
  scale_y_continuous(n.breaks = 8)+
  xlab("Chromosome")+
  ylab("Position (Mb)")
print(p)
ggsave("sizedQTLpeaks.png", plot = last_plot(), width = 20, height = 12, units = c("cm"), dpi = 320)



p <- ggplot(data=BSAsnps, aes(x=CHR_POSpercent, y=Gprime))+
geom_jitter(aes(colour=BSAsnps[,parent]),alpha=5/10,shape=20, na.rm=T)+
guides(color= guide_legend())+
#scale_x_continuous(breaks=seq(0,29,1))+
#scale_x_discrete(breaks=as.factor(seq(1,29,1)), labels=as.factor(seq(1,29,1)))+
scale_x_discrete(breaks=as.character(seq(1,29,1)), labels=as.character(seq(1,29,1)))+
theme(plot.title =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  legend.text    =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = .5, face = "plain"),
  legend.title   =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = 0, face = "plain"),
  axis.text.x    =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  axis.text.y    =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = 0, face = "plain"),  
  axis.title.x   =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  axis.title.y   =element_text(color = "grey20", size = 12, angle = 90, hjust = .5, vjust = .5, face = "plain"))+
  xlab("Chromosome")+
  ylab("Gprime")
print(p)
ggsave("colourparentsQTLpeakoverlap.png", plot = last_plot(), width = 20, height = 12, units = c("cm"), dpi = 320)


#or plot region overlap from snp data
p <- ggplot(data=BSAsnps, aes(x=CHR_POSpercent, y=Gprime))+
geom_point(aes(colour=BSAsnps[,overlap],size=BSAsnps[,overlap]),alpha=5/10, na.rm=T)+#,size=BSAsnps[,overlap]
scale_colour_continuous(name=str_wrap("Number of overlapping SNPs from QTL",20),low="black", high="green", limits=c(1, 4), breaks=seq(1, 4, by=1)) +
scale_size_continuous(name=str_wrap("Number of overlapping SNPs from QTL",20), limits=c(1, 4), breaks=seq(1, 4, by=1)) +
guides(color= guide_legend(), size=guide_legend())+
  
theme(plot.title =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  legend.text    =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = .5, face = "plain"),
  legend.title   =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = 0, face = "plain"),
  axis.text.x    =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  axis.text.y    =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = 0, face = "plain"),  
  axis.title.x   =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  axis.title.y   =element_text(color = "grey20", size = 12, angle = 90, hjust = .5, vjust = .5, face = "plain"))+
  xlab("Chromosome")+
  ylab("Gprime")
print(p)
ggsave("sizedQTLpeakoverlap.png", plot = last_plot(), width = 20, height = 12, units = c("cm"), dpi = 320)


#or plot region overlap from snp data
p <- ggplot(data=BSAsnps, aes(x=CHROM, y=POSmb))+
geom_point(aes(colour=BSAsnps[,sig],size=BSAsnps[,overlap]),alpha=5/10, na.rm=T)+#,size=BSAsnps[,overlap]
scale_size_continuous(name=str_wrap("Number of overlapping SNPs from QTL",20), limits=c(1, 4), breaks=seq(1, 4, by=1)) +
guides(color= guide_legend(), size=guide_legend())+
theme(plot.title =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  legend.text    =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = .5, face = "plain"),
  legend.title   =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = 0, face = "plain"),
  axis.text.x    =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  axis.text.y    =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = 0, face = "plain"),  
  axis.title.x   =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  axis.title.y   =element_text(color = "grey20", size = 12, angle = 90, hjust = .5, vjust = .5, face = "plain"))+
    labs(colour=str_wrap("Repeated measures (q-0.05) result",20))+
  xlab("Chromosome")+
  ylab("Gprime")
print(p)
ggsave("sizedQTLpeakoverlapsigs.png", plot = last_plot(), width = 20, height = 12, units = c("cm"), dpi = 320)



# plot only snps that were in 3pools
BSA6snps <- BSAsnps[overlap>=3,]
p <- ggplot(data=BSA6snps, aes(x=CHROM, y=POSmb))+
geom_jitter(aes(colour=BSA6snps[,sig]),alpha=5/10, na.rm=T, size=0.5)+
  theme(plot.title =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  legend.text    =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = .5, face = "plain"),
  legend.title   =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = 0, face = "plain"),
  axis.text.x    =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  axis.text.y    =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = 0, face = "plain"),  
  axis.title.x   =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  axis.title.y   =element_text(color = "grey20", size = 12, angle = 90, hjust = .5, vjust = 0, face = "plain"))+
  labs(colour=str_wrap("Repeated measures (q-0.05) result",20))+
  xlab("Chromosome")+
  ylab("Position (Mb)")
print(p)

ggsave("3x4xoverlap_QTLpeakoverlap.png", plot = last_plot(), width = 20, height = 12, units = c("cm"), dpi = 320)
```

```{r just use the BSA3snps positions without the bedfile since there arent many, eval=FALSE}
BSA3snps <- fread("BSA0.05snps.csv")
fwrite(BSA3snps, file="BSA0.05_3snps.csv", append=F)
BSA3snps <- BSA3snps[overlap>=3,]
BSA3snps <- unique(BSA3snps[, c("CHROM", "POS", "ALT")])
write.table(BSA3snps, file="BSA3snps.csv",quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")

```



```{r make bedfile of the snps in 3-4 pools for analysis, eval=FALSE}
BSA3snps <- fread("BSAsnps.csv")
BSA3snps <- BSA3snps[overlap>=3,]
minregions <- BSA3snps[, .("min"=min(POS)), by=.(parent,CHROM)]
maxregions <- BSA3snps[, .("max"=max(POS)), by=.(parent,CHROM)]
BSA3regions <- cbind(minregions,maxregions)
BSA3regions <- BSA3regions[,c(2,3,6)]
rm(minregions,maxregions)
write.table(BSA3regions, file="BSAregions_with3commonsnps.bed",quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")

#just use the BSA3snps positions without the bedfile since there arent many


```


```{r regions of various sizes based off peaks for narrower bedfiles for each peak}
BSAsnps <- fread("top0.5percandsig_CK_snps.csv")
BSAsnps <- BSAsnps[top=="Top 0.5 % SNPs",]
BSAsnps <- BSAsnps[, "Gp_max" := max(Gprime), by=.(parent,CHROM)]
#give second Gp_max at max GP for P3 chr11 which has 2 peack on the same chr to split
BSAsnpst<- BSAsnps[parent=="P3sp_bsa" & CHROM=="chr11" & POS==17004370, Gp_max := Gprime]

BSAsnpsunique <- unique(BSAsnps, by= c("Gp_max","parent","CHROM"))

splist <- c("P1sp_bsa","P2sp_bsa","P3sp_bsa","P4sp_bsa","P9sp_bsa","P10sp_bsa","P11sp_bsa","P12sp_bsa")
regions_G5000 <- BSAsnpsunique[, "G5000l" := POS - 5000, by=.(parent,CHROM,Gp_max)
                              ][, "G5000h" := POS + 5000, by=.(parent,CHROM,Gp_max)]
regions_G5000 <- regions_G5000[, "qtl_name" := paste(str_replace_all(parent, c("_bsa" = "")),CHROM,POS,sep="_")]
BSAregions_G5000 <- regions_G5000[, .(CHROM, G5000l,G5000h, qtl_name, parent)]

for (i in splist){
  assign(paste0("BSAregions_G5000_",{i}), BSAregions_G5000[parent== {i}, c(1,2,3,4)])
  dt_name <- get(paste0("BSAregions_G5000_",{i}), envir = .GlobalEnv)
  write.table(dt_name, file=paste0("BSAregions_G5000_",{i}, ".bed"),quote=FALSE, col.names=FALSE, row.names=FALSE, sep="\t")
}

regions_G10000 <- BSAsnpsunique[, "G10000l" := POS - 10000, by=.(parent,CHROM,Gp_max)
                               ][, "G10000h" := POS + 10000, by=.(parent,CHROM,Gp_max)]
regions_G10000 <- regions_G10000[, "qtl_name" := paste(str_replace_all(parent, c("_bsa" = "")),CHROM,POS,sep="_")]
BSAregions_G10000 <- regions_G10000[, .(CHROM, G10000l,G10000h, qtl_name, parent)]

for (i in splist){
  assign(paste0("BSAregions_G10000_",{i}), BSAregions_G10000[parent== {i}, c(1,2,3,4)])
  dt_name <- get(paste0("BSAregions_G10000_",{i}), envir = .GlobalEnv)
  write.table(dt_name, file=paste0("BSAregions_G10000_",{i}, ".bed"),quote=FALSE, col.names=FALSE, row.names=FALSE, sep="\t")
}

regions_G50000 <- BSAsnpsunique[, "G50000l" := POS - 50000, by=.(parent,CHROM,Gp_max)
                               ][, "G50000h" := POS + 50000, by=.(parent,CHROM,Gp_max)]
regions_G50000 <- regions_G50000[, "qtl_name" := paste(str_replace_all(parent, c("_bsa" = "")),CHROM,POS,sep="_")]
BSAregions_G50000 <- regions_G50000[, .(CHROM, G50000l,G50000h, qtl_name, parent)]

for (i in splist){
  assign(paste0("BSAregions_G50000_",{i}), BSAregions_G50000[parent== {i}, c(1,2,3,4)])
  dt_name <- get(paste0("BSAregions_G50000_",{i}), envir = .GlobalEnv)
  write.table(dt_name, file=paste0("BSAregions_G50000_",{i}, ".bed"),quote=FALSE, col.names=FALSE, row.names=FALSE, sep="\t")
}

regions_G100000 <- BSAsnpsunique[, "G100000l" := POS - 100000, by=.(parent,CHROM,Gp_max)
                                ][, "G100000h" := POS + 100000, by=.(parent,CHROM,Gp_max)]
regions_G100000 <- regions_G100000[, "qtl_name" := paste(str_replace_all(parent, c("_bsa" = "")),CHROM,POS,sep="_")]
BSAregions_G100000 <- regions_G100000[, .(CHROM, G100000l,G100000h, qtl_name, parent)]

for (i in splist){
  assign(paste0("BSAregions_G100000_",{i}), BSAregions_G100000[parent== {i}, c(1,2,3,4)])
  dt_name <- get(paste0("BSAregions_G100000_",{i}), envir = .GlobalEnv)
  write.table(dt_name, file=paste0("BSAregions_G100000_",{i}, ".bed"),quote=FALSE, col.names=FALSE, row.names=FALSE, sep="\t")
}

# regions_G500000 <- BSAsnpsunique[, "G500000l" := POS - 500000, by=.(parent,CHROM,Gp_max)
#                                 ][, "G500000h" := POS + 500000, by=.(parent,CHROM,Gp_max)]
# regions_G500000 <- regions_G500000[, "qtl_name" := paste(str_replace_all(parent, c("_bsa" = "")),CHROM,POS,sep="_")]
# BSAregions_G500000 <- regions_G500000[, .(CHROM, G500000l,G500000h, qtl_name, parent)]

# for (i in splist){
#   assign(paste0("BSAregions_G500000_",{i}), BSAregions_G500000[parent== {i}, c(1,2,3,4)])
#   dt_name <- get(paste0("BSAregions_G500000_",{i}), envir = .GlobalEnv)
#   write.table(dt_name, file=paste0("BSAregions_G500000_",{i}, ".bed"),quote=FALSE, col.names=FALSE, row.names=FALSE, sep="\t")
# }
#continued at QTLseqR/10.QTLseqrPSAqtl/QTLseqrPSAqtl.Rmd
```

# old bed regions not used to split chrs with more than one peak at the 1% level if the 1% level is needed .
```{r read all top1% threshold snps and write to bedfile, eval=FALSE}
BSAsnps <- fread("top1percandsig_CK_snps.csv")
BSAsnps <- BSAsnps[top=="Top 1 % SNPs",]
minregions <- BSAsnps[, .("min"=min(POS)), by=.(parent,CHROM)]
maxregions <- BSAsnps[, .("max"=max(POS)), by=.(parent,CHROM)]

BSAregions <- cbind(minregions,maxregions)
BSAregions <- BSAregions[,c(1,2,3,6)]
rm(minregions,maxregions)

#manually split chromosome regions with multiple peaks.
#finder <- BSAsnps[parent=='P2_bsa',][CHROM=='chr27',]
#qplot(finder$POS)
BSAregions <- BSAregions[min!='4287366']
BSAregions <-rbind(BSAregions, list('P11_bsa', 'chr22', '4287366', '9999291'))
BSAregions <-rbind(BSAregions, list('P11_bsa', 'chr22', '16057920', '18197508'))
BSAregions <- BSAregions[min!='9842636']
BSAregions <-rbind(BSAregions, list('P10_bsa', 'chr17', '9842636', '11005178'))
BSAregions <-rbind(BSAregions, list('P10_bsa', 'chr17', '12928921', '13606201'))
BSAregions <-BSAregions[min!='7902242']
BSAregions <-rbind(BSAregions, list('P9_bsa', 'chr27', '7902242', '8923443'))
BSAregions <-rbind(BSAregions, list('P9_bsa', 'chr27', '12634193', '12805267'))
BSAregions <-BSAregions[min!='7316320']
BSAregions <-rbind(BSAregions, list('P9_bsa', 'chr22', '7316320', '8399998'))
BSAregions <-rbind(BSAregions, list('P9_bsa', 'chr22', '17461139', '18115317'))
BSAregions <-BSAregions[min!='4738191']
BSAregions <-rbind(BSAregions, list('P4_bsa', 'chr22', '4738191', '8801011'))
BSAregions <-rbind(BSAregions, list('P4_bsa', 'chr22', '17851237', '18018931'))
BSAregions <-BSAregions[min!='6866334']
BSAregions <-rbind(BSAregions, list('P3_bsa', 'chr11', '6866334', '8863236'))
BSAregions <-rbind(BSAregions, list('P3_bsa', 'chr11', '16667753', '17004370'))
BSAregions <-BSAregions[min!='4755']
BSAregions <-rbind(BSAregions, list('P2_bsa', 'chr27', '4755', '448377'))
BSAregions <-rbind(BSAregions, list('P2_bsa', 'chr27', '16046728', '16922169'))
BSAregions <- BSAregions[,c(2,3,4)]
write.table(BSAregions, file="allBSAregions.bed",quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t") #the sample name wont work with wide format vcffiles containing multiple samples especially since these overlap.

#seperate bed files into chromosomes to keep ram load down before selecting SNP sites
chrs <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chr23","chr24","chr25","chr26","chr27","chr28","chr29")

for (i in seq_along(chrs)){
write.table(BSAregions[CHROM==chrs[i],], file=paste0("allBSAregions_", chrs[i], ".bed"),quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
}

```

##male percentages from sorting bulks and working out contributions.
```{r male percentages}
mp <- fread("male_percentage.txt")
mp1 <- mp[1:36,]
p <- ggplot(data=mp1, aes(x=row, y=percent))+
geom_point(aes(colour=factor(qtl),fill=factor(qtl), size=0.1))+#,alpha=5/10, , size=0.5,,shape=factor(threshold) position=position_dodge(width=1)
  scale_colour_manual(values = c("tomato", "darkgreen"))+
  #geom_hline(yintercept = 20)+
  #geom_text(aes(0,20,label = "80% accuracy", vjust = -1, hjust = 0.1))+
  scale_y_continuous(labels = function(x) paste0(x, "%"))+
  labs(title =str_wrap("QTL detection using the top 0.5 % of Gprime values as a threshold",40))+
  theme(plot.title =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  legend.text    =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = .5, face = "plain"),
  legend.title   =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = .5, face = "plain"),
  legend.position="bottom",
  axis.text.x    =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  axis.text.y    =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = 0, face = "plain"),  
  axis.title.x   =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  axis.title.y   =element_text(color = "grey20", size = 12, angle = 90, hjust = .5, vjust = 0, face = "plain"))+
  guides(size = "none")+
  labs(colour=str_wrap("QTL detected on chromosome 25",200),fill=str_wrap("QTL detected on chromosome 25",200))+
  xlab("Bulk number")+
  ylab(str_wrap("The expected difference in sex locus allele frequency between bulks",35))
print(p)

ggsave("Perc_dif_between_bulks_top5.png", plot = p, width = 15, height = 10, units = c("cm"), dpi = 320)

mp2 <- mp[37:72,]
p <- ggplot(data=mp2, aes(x=row, y=percent))+
geom_point(aes(colour=factor(qtl),fill=factor(qtl), size=0.1))+#,alpha=5/10, na.rm=T, size=0.5, position=position_dodge(width=1)
  scale_colour_manual(values = c("tomato", "darkgreen"))+
  #geom_hline(yintercept = 43)+
  #geom_text(aes(0,43,label = "80% accuracy", vjust = -1, hjust = 0.1))+
  scale_y_continuous(labels = function(x) paste0(x, "%"))+
  labs(title="QTL detection using q=0.05 as a threshold")+
  theme(plot.title =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  legend.text    =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = .5, face = "plain"),
  legend.title   =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = .5, face = "plain"),
  legend.position="bottom",
  axis.text.x    =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  axis.text.y    =element_text(color = "grey20", size = 12, angle = 0, hjust = 0, vjust = 0, face = "plain"),  
  axis.title.x   =element_text(color = "grey20", size = 12, angle = 0, hjust = .5, vjust = 0, face = "plain"),
  axis.title.y   =element_text(color = "grey20", size = 12, angle = 90, hjust = .5, vjust = 0, face = "plain"))+
  guides(size = "none")+
  labs(colour=str_wrap("QTL detected on chromosome 25",200),fill=str_wrap("QTL detected on chromosome 25",200) )+
  xlab("Bulk number")+
  ylab(str_wrap("The expected difference in sex locus allele frequency between bulks",35))
print(p)

ggsave("Perc_dif_between_bulks_p05.png", plot = p, width = 15, height = 10, units = c("cm"), dpi = 320)

```

