---
title: "malepoolmpile"
author: "Casey Flay"
date: "30/09/2020"
output:
  html_document: default
  word_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(RLinuxModules)
library(data.table)
library(glue)
library(here)
library(tidyverse)
library(magrittr)
library(GenomicRanges)
library(tools)
library(future)
library(future.batchtools)
library(here)
plan(batchtools_openlava, resources = list(threads=1))
setDTthreads(1)
module("load openlava asub/2.2 bwa/0.7.12 samtools/1.9 bcftools/1.10.2 ")#load old samtools module loaded for popoolation2 perlbrew/0.76
here()
```
**parent files pulled from database, merged, sorted and mpiled with bcftools**
```{r parents NCBI aligned bamfiles from male pool and testpools}
Rualn_bamDir <- "/output/genomic/plant/Actinidia/chinensis/Resequencing/Alignments/Russell_V2a.chromosomes.and.unassigned"
Rualn_bams  <- as.data.table(grep(paste('.bam$', collapse = "|"), list.files(path=Rualn_bamDir), invert=FALSE, value=TRUE))
Rualn_bams  <- Rualn_bams[, lapply(.SD, gsub, pattern = ("_Russell_V2a"), replacement = "&_Russell_V2a&"), .SDcols = colnames(Rualn_bams)][, c("er","reference",  "ext") := tstrsplit(V1, "&", fixed=TRUE)]
ER  <- Rualn_bams[, lapply(.SD, gsub, pattern = ("ER\\d{5}_"), replacement = ""), .SDcols = colnames(Rualn_bams)]
ER  <- ER[, lapply(.SD, gsub, pattern = ("_ER\\d{5}"), replacement = ""), .SDcols = colnames(Rualn_bams)]
ER  <- ER[, lapply(.SD, gsub, pattern = ("SRA_"), replacement = ""), .SDcols = colnames(Rualn_bams)]
ER  <- ER[, lapply(.SD, gsub, pattern = ("UNK_"), replacement = ""), .SDcols = colnames(Rualn_bams)]
setnames(ER, c("er"), c("PedigreeItem"))
Rualn_bams  <-cbind(Rualn_bams$er, ER)
ER  <- Rualn_bams[, c(1,3,4,5)]

ER  <- ER[, "er1" := str_extract(Rualn_bams$V1, "ER\\d{5}_")][, "er2" := str_extract(Rualn_bams$V1, "_ER\\d{5}")][, "er3" := str_extract(Rualn_bams$V1, "SRA_")][, "er4" := as.data.table(str_extract(Rualn_bams$V1, "UNK_"))]
ER  <- ER[, "er5" := paste0(er1,er2,er3,er4)]
ER5  <- ER[, lapply(.SD, gsub, pattern = ("NA"), replacement = ""), .SDcols = "er5"]
Rualn_bams  <-cbind(ER, ER5)
Rualn_bams  <- Rualn_bams[, c(1,2,3,4,10)]
Rualn_bams <- Rualn_bams[, "full" := paste0(er5,PedigreeItem,reference,ext)]
Rualn_bams <- Rualn_bams[, "dir" := Rualn_bamDir]
Rualn_bams <- Rualn_bams[, "pathandfile" := paste(dir, full, sep="/")]
t<- as.data.table(Rualn_bams$full)

Rualn_bams <- Rualn_bams[, "row" := seq_along(full)]
Rualn_bams <- Rualn_bams[PedigreeItem=="Red19", PedigreeItem := ("Red 19")]
Rualn_bams <- Rualn_bams[, "isfile" := (file.exists(pathandfile))]
Rualn_bams <- Rualn_bams[isfile != FALSE,]
rm(ER,ER5)
```

```{r}
#male pool parents are all the same for each pool e.g A1, B2 etc. unless adding multiple copies of each.
Russell_tp_bams <- Rualn_bams[PedigreeItem == 'CK02_04' | PedigreeItem == 'CK06_01' | PedigreeItem == 'CK09_06'| PedigreeItem == 'CK13_02'| PedigreeItem =='CK20_01'| PedigreeItem =='CK51_06'| PedigreeItem =='Russell'][, `:=` ("pool_parent"= 'Russell', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]# only found "CK06_01","CK20_01","CK51_06". missing 'CK02_04', 'CK13_02','CK09_06'.

#adjusted russell to a single input as it should be.
Russell_bam <- Rualn_bams[V1 =='ER10522_Russell'][, `:=` ("pool_parent"= 'Russell', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]

#Full pool parents these tables are also needed in the merge step but cant be used in a single table.
CK02_01_tp_bams <- Rualn_bams[PedigreeItem=='CK02_01'| PedigreeItem=='CK15_03'| PedigreeItem=='CK17_02'| PedigreeItem=='CK20_01'| PedigreeItem=='Bejing #36'| PedigreeItem == 'CK51_01'][, `:=` ("pool_parent"= 'CK02_01', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK02_01
CK10_02_tp_bams <- Rualn_bams[PedigreeItem=='CK10_02'| PedigreeItem=='CK17_03'| PedigreeItem=='CK18_01'| PedigreeItem=='CK19_03'| PedigreeItem=='CK51_09'][, `:=` ("pool_parent"= 'CK10_02', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK10_02
CK19_03_tp_bams <- Rualn_bams[PedigreeItem=='CK19_03'| PedigreeItem=='CK10_02'| PedigreeItem=='CK10_05'| PedigreeItem=='CK13_02'| PedigreeItem=='CK15_04'| PedigreeItem=='Bejing #36'| PedigreeItem=='CK51_01'][, `:=` ("pool_parent"= 'CK19_03', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK19_03
CK20_01_tp_bams <- Rualn_bams[PedigreeItem=='CK20_01'| PedigreeItem=='CK02_01'| PedigreeItem=='CK09_11'| PedigreeItem=='CK10_03'][, `:=` ("pool_parent"= 'CK20_01', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK20_01
CK22_03_tp_bams <- Rualn_bams[PedigreeItem=='CK22_03'| PedigreeItem=='CK23_08'| PedigreeItem=='CK51_09'][, `:=` ("pool_parent"= 'CK22_03', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK22_03
CK23_08_tp_bams <- Rualn_bams[PedigreeItem=='CK23_08'| PedigreeItem=='CK01_01_01_01'| PedigreeItem=='CK15_03'| PedigreeItem=='CK16_01'| PedigreeItem=='CK22_03'][, `:=` ("pool_parent"= 'CK23_08', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK23_08
CK51_01_tp_bams <- Rualn_bams[PedigreeItem=='CK51_01'| PedigreeItem=='CK02_01'| PedigreeItem=='CK17_03'| PedigreeItem=='CK19_03'][, `:=` ("pool_parent"= 'CK51_01', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK51_01
CK51_09_tp_bams <- Rualn_bams[PedigreeItem=='CK51_09'| PedigreeItem=='CK02_04'| PedigreeItem=='CK06_01'| PedigreeItem=='CK10_02'| PedigreeItem=='CK15_04'| PedigreeItem=='CK22_03'][, `:=` ("pool_parent"= 'CK51_09', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK51_09
parent_bams <- rbindlist(list(Russell_tp_bams, CK02_01_tp_bams, CK10_02_tp_bams, CK19_03_tp_bams, CK20_01_tp_bams, CK22_03_tp_bams, CK23_08_tp_bams, CK51_01_tp_bams, CK51_09_tp_bams))

parent_bams[, "psymlink_dir" := "/powerplant/workspace/hrtcdf/github/FSTs/parentpools"]
parent_bams <- parent_bams[er5=="ER10508_" |er5=="ER10522_",][, "pbams" := paste0(er5,PedigreeItem,reference, ext, sep="")]

#file.symlink(unique(parent_bams$pathandfile), here('QTLseqR'))
```

```{r index}
bamin          <- paste0(unique(parent_bams$pbams))
file.exists(bamin)
for (i in bamin){
cmd        <- glue::glue("samtools index {i}")
Log          <- "80" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
#system(bsub_cmd)
}
```

```{r picard merge bams into parent pools keeps read groups for later seperation which i dont want using samtools instead.}
# for (i in unique(parent_bams$pool_parent)){ 
# x       <- parent_bams[pool_parent=={i}, pbams]
# infile  <- str_c(paste("I=", x, sep=""), sep="", collapse=" ")
# outfile <- paste({i}, "_mrgdparents.bam", sep="")
# file.exists(x)
# cmd         <- glue("java -jar /software/bioinformatics/picard-tools-2.9.4/picard.jar MergeSamFiles {infile} O={outfile}")
# Log          <- "92" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
# bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
# #system(bsub_cmd)
# }
```

```{r samtools 1.9 merge}
for (i in unique(parent_bams$pool_parent)){ 
infile       <- parent_bams[pool_parent=={i}, pbams]
outfile <- paste({i}, "_mrgdparents.bam", sep="")

file.exists(infile)
cmd         <- glue("samtools merge -f -r {outfile} {infile}")
Log          <- "merge92b" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
system(bsub_cmd)
}                                           
```

```{r sort }
parent_bams <- parent_bams[, `:=` ("mrgedparents"= paste0(parent_bams$pool_parent, "_mrgdparents", sep=""), "bam"=".bam")]
bamin          <- parent_bams[, unique(mrgedparents)]

for (i in bamin){
cmd          <- glue::glue("samtools sort -o {i}_sort.bam {i}.bam")
Log          <- "si103" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
system(bsub_cmd)
}
```

```{r index }
bamin          <- parent_bams[, unique(mrgedparents)]
for (i in bamin){
cmd        <- glue::glue("samtools index {i}_sort.bam")
Log          <- "index116" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
system(bsub_cmd)
}
```

```{r run bcftools mpileup}
#-r 'chr25' #
submit_mpileup <- function(reference = "*.fa.gz", output = c(), bamfiles = c(), job.init = "asub", job.opts = "-q normal") {
  cmd          <- "bcftools mpileup -B -I -Q 20 -f {reference} --excl-flags 'UNMAP,SECONDARY,QCFAIL,DUP' -a {mpileuptags} {bamfiles} -Ou  |bcftools call -f {bcfcalltags} -p {pvalThres} -m -A -Oz -o {output} && bcftools index {output}" 
  tmpf         <- tempfile(pattern = "submit_mileup.", fileext = ".sh")
  writeLines(text = c(glue(cmd)), con = tmpf)
  submit       <- "{job.init} {job.opts} {tmpf}"
  system(glue(submit), intern = TRUE)
}
mpileuptags    <- "DP,AD,SP,INFO/AD"
bcfcalltags    <- "GQ,GP"
pvalThres      <- 0.99
PedigreeItem   <- parent_bams[, unique(pool_parent)]
bamfiles       <- parent_bams[, unique(paste0(mrgedparents, "_sort.bam"))]
output         <- paste("bcfmpile_parents_", PedigreeItem, ".vcf", sep = "")
reference      <- dir("/workspace/hrards/projects/hrtcdf/genome/fasta", pattern = ".*\\.gz$", full.names = TRUE)
file.exists(bamfiles)
file.exists(reference)
submit_mpileup(reference = reference, output = output, bamfiles = bamfiles)
```

```{r sort }
bamin        <- paste("bcfmpile_parents_", parent_bams[, unique(pool_parent)], ".vcf", sep = "")
output       <- parent_bams[, unique(mrgedparents)]      
          
for (i in bamin){
cmd          <- glue::glue("samtools sort -o {i}_sort.bam {i}.bam")
Log          <- "si103" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
system(bsub_cmd)
}
```

***male pools and sample pools from hrtcdf/github/FSTs/malepool/malepool.Rmd***
```{r symlink male pool bam files}
royDir <- "/workspace/hrards/projects/hrtcdf/10.alignments/bwa"
outDir <- "/powerplant/workspace/hrtcdf/github/FSTs/QTLseqR"
mspoolfiles       <- dir(royDir, pattern = ".bam", full.names = TRUE, include.dirs = TRUE)

#file.symlink(mspoolfiles, outDir)
```

```{r transform filenames to sample, lane and suffix for data.table from symlink}
mpooldt_bam <- as.data.table(grep(paste("^HW5YYDSXX"), list.files(path="."), invert=FALSE, value=TRUE))
mpooldt_bam <- mpooldt_bam[, "pooltype" := "variablemale"]
spooldt_bam <- as.data.table(grep(paste("^HL55WDRXX"), list.files(path="."), invert=FALSE, value=TRUE))
spooldt_bam <- spooldt_bam[, "pooltype" := "lowPSA"]
pooldt_bam  <- rbind(mpooldt_bam, spooldt_bam)

pooldt_bam <- pooldt_bam[, V2 := str_replace_all(V1, "_L" , "&_L")
               ][, V2 := str_replace_all(V2, "XX_" , "XX_&")
               ][, V2 := str_replace_all(V2, ".bam" , "&.bam") 
               ][, c("H", "sample", "lane", "ftype") := tstrsplit(V2, "&", fixed=TRUE)
               ][, V2:=NULL]
```

```{r index}
bamin <- pooldt_bam[ftype==".bam", V1]

for (i in bamin){
cmd         <- glue::glue("samtools index {i}")
Log          <- "170" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
system(bsub_cmd)
}
```

```{r picard merge lanes for sample lowPSA bam files}
for (i in unique(pooldt_bam$sample)){ 
x       <- pooldt_bam[sample=={i} & ftype==".bam", V1]
infile  <- str_c(paste("I=", x, sep=""), sep="", collapse=" ")
outfile <- paste({i}, "_picmrg_lowPSA.bam", sep="")

file.exists(x)
cmd         <- glue("java -jar /software/bioinformatics/picard-tools-2.9.4/picard.jar MergeSamFiles --CREATE_INDEX {infile} O={outfile}")
Log          <- "78" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
#system(bsub_cmd)
}
```

```{r}
picmrg_lowPSApooldt_bam <- as.data.table(paste0(unique(pooldt_bam$sample), "_picmrg_lowPSA.bam"))
mpilepoolbams <- picmrg_lowPSApooldt_bam[, `:=` (sample= str_replace_all(V1, "_picmrg_lowPSA.bam", ""), "ftype"=".bam")]
mpilepoolbams <- mpilepoolbams[, `:=` (sample= str_replace_all(V1, "_picmrg_lowPSA.bam", ""), "ftype"=".bam", "sample2"=sample)]
mpilepoolbams <- mpilepoolbams[, `:=` (sample2= str_replace_all(sample2, "CK19_03B", "CK19_03"))]
#variablemale     <- pooldt_bam[pooltype=="variablemale", .(V1, sample, ftype)]
#mpilepoolbams    <- rbind(picmrg_lowPSApooldt_bam, variablemale)
```

```{r sort}
bamin     <- mpilepoolbams[ftype==".bam",V1]
for (i in bamin){
cmd       <- glue::glue("samtools sort -o s_{i} {i}")
Log       <- "si298";Ol<-paste0(here('PSApopoolation/logD/'),Log,".out");Oe<-paste0(here('PSApopoolation/logD/'),Log,".err")
bsub_cmd  <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
system(bsub_cmd)
}
```

```{r index}
bamin     <- mpilepoolbams[ftype==".bam",V1]
for (i in bamin){
cmd       <- glue::glue("samtools index s_{i}")
Log       <- "i232";Ol<-paste0(here('PSApopoolation/logD/'),Log,".out");Oe<-paste0(here('PSApopoolation/logD/'),Log,".err")
bsub_cmd  <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
system(bsub_cmd)
}
mpilepoolbams     <- mpilepoolbams[ftype==".bam",'s_' := "s_"]
```

```{r run bcftools mpileup for chr25 for QTLseqR}
submit_mpileup <- function(reference = "*.fa.gz", output = c(), bamfiles = c(), job.init = "asub", job.opts = "-q normal") {
  cmd          <- "bcftools mpileup -B -I -Q 20 -f {reference} --excl-flags 'UNMAP,SECONDARY,QCFAIL,DUP' -a {mpileuptags} {bamfiles} -Ou | bcftools call  -f {bcfcalltags} -p {pvalThres} -m -v -A -Oz -o {output} && bcftools index -f {output}" 
  tmpf         <- tempfile(pattern = "submit_mileup.", fileext = ".sh")
  writeLines(text = c(glue(cmd)), con = tmpf)
  submit       <- "{job.init} {job.opts} {tmpf}"
  system(glue(submit), intern = TRUE)
}
mpileuptags    <- "DP,AD,SP,INFO/AD"
bcfcalltags    <- "GQ,GP"
pvalThres      <- 0.99
PedigreeItem   <- mpilepoolbams[ftype==".bam",sample2]
bamfiles       <- mpilepoolbams[ftype==".bam",paste0("s_", V1)]
output         <- paste("bcfmpile_samples_", PedigreeItem, ".vcf", sep = "")
reference      <- dir("/workspace/hrards/projects/hrtcdf/genome/fasta", pattern = ".*\\.gz$", full.names = TRUE)
file.exists(bamfiles)
file.exists(reference)
submit_mpileup(reference = reference, output = output, bamfiles = bamfiles)
```

```{r remove unneeded files}
r1     <- paste0(merged_parent_bams$mrgedparents)
bamin1          <- paste(mpilepoolbams$V1)
bamin2          <- paste(pooldt_bam$V1)
bai <- grep(paste("*.bai"), list.files(path="."), invert=FALSE, value=TRUE)

#file.remove(r1, bamin1, bamin2, bai)
```

***make sure the correct russell is used for the parent in the male pools***

```{r index}
cmd          <- "samtools index 'ER10522_Russell_Russell_V2a.chromosomes.and.unassigned.fsa.bam'"
Log          <- "i273" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
system(bsub_cmd)
```


```{r sort}
cmd       <- glue::glue("samtools sort -o 's_ER10522_Russell_Russell_V2a.chromosomes.and.unassigned.fsa.bam' 'ER10522_Russell_Russell_V2a.chromosomes.and.unassigned.fsa.bam'")
Log       <- "s280";Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd  <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
system(bsub_cmd)
```

```{r index}
cmd       <- glue::glue("samtools index 's_ER10522_Russell_Russell_V2a.chromosomes.and.unassigned.fsa.bam'")
Log       <- "i287";Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd  <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
system(bsub_cmd)
```

```{r run bcftools mpileup for male pools QTLseqR}
submit_mpileup <- function(reference = "*.fa.gz", output = c(), bamfiles = c(), job.init = "asub", job.opts = "-q normal") {
  cmd          <- "bcftools mpileup -B -I -Q 20 -f {reference} --excl-flags 'UNMAP,SECONDARY,QCFAIL,DUP' -a {mpileuptags} {bamfiles} -Ou | bcftools call  -f {bcfcalltags} -p {pvalThres} -m -v -A -Oz -o {output} && bcftools index -f {output}" 
  tmpf         <- tempfile(pattern = "submit_mileup.", fileext = ".sh")
  writeLines(text = c(glue(cmd)), con = tmpf)
  submit       <- "{job.init} {job.opts} {tmpf}"
  system(glue(submit), intern = TRUE)
}
mpileuptags    <- "DP,AD,SP,INFO/AD"
bcfcalltags    <- "GQ,GP"
pvalThres      <- 0.99
bamfiles       <- 's_ER10522_Russell_Russell_V2a.chromosomes.and.unassigned.fsa.bam'
output         <- 'Russell_for_malepools.vcf.gz'
reference      <- dir("/workspace/hrards/projects/hrtcdf/genome/fasta", pattern = ".*\\.gz$", full.names = TRUE)
file.exists(bamfiles)
file.exists(reference)
submit_mpileup(reference = reference, output = output, bamfiles = bamfiles)
```

**split files for male pools into chromosomes and recombine after filtering for low parent bulks**
```{r bamfiles made in PSApopoolationfiles.Rmd, eval=FALSE}
bcfmpile_parents     <- as.data.table(grep(paste("bcfmpile_parents"), list.files(path="."), invert=FALSE, value=TRUE))
bcfmpile_samples     <- as.data.table(grep("bcfmpile_samples", list.files(path="."), invert=FALSE, value=TRUE))
bcfmpile_mrgdparents <- as.data.table(grep("mrgdparents_sort", list.files(path="."), invert=FALSE, value=TRUE))
abfiles              <- rbindlist(list(bcfmpile_parents,bcfmpile_samples,bcfmpile_mrgdparents))

abfiles <- abfiles[, V2 := str_replace_all(V1, ".vcf$" , "&.vcf")
                 ][, V2 := str_replace_all(V2, "bcfmpile_parents_" , "bcfmpile_parents_&")
                 ][, V2 := str_replace_all(V2, "bcfmpile_samples_" , "bcfmpile_samples_&")
                 ][, V2 := str_replace_all(V2, "_chr25" , "&_chr25")
                 ][, c("V3", "x2", "x3", "x4") := tstrsplit(V2, "&", fixed=TRUE)]
abfiles <- abfiles[x3=="_chr25",]
#abfiles <- abfiles[V3=="Russell", V1 := "Russell_mrgdparents.bam"]
abfiles <-  abfiles[x2=="A1", "malepercentage" := "A1m90"][x2=="B2", "malepercentage" := "B2m75"][x2=="C3", "malepercentage" := "C3m60"][x2=="D6", "malepercentage" := "D6m44"][x2=="E11", "malepercentage" := "E11m80"][x2=="F12", "malepercentage" := "F12m50"][x2=="G13", "malepercentage" := "G13m45"][x2=="H14", "malepercentage" := "H14m68"][x2=="I15", "malepercentage" := "I15m21"][x2=="J16", "malepercentage" := "J16m11"]
abfiles <-  abfiles[, "dir" := "/powerplant/workspace/hrtcdf/github/FSTs/QTLseqR/"]
mps1 <-     abfiles[malepercentage %in% na.omit(abfiles$malepercentage),]
mps2 <-     abfiles[V1=="bcfmpile_samples_Russell_chr25.vcf", ][,malepercentage :="Russell"]
mp  <-     rbind(mps1,mps2)
rm(abfiles, bcfmpile_parents,bcfmpile_mrgdparents, bcfmpile_samples, mps1, mps2)
```

```{r makes a data file for each chromosome for each file, eval=FALSE}
vcfin   <- mp[,V1]
chrs   <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chr23","chr24","chr25","chr26","chr27","chr28","chr29")
for (i in vcfin){
  for(x in chrs){
  cmd          <- glue::glue("bcftools query {i} -r '{x}' -e 'INFO/DP<20 || INFO/DP>200' -f '[[%CHROM]]\t%POS\t%REF\t%ALT\t%AD{{0}}\t%AD{{1}}\t%DP[[\t%GQ\t%PL]] \n' -o '{i}_{x}.data'")
  Log          <- "q47b" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
  bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
  system(bsub_cmd)
  }
}
```
##Calculate AF over all pools
```{r copied from newer version for individual files, eval=FALSE}
datain <- as.data.table(grep ("[.]data", list.files(path="."), invert=FALSE, value=TRUE))
datain <-  datain[, V2 := str_replace_all(V1, "_chr25.vcf_" , "&_chr25.vcf_&")
                 ][, V2 := str_replace_all(V2, ".data" , "&.data")
                 ][, V2 := str_replace_all(V2, "bcfmpile_samples_" , "bcfmpile_samples_&")
                 ][, c("V3", "V4", "V5", "V6", "V7") := tstrsplit(V2, "&", fixed=TRUE)]
datain <-  unique(datain[, V4])

for (x in datain){
dt <-  data.table('s1CHROM'=character(),'s1POS'=integer(),'s1REF'=character(),'s1ALT'=character(),'s1AD0'=integer(),'s1AD1'=integer(),'s1DP'=integer(),'s1GQ'=character(),'s1PL'=character(),'r1CHROM'=character(),'r1POS'=integer(),'r1REF'=character(),'r1ALT'=character(),'r1AD0'=integer(),'r1AD1'=integer(),'r1DP'=integer(),'r1GQ'=character(),'r1PL'=character(),'Cpos'==integer())  

chrs   <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chr23","chr24","chr25","chr26","chr27","chr28","chr29")

  for (i in chrs){
  s1<- fread(paste0("bcfmpile_samples_",glue({x}), "_chr25.vcf_", glue({i}),".data"), col.names = c('s1CHROM','s1POS','s1REF','s1ALT','s1AD0','s1AD1','s1DP','s1GQ','s1PL'))
  s1 <- s1[, "Cpos" := str_c(glue({i}), s1POS, sep=".")]
  setkey(s1, Cpos)
  r1 <- fread(paste0("bcfmpile_samples_Russell_chr25.vcf_", glue({i}),".data"), col.names = c('r1CHROM','r1POS','r1REF','r1ALT','r1AD0','r1AD1','r1DP','r1GQ','r1PL'))
  r1 <- r1[, "Cpos" := str_c(glue({i}), r1POS, sep=".")]
  setkey(r1, Cpos)
  dt1 <- s1[r1]
  rm(r1,s1)
  dt1 <- dt1[!is.na(s1CHROM) & !is.na(r1CHROM),]
  dt1 <- dt1[s1ALT!="."]
  dt <- rbind(dt,dt1,fill=TRUE )
  rm(dt1)
  }
  #SNP-indexper bulk = Alternate allele depth divided by Total read depth
  dt <-dt[,'s1SNPi':=s1AD1/s1DP][,'r1SNPi':=r1AD1/r1DP]

  #Reference allele frequency = Ref allele depth HighBulk + Ref allele depth LowBulk divided by the Total read depth for both bulks
  #in this case       REF_FRQ = ref allele depth parent + Ref allele depth xMalePool/  DPrussell + DPxMalePool
  dt <-dt[,'REF_FRQ':=((s1AD0+r1AD0)/(s1DP+r1DP))]

  #delta(SNP-index) = SNP-indexHighBulk minus SNP-indexLowBulk
  dt<-dt[,'deltaSNP':=s1SNPi-r1SNPi]
  #sapply(dt,class)
  fwrite(dt, file=paste0(glue({x}), "_malepoolSNPS_dt.csv"), append=FALSE)
  rm(dt)
}
```

**split files for whole sample pools into chromosomes and recombine after filtering **
```{r bamfiles made in PSApopoolationfiles.Rmd, eval=FALSE}
bcfmpile_parents     <- as.data.table(grep(paste("bcfmpile_parents"), list.files(path="."), invert=FALSE, value=TRUE))
bcfmpile_samples     <- as.data.table(grep("bcfmpile_samples", list.files(path="."), invert=FALSE, value=TRUE))
bcfmpile_mrgdparents <- as.data.table(grep("mrgdparents_sort", list.files(path="."), invert=FALSE, value=TRUE))
abfiles              <- rbindlist(list(bcfmpile_parents,bcfmpile_samples,bcfmpile_mrgdparents))

abfiles <- abfiles[, V2 := str_replace_all(V1, ".vcf$" , "&.vcf")
                 ][, V2 := str_replace_all(V2, "bcfmpile_parents_" , "bcfmpile_parents_&")
                 ][, V2 := str_replace_all(V2, "bcfmpile_samples_" , "bcfmpile_samples_&")
                 ][, V2 := str_replace_all(V2, "_chr25" , "&_chr25")
                 ][, c("V3", "x2", "x3", "x4") := tstrsplit(V2, "&", fixed=TRUE)]
abfiles <- abfiles[x3=="_chr25",]
abfiles <- abfiles[, "CKs" := str_extract(x2, "CK")][V1=="bcfmpile_parents_Russell_chr25.vcf", CKs := "CK"][V1=="bcfmpile_samples_Russell_chr25.vcf", CKs := "CK"]
mp  <-  abfiles[, "dir" := "/powerplant/workspace/hrtcdf/github/FSTs/QTLseqR/"]

rm(abfiles, bcfmpile_parents,bcfmpile_mrgdparents, bcfmpile_samples)
```

```{r makes a data file for each chromosome for each file, eval=FALSE}
vcfin   <- mp[,V1]
chrs   <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chr23","chr24","chr25","chr26","chr27","chr28","chr29")
for (i in vcfin){
  for(x in chrs){
  cmd          <- glue::glue("bcftools query {i} -r '{x}' -e 'INFO/DP<20 || INFO/DP>200' -f '[[%CHROM]]\t%POS\t%REF\t%ALT\t%AD{{0}}\t%AD{{1}}\t%DP[[\t%GQ\t%PL]] \n' -o '{i}_{x}.data'")
  Log          <- "q47b" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
  bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
  system(bsub_cmd)
  }
}
```

##Calculate AF over all pools where high bulk equals parents and samples and male% as low bulks
```{r pull data from queried data, eval=FALSE}
datain <- as.data.table(grep ("[.]vcfdata", list.files(path="."), invert=FALSE, value=TRUE))
datain <-  datain[, V2 := str_replace_all(V1, "_chr25.vcfdata" , "&_chr25.vcf_&")
                 ][, V2 := str_replace_all(V2, "bcfmpile_samples_" , "bcfmpile_samples_&")
                 ][, V2 := str_replace_all(V2, "bcfmpile_parents_" , "bcfmpile_parents_&")                   
                 ][, c("V3", "V4","V5") := tstrsplit(V2, "&", fixed=TRUE)]
datain <-  unique(datain[, V4])

for (x in datain){
dt <-  data.table('s1CHROM'=character(),'s1POS'=integer(),'s1REF'=character(),'s1ALT'=character(),'s1AD0'=integer(),'s1AD1'=integer(),'s1DP'=integer(),'s1GQ'=character(),'s1PL'=character(),'p1CHROM'=character(),'p1POS'=integer(),'p1REF'=character(),'p1ALT'=character(),'p1AD0'=integer(),'p1AD1'=integer(),'p1DP'=integer(),'p1GQ'=character(),'p1PL'=character(),'Cpos'==integer())  

chrs   <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chr23","chr24","chr25","chr26","chr27","chr28","chr29")

  for (i in chrs){
  s1<- fread(paste0("bcfmpile_samples_",glue({x}), "_chr25.vcf_", glue({i})), col.names = c('s1CHROM','s1POS','s1REF','s1ALT','s1AD0','s1AD1','s1DP','s1GQ','s1PL'))
  s1 <- s1[, "Cpos" := str_c(glue({i}), s1POS, sep=".")]
  setkey(s1, Cpos)
  r1 <- fread(paste0("bcfmpile_parents_",glue({x}), "_chr25.vcf_", glue({i})), col.names = c('r1CHROM','r1POS','r1REF','r1ALT','r1AD0','r1AD1','r1DP','r1GQ','r1PL'))
  r1 <- r1[, "Cpos" := str_c(glue({i}), r1POS, sep=".")]
  setkey(r1, Cpos)
  dt1 <- s1[r1]
  rm(r1,s1)
  dt1 <- dt1[!is.na(s1CHROM) & !is.na(r1CHROM),]
  dt1 <- dt1[s1ALT!="."]
  dt <- rbind(dt,dt1,fill=TRUE )
  rm(dt1)
  }
  #SNP-indexper bulk = Alternate allele depth divided by Total read depth
  dt <-dt[,'s1SNPi':=s1AD1/s1DP][,'r1SNPi':=r1AD1/r1DP]

  #Reference allele frequency = Ref allele depth HighBulk + Ref allele depth LowBulk divided by the Total read depth for both bulks
  #in this case       REF_FRQ = ref allele depth parent + Ref allele depth xMalePool/  DPrussell + DPxMalePool
  dt <-dt[,'REF_FRQ':=((s1AD0+r1AD0)/(s1DP+r1DP))]

  #delta(SNP-index) = SNP-indexHighBulk minus SNP-indexLowBulk
  dt<-dt[,'deltaSNP':=s1SNPi-r1SNPi]
  #sapply(dt,class)
  fwrite(dt, file=paste0(glue({x}), "_samplepoolSNPs_dt.csv"), append=FALSE)
  rm(dt)
}
```

**compare male pool files**
```{r bamfiles made in PSApopoolationfiles.Rmd, eval=FALSE}
bcfmpile_parents     <- as.data.table(grep(paste("bcfmpile_parents"), list.files(path="."), invert=FALSE, value=TRUE))
bcfmpile_samples     <- as.data.table(grep("bcfmpile_samples", list.files(path="."), invert=FALSE, value=TRUE))
bcfmpile_mrgdparents <- as.data.table(grep("mrgdparents_sort", list.files(path="."), invert=FALSE, value=TRUE))
abfiles              <- rbindlist(list(bcfmpile_parents,bcfmpile_samples,bcfmpile_mrgdparents))

abfiles <- abfiles[, V2 := str_replace_all(V1, ".vcf$" , "&.vcf")
                 ][, V2 := str_replace_all(V2, "bcfmpile_parents_" , "bcfmpile_parents_&")
                 ][, V2 := str_replace_all(V2, "bcfmpile_samples_" , "bcfmpile_samples_&")
                 ][, V2 := str_replace_all(V2, "_chr25" , "&_chr25")
                 ][, c("V3", "x2", "x3", "x4") := tstrsplit(V2, "&", fixed=TRUE)]
abfiles <- abfiles[x3=="_chr25",]
#abfiles <- abfiles[V3=="Russell", V1 := "Russell_mrgdparents.bam"]
abfiles <-  abfiles[x2=="A1", "malepercentage" := "A1m90"][x2=="B2", "malepercentage" := "B2m75"][x2=="C3", "malepercentage" := "C3m60"][x2=="D6", "malepercentage" := "D6m44"][x2=="E11", "malepercentage" := "E11m80"][x2=="F12", "malepercentage" := "F12m50"][x2=="G13", "malepercentage" := "G13m45"][x2=="H14", "malepercentage" := "H14m68"][x2=="I15", "malepercentage" := "I15m21"][x2=="J16", "malepercentage" := "J16m11"]
abfiles <-  abfiles[, "dir" := "/powerplant/workspace/hrtcdf/github/FSTs/QTLseqR/"]
mps1 <-     abfiles[malepercentage %in% na.omit(abfiles$malepercentage),]
#mps2 <-     abfiles[V1=="bcfmpile_samples_Russell_chr25.vcf", ][,malepercentage :="Russell"] #not including russell
#mp  <-     rbind(mps1,mps2)
mp <- mps1
rm(abfiles, bcfmpile_parents,bcfmpile_mrgdparents, bcfmpile_samples, mps1, mps2)
```

##Calculate AF over all male pools to compare between each
```{r copied from newer version for individual files, eval=FALSE}
datain <- as.data.table(grep ("[.]data", list.files(path="."), invert=FALSE, value=TRUE))
datain <-  datain[, V2 := str_replace_all(V1, "_chr25.vcf_" , "&_chr25.vcf_&")
                 ][, V2 := str_replace_all(V2, ".data" , "&.data")
                 ][, V2 := str_replace_all(V2, "bcfmpile_samples_" , "bcfmpile_samples_&")
                 ][, c("V3", "V4", "V5", "V6", "V7") := tstrsplit(V2, "&", fixed=TRUE)]
datain <-  unique(datain[V4!="Russell", V4])

for (x in datain){
  for (y in datain){
dt <-  data.table('s1CHROM'=character(),'s1POS'=integer(),'s1REF'=character(),'s1ALT'=character(),'s1AD0'=integer(),'s1AD1'=integer(),'s1DP'=integer(),'s1GQ'=character(),'s1PL'=character(),'r1CHROM'=character(),'r1POS'=integer(),'r1REF'=character(),'r1ALT'=character(),'r1AD0'=integer(),'r1AD1'=integer(),'r1DP'=integer(),'r1GQ'=character(),'r1PL'=character(),'Cpos'==integer())  

chrs   <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chr23","chr24","chr25","chr26","chr27","chr28","chr29")

    for (i in chrs){
    s1<- fread(paste0("bcfmpile_samples_",glue({x}), "_chr25.vcf_", glue({i}),".data"), col.names = c('s1CHROM','s1POS','s1REF','s1ALT','s1AD0','s1AD1','s1DP','s1GQ','s1PL'))
    s1 <- s1[, "Cpos" := str_c(glue({i}), s1POS, sep=".")]
    setkey(s1, Cpos)
    r1 <- fread(paste0("bcfmpile_samples_",glue({y}), "_chr25.vcf_", glue({i}),".data"), col.names = c('r1CHROM','r1POS','r1REF','r1ALT','r1AD0','r1AD1','r1DP','r1GQ','r1PL'))
    r1 <- r1[, "Cpos" := str_c(glue({i}), r1POS, sep=".")]
    setkey(r1, Cpos)
    dt1 <- s1[r1]
    rm(r1,s1)
    dt1 <- dt1[!is.na(s1CHROM) & !is.na(r1CHROM),]
    dt1 <- dt1[s1ALT!="." | r1ALT!="."]
    dt <- rbind(dt,dt1,fill=TRUE )
    rm(dt1)
    }
    #SNP-indexper bulk = Alternate allele depth divided by Total read depth
    dt <-dt[,'s1SNPi':=s1AD1/s1DP][,'r1SNPi':=r1AD1/r1DP]

    #Reference allele frequency = Ref allele depth HighBulk + Ref allele depth LowBulk divided by the Total read depth for both bulks
    #in this case       REF_FRQ = ref allele depth parent + Ref allele depth xMalePool/  DPrussell + DPxMalePool 
    dt <-dt[,'REF_FRQ':=((s1AD0+r1AD0)/(s1DP+r1DP))]

    #delta(SNP-index) = SNP-indexHighBulk minus SNP-indexLowBulk
    dt<-dt[,'deltaSNP':=s1SNPi-r1SNPi]
    #sapply(dt,class)
    fwrite(dt, file=paste0(glue({x}),glue({y}), "_malepool_compare.csv"), append=FALSE)
    rm(dt)
  }
}
file.remove("B2A1_malepool_compare.csv", "A1A1_malepool_compare.csv", "A1Russell_malepool_compare.csv", "B2B2_malepool_compare.csv", "B2Russell_malepool_compare.csv", "C3A1_malepool_compare.csv","C3B2_malepool_compare.csv", "C3C3_malepool_compare.csv", "D6A1_malepool_compare.csv", "D6B2_malepool_compare.csv", "D6C3_malepool_compare.csv", "D6D6_malepool_compare.csv", "D6Russell_malepool_compare.csv", "E11A1_malepool_compare.csv", "E11B2_malepool_compare.csv", "E11C3_malepool_compare.csv", "E11D6_malepool_compare.csv", "E11E11_malepool_compare.csv", "E11Russell_malepool_compare.csv", "F12A1_malepool_compare.csv", "F12B2_malepool_compare.csv", "F12C3_malepool_compare.csv", "F12D6_malepool_compare.csv", "F12E11_malepool_compare.csv", "F12F12_malepool_compare.csv", "F12Russell_malepool_compare.csv", "G13A1_malepool_compare.csv", "G13B2_malepool_compare.csv", "G13C3_malepool_compare.csv", "G13D6_malepool_compare.csv", "G13E11_malepool_compare.csv", "G13F12_malepool_compare.csv", "G13G13_malepool_compare.csv", "G13Russell_malepool_compare.csv", "H14A1_malepool_compare.csv", "H14B2_malepool_compare.csv", "H14C3_malepool_compare.csv", "H14D6_malepool_compare.csv", "H14E11_malepool_compare.csv", "H14F12_malepool_compare.csv", "H14G13_malepool_compare.csv", "H14H14_malepool_compare.csv", "H14Russell_malepool_compare.csv", "I15A1_malepool_compare.csv", "I15B2_malepool_compare.csv", "I15C3_malepool_compare.csv", "I15D6_malepool_compare.csv", "I15E11_malepool_compare.csv", "I15F12_malepool_compare.csv", "I15G13_malepool_compare.csv", "I15H14_malepool_compare.csv","I15I15_malepool_compare.csv", "I15Russell_malepool_compare.csv", "J16A1_malepool_compare.csv", "J16B2_malepool_compare.csv", "J16C3_malepool_compare.csv", "J16D6_malepool_compare.csv", "J16E11_malepool_compare.csv", "J16F12_malepool_compare.csv", "J16G13_malepool_compare.csv", "J16H14_malepool_compare.csv","J16J16_malepool_compare.csv", "I15Russell_malepool_compare.csv")
```

**compare sample pool files**

##Calculate AF over all male pools to compare between each - not used since pools are compared with parent bulk.
```{r copied from newer version for individual files, eval=FALSE}
datain <- as.data.table(grep ("bcfmpile_samples", list.files(path="."), invert=FALSE, value=TRUE))
datain <-  datain[, V2 := str_replace_all(V1, "_chr25.vcfdata" , "&_chr25.vcf&")
][, V2 := str_replace_all(V2, "bcfmpile_samples_" , "bcfmpile_samples_&")                  
                 ][, V2 := str_replace_all(V2, "chr" , "&_chr")
                 ][, V2 := str_replace_all(V2, ".data" , "&.data")                   
                 ][, V2 := str_replace_all(V2, "bcfmpile_parents_" , "bcfmpile_parents_&")                   
                 ][, c("V3", "V4","V5","V6","V7") := tstrsplit(V2, "&", fixed=TRUE)]
datain <-  datain[ V4=="CK02_01_"| V4=="CK10_02_"| V4=="CK19_03_"| V4=="CK22_01_"| V4=="CK22_03_"| V4=="CK23_08_"| V4=="CK51_01_"| V4=="CK51_09_"| V4=="Russell_",]
datain <-  unique(datain[is.na(V7)][!is.na(V6), V4])

#	bcfmpile_samples_CK02_01_chr25.vcf_chr1
for (x in datain){
  for (y in datain){
dt <-  data.table('s1CHROM'=character(),'s1POS'=integer(),'s1REF'=character(),'s1ALT'=character(),'s1AD0'=integer(),'s1AD1'=integer(),'s1DP'=integer(),'s1GQ'=character(),'s1PL'=character(),'r1CHROM'=character(),'r1POS'=integer(),'r1REF'=character(),'r1ALT'=character(),'r1AD0'=integer(),'r1AD1'=integer(),'r1DP'=integer(),'r1GQ'=character(),'r1PL'=character(),'Cpos'==integer())  

chrs   <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chr23","chr24","chr25","chr26","chr27","chr28","chr29")

    for (i in chrs){
    s1<- fread(paste0("bcfmpile_samples_",glue({x}), "chr25.vcf_", glue({i})), col.names = c('s1CHROM','s1POS','s1REF','s1ALT','s1AD0','s1AD1','s1DP','s1GQ','s1PL'))
    s1 <- s1[, "Cpos" := str_c(glue({i}), s1POS, sep=".")]
    setkey(s1, Cpos)
    r1 <- fread(paste0("bcfmpile_samples_",glue({y}), "chr25.vcf_", glue({i})), col.names = c('r1CHROM','r1POS','r1REF','r1ALT','r1AD0','r1AD1','r1DP','r1GQ','r1PL'))
    r1 <- r1[, "Cpos" := str_c(glue({i}), r1POS, sep=".")]
    setkey(r1, Cpos)
    dt1 <- s1[r1]
    rm(r1,s1)
    dt1 <- dt1[!is.na(s1CHROM) & !is.na(r1CHROM),]
    dt1 <- dt1[s1ALT!="." | r1ALT!="."]
    dt <- rbind(dt,dt1,fill=TRUE )
    rm(dt1)
    }
    #SNP-indexper bulk = Alternate allele depth divided by Total read depth
    dt <-dt[,'s1SNPi':=s1AD1/s1DP][,'r1SNPi':=r1AD1/r1DP]

    #Reference allele frequency = Ref allele depth HighBulk + Ref allele depth LowBulk divided by the Total read depth for both bulks
    #in this case       REF_FRQ = ref allele depth parent + Ref allele depth xMalePool/  DPrussell + DPxMalePool
    dt <-dt[,'REF_FRQ':=((s1AD0+r1AD0)/(s1DP+r1DP))]

    #delta(SNP-index) = SNP-indexHighBulk minus SNP-indexLowBulk
    dt<-dt[,'deltaSNP':=s1SNPi-r1SNPi]
    #sapply(dt,class)
    fwrite(dt, file=paste0(glue({x}),glue({y}), "_samplepool_compare.csv"), append=FALSE)
    rm(dt)
  }
}
file.remove("CK02_01_CK02_01__samplepool_compare.csv", "CK10_02_CK02_01__samplepool_compare.csv", "CK10_02_CK10_02__samplepool_compare.csv", "CK19_03_CK02_01__samplepool_compare.csv", "CK19_03_CK10_02__samplepool_compare.csv", "CK19_03_CK19_03__samplepool_compare.csv", "CK22_03_CK02_01__samplepool_compare.csv", "CK22_03_CK10_02__samplepool_compare.csv", "CK22_03_CK19_03__samplepool_compare.csv", "CK22_03_CK22_03__samplepool_compare.csv", "CK23_08_CK02_01__samplepool_compare.csv", "CK23_08_CK10_02__samplepool_compare.csv", "CK23_08_CK19_03__samplepool_compare.csv", "CK23_08_CK22_03__samplepool_compare.csv", "CK23_08_CK23_08__samplepool_compare.csv", "CK51_01_CK02_01__samplepool_compare.csv", "CK51_01_CK10_02__samplepool_compare.csv", "CK51_01_CK19_03__samplepool_compare.csv", "CK51_01_CK22_03__samplepool_compare.csv", "CK51_01_CK23_08__samplepool_compare.csv", "CK51_01_CK51_01__samplepool_compare.csv", "CK51_09_CK02_01__samplepool_compare.csv", "CK51_09_CK10_02__samplepool_compare.csv", "CK51_09_CK22_03__samplepool_compare.csv", "CK51_09_CK23_08__samplepool_compare.csv", "CK51_09_CK51_01__samplepool_compare.csv", "CK51_09_CK51_09__samplepool_compare.csv", "Russell_CK02_01__samplepool_compare.csv", "Russell_CK10_02__samplepool_compare.csv", "Russell_CK19_03__samplepool_compare.csv", "Russell_CK22_03__samplepool_compare.csv", "Russell_CK23_08__samplepool_compare.csv", "Russell_CK51_01__samplepool_compare.csv", "Russell_CK51_09__samplepool_compare.csv", "Russell_Russell__samplepool_compare.csv")
```
