---
title: "malepoolmpile"
author: "Casey Flay"
date: "30/09/2020"
output:
  html_document: default
  word_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(RLinuxModules)
library(data.table)
library(glue)
library(here)
library(tidyverse)
library(magrittr)
library(GenomicRanges)
library(tools)
library(future)
library(future.batchtools)
library(here)
plan(batchtools_openlava, resources = list(threads=1))
setDTthreads(1)
module("load openlava asub/2.2 bwa/0.7.12 samtools/1.9 bcftools/1.10.2 popoolation2 perlbrew/0.76")#old samtools module loaded for popoolation2
here()
```
**parent files pulled from database, merged with picard, sorted and mpiled with bcftools**
```{r parents NCBI aligned bamfiles from male pool and testpools}
Rualn_bamDir <- "/output/genomic/plant/Actinidia/chinensis/Resequencing/Alignments/Russell_V2a.chromosomes.and.unassigned"
Rualn_bams  <- as.data.table(grep(paste('.bam$', collapse = "|"), list.files(path=Rualn_bamDir), invert=FALSE, value=TRUE))
Rualn_bams  <- Rualn_bams[, lapply(.SD, gsub, pattern = ("_Russell_V2a"), replacement = "&_Russell_V2a&"), .SDcols = colnames(Rualn_bams)][, c("er","reference",  "ext") := tstrsplit(V1, "&", fixed=TRUE)]
ER  <- Rualn_bams[, lapply(.SD, gsub, pattern = ("ER\\d{5}_"), replacement = ""), .SDcols = colnames(Rualn_bams)]
ER  <- ER[, lapply(.SD, gsub, pattern = ("_ER\\d{5}"), replacement = ""), .SDcols = colnames(Rualn_bams)]
ER  <- ER[, lapply(.SD, gsub, pattern = ("SRA_"), replacement = ""), .SDcols = colnames(Rualn_bams)]
ER  <- ER[, lapply(.SD, gsub, pattern = ("UNK_"), replacement = ""), .SDcols = colnames(Rualn_bams)]
setnames(ER, c("er"), c("PedigreeItem"))
Rualn_bams  <-cbind(Rualn_bams$er, ER)
ER  <- Rualn_bams[, c(1,3,4,5)]

ER  <- ER[, "er1" := str_extract(Rualn_bams$V1, "ER\\d{5}_")][, "er2" := str_extract(Rualn_bams$V1, "_ER\\d{5}")][, "er3" := str_extract(Rualn_bams$V1, "SRA_")][, "er4" := as.data.table(str_extract(Rualn_bams$V1, "UNK_"))]
ER  <- ER[, "er5" := paste0(er1,er2,er3,er4)]
ER5  <- ER[, lapply(.SD, gsub, pattern = ("NA"), replacement = ""), .SDcols = "er5"]
Rualn_bams  <-cbind(ER, ER5)
Rualn_bams  <- Rualn_bams[, c(1,2,3,4,10)]
Rualn_bams <- Rualn_bams[, "full" := paste0(er5,PedigreeItem,reference,ext)]
Rualn_bams <- Rualn_bams[, "dir" := Rualn_bamDir]
Rualn_bams <- Rualn_bams[, "pathandfile" := paste(dir, full, sep="/")]
t<- as.data.table(Rualn_bams$full)

Rualn_bams <- Rualn_bams[, "row" := seq_along(full)]
Rualn_bams <- Rualn_bams[PedigreeItem=="Red19", PedigreeItem := ("Red 19")]
Rualn_bams <- Rualn_bams[, "isfile" := (file.exists(pathandfile))]
Rualn_bams <- Rualn_bams[isfile != FALSE,]
rm(ER,ER5)
```

```{r}
#male pool parents are all the same for each pool e.g A1, B2 etc. unless adding multiple copies of each.
Russell_tp_bams <- Rualn_bams[PedigreeItem == 'CK02_04' | PedigreeItem == 'CK06_01' | PedigreeItem == 'CK09_06'| PedigreeItem == 'CK13_02'| PedigreeItem =='CK20_01'| PedigreeItem =='CK51_06'| PedigreeItem =='Russell'][, `:=` ("pool_parent"= 'Russell', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]# only found "CK06_01","CK20_01","CK51_06". missing 'CK02_04', 'CK13_02','CK09_06'.

#adjusted russell to a single input as it should be.
Russell_bam <- Rualn_bams[V1 =='ER10522_Russell'][, `:=` ("pool_parent"= 'Russell', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]

#Full pool parents these tables are also needed in the merge step but cant be used in a single table.
CK02_01_tp_bams <- Rualn_bams[PedigreeItem=='CK02_01'| PedigreeItem=='CK15_03'| PedigreeItem=='CK17_02'| PedigreeItem=='CK20_01'| PedigreeItem=='Bejing #36'| PedigreeItem == 'CK51_01'][, `:=` ("pool_parent"= 'CK02_01', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK02_01
CK10_02_tp_bams <- Rualn_bams[PedigreeItem=='CK10_02'| PedigreeItem=='CK17_03'| PedigreeItem=='CK18_01'| PedigreeItem=='CK19_03'| PedigreeItem=='CK51_09'][, `:=` ("pool_parent"= 'CK10_02', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK10_02
CK19_03_tp_bams <- Rualn_bams[PedigreeItem=='CK19_03'| PedigreeItem=='CK10_02'| PedigreeItem=='CK10_05'| PedigreeItem=='CK13_02'| PedigreeItem=='CK15_04'| PedigreeItem=='Bejing #36'| PedigreeItem=='CK51_01'][, `:=` ("pool_parent"= 'CK19_03', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK19_03
CK20_01_tp_bams <- Rualn_bams[PedigreeItem=='CK20_01'| PedigreeItem=='CK02_01'| PedigreeItem=='CK09_11'| PedigreeItem=='CK10_03'][, `:=` ("pool_parent"= 'CK20_01', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK20_01
CK22_03_tp_bams <- Rualn_bams[PedigreeItem=='CK22_03'| PedigreeItem=='CK23_08'| PedigreeItem=='CK51_09'][, `:=` ("pool_parent"= 'CK22_03', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK22_03
CK23_08_tp_bams <- Rualn_bams[PedigreeItem=='CK23_08'| PedigreeItem=='CK01_01_01_01'| PedigreeItem=='CK15_03'| PedigreeItem=='CK16_01'| PedigreeItem=='CK22_03'][, `:=` ("pool_parent"= 'CK23_08', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK23_08
CK51_01_tp_bams <- Rualn_bams[PedigreeItem=='CK51_01'| PedigreeItem=='CK02_01'| PedigreeItem=='CK17_03'| PedigreeItem=='CK19_03'][, `:=` ("pool_parent"= 'CK51_01', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK51_01
CK51_09_tp_bams <- Rualn_bams[PedigreeItem=='CK51_09'| PedigreeItem=='CK02_04'| PedigreeItem=='CK06_01'| PedigreeItem=='CK10_02'| PedigreeItem=='CK15_04'| PedigreeItem=='CK22_03'][, `:=` ("pool_parent"= 'CK51_09', 'vcf' = paste(PedigreeItem, '_chr25_pile.vcf.gz',sep=""))]#Full pool parents CK51_09
parent_bams <- rbindlist(list(Russell_tp_bams, CK02_01_tp_bams, CK10_02_tp_bams, CK19_03_tp_bams, CK20_01_tp_bams, CK22_03_tp_bams, CK23_08_tp_bams, CK51_01_tp_bams, CK51_09_tp_bams))

parent_bams[, "psymlink_dir" := "/powerplant/workspace/hrtcdf/github/FSTs/parentpools"]
parent_bams <- parent_bams[er5=="ER10508_" |er5=="ER10522_",][, "pbams" := paste0(er5,PedigreeItem,reference, ext, sep="")]

#file.symlink(unique(parent_bams$pathandfile), here('QTLseqR'))
```

```{r index}
bamin          <- paste0(unique(parent_bams$pbams))
file.exists(bamin)
for (i in bamin){
cmd        <- glue::glue("samtools index {i}")
Log          <- "80" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
#system(bsub_cmd)
}
```

```{r picard merge bams into parent pools keeps read groups for later seperation which i dont want using samtools instead.}
for (i in unique(parent_bams$pool_parent)){ 
x       <- parent_bams[pool_parent=={i}, pbams]
infile  <- str_c(paste("I=", x, sep=""), sep="", collapse=" ")
outfile <- paste({i}, "_mrgdparents.bam", sep="")

file.exists(x)
cmd         <- glue("java -jar /software/bioinformatics/picard-tools-2.9.4/picard.jar MergeSamFiles {infile} O={outfile}")
Log          <- "92" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
#system(bsub_cmd)
}
```

```{r samtools 1.9 merge}
for (i in unique(parent_bams$pool_parent)){ 
infile       <- parent_bams[pool_parent=={i}, pbams]
outfile <- paste({i}, "_mrgdparents.bam", sep="")

file.exists(infile)
cmd         <- glue("samtools merge -f -r {outfile} {infile}")
Log          <- "merge92b" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
system(bsub_cmd)
}
```

```{r sort }
parent_bams <- parent_bams[, `:=` ("mrgedparents"= paste0(parent_bams$pool_parent, "_mrgdparents", sep=""), "bam"=".bam")]
bamin          <- parent_bams[, unique(mrgedparents)]

for (i in bamin){
cmd          <- glue::glue("samtools sort -o {i}_sort.bam {i}.bam")
Log          <- "si103" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
system(bsub_cmd)
}
```

```{r index }
bamin          <- parent_bams[, unique(mrgedparents)]
for (i in bamin){
cmd        <- glue::glue("samtools index {i}_sort.bam")
Log          <- "index116" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
system(bsub_cmd)
}
```

```{r run bcftools mpileup for chr25}
#-r 'chr25'
submit_mpileup <- function(reference = "*.fa.gz", output = c(), bamfiles = c(), job.init = "asub", job.opts = "-q normal") {
  cmd          <- "bcftools mpileup -B -I -Q 20 -f {reference} --excl-flags 'UNMAP,SECONDARY,QCFAIL,DUP' -a {mpileuptags} {bamfiles} -Ou | bcftools call -f {bcfcalltags} -p {pvalThres} -m -A -Oz -o {output} && bcftools index {output}" 
  tmpf         <- tempfile(pattern = "submit_mileup.", fileext = ".sh")
  writeLines(text = c(glue(cmd)), con = tmpf)
  submit       <- "{job.init} {job.opts} {tmpf}"
  system(glue(submit), intern = TRUE)
}
mpileuptags    <- "DP,AD,SP,INFO/AD"
bcfcalltags    <- "GQ,GP"
pvalThres      <- 0.99
PedigreeItem   <- parent_bams[, unique(pool_parent)]
bamfiles       <- parent_bams[, unique(paste0(mrgedparents, "_sort.bam"))]
output         <- paste("bcfmpile_parents", PedigreeItem, ".vcf", sep = "_")
reference      <- dir("/workspace/hrards/projects/hrtcdf/genome/fasta", pattern = ".*\\.gz$", full.names = TRUE)
file.exists(bamfiles)
file.exists(reference)
submit_mpileup(reference = reference, output = output, bamfiles = bamfiles)
```

***male pools from hrtcdf/github/FSTs/malepool/malepool.Rmd***
```{r symlink male pool bam files}
royDir <- "/workspace/hrards/projects/hrtcdf/10.alignments/bwa"
outDir <- "/powerplant/workspace/hrtcdf/github/FSTs/QTLseqR"
mspoolfiles       <- dir(royDir, pattern = ".bam", full.names = TRUE, include.dirs = TRUE)

#file.symlink(mspoolfiles, outDir)
```

```{r transform filenames to sample, lane and suffix for data.table from symlink}
mpooldt_bam <- as.data.table(grep(paste("^HW5YYDSXX"), list.files(path="."), invert=FALSE, value=TRUE))
mpooldt_bam <- mpooldt_bam[, "pooltype" := "variablemale"]
spooldt_bam <- as.data.table(grep(paste("^HL55WDRXX"), list.files(path="."), invert=FALSE, value=TRUE))
spooldt_bam <- spooldt_bam[, "pooltype" := "lowPSA"]
pooldt_bam  <- rbind(mpooldt_bam, spooldt_bam)

pooldt_bam <- pooldt_bam[, V2 := str_replace_all(V1, "_L" , "&_L")
               ][, V2 := str_replace_all(V2, "XX_" , "XX_&")
               ][, V2 := str_replace_all(V2, ".bam" , "&.bam") 
               ][, c("H", "sample", "lane", "ftype") := tstrsplit(V2, "&", fixed=TRUE)
               ][, V2:=NULL]
```

```{r index}
bamin <- pooldt_bam[ftype==".bam", V1]

for (i in bamin){
cmd         <- glue::glue("samtools index {i}")
Log          <- "170" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
system(bsub_cmd)
}
```

```{r picard merge lanes for sample lowPSA bam files}
for (i in unique(pooldt_bam$sample)){ 
x       <- pooldt_bam[sample=={i} & ftype==".bam", V1]
infile  <- str_c(paste("I=", x, sep=""), sep="", collapse=" ")
outfile <- paste({i}, "_picmrg_lowPSA.bam", sep="")

file.exists(x)
cmd         <- glue("java -jar /software/bioinformatics/picard-tools-2.9.4/picard.jar MergeSamFiles --CREATE_INDEX {infile} O={outfile}")
Log          <- "78" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
#system(bsub_cmd)
}
```

```{r}
picmrg_lowPSApooldt_bam <- as.data.table(paste0(unique(pooldt_bam$sample), "_picmrg_lowPSA.bam"))
mpilepoolbams <- picmrg_lowPSApooldt_bam[, `:=` (sample= str_replace_all(V1, "_picmrg_lowPSA.bam", ""), "ftype"=".bam")]

#variablemale     <- pooldt_bam[pooltype=="variablemale", .(V1, sample, ftype)]
#mpilepoolbams    <- rbind(picmrg_lowPSApooldt_bam, variablemale)
```

```{r sort}
bamin     <- mpilepoolbams[ftype==".bam",V1]
for (i in bamin){
cmd       <- glue::glue("samtools sort -o s_{i} {i}")
Log       <- "si298";Ol<-paste0(here('PSApopoolation/logD/'),Log,".out");Oe<-paste0(here('PSApopoolation/logD/'),Log,".err")
bsub_cmd  <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
system(bsub_cmd)
}
```

```{r index}
bamin     <- mpilepoolbams[ftype==".bam",V1]
for (i in bamin){
cmd       <- glue::glue("samtools index s_{i}")
Log       <- "i232";Ol<-paste0(here('PSApopoolation/logD/'),Log,".out");Oe<-paste0(here('PSApopoolation/logD/'),Log,".err")
bsub_cmd  <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
system(bsub_cmd)
}
mpilepoolbams     <- mpilepoolbams[ftype==".bam",'s_' := "s_"]
```

```{r run bcftools mpileup for chr25 for QTLseqR}
submit_mpileup <- function(reference = "*.fa.gz", output = c(), bamfiles = c(), job.init = "asub", job.opts = "-q normal") {
  cmd          <- "bcftools mpileup -B -I -Q 20 -f {reference} --excl-flags 'UNMAP,SECONDARY,QCFAIL,DUP' -a {mpileuptags} {bamfiles} -Ou | bcftools call  -f {bcfcalltags} -p {pvalThres} -m -v -A -Oz -o {output} && bcftools index -f {output}" 
  tmpf         <- tempfile(pattern = "submit_mileup.", fileext = ".sh")
  writeLines(text = c(glue(cmd)), con = tmpf)
  submit       <- "{job.init} {job.opts} {tmpf}"
  system(glue(submit), intern = TRUE)
}
mpileuptags    <- "DP,AD,SP,INFO/AD"
bcfcalltags    <- "GQ,GP"
pvalThres      <- 0.99
PedigreeItem   <- mpilepoolbams[ftype==".bam",sample]
bamfiles       <- mpilepoolbams[ftype==".bam",paste0("s_", V1)]
output         <- paste("bcfmpile_samples_", PedigreeItem, ".vcf", sep = "")
reference      <- dir("/workspace/hrards/projects/hrtcdf/genome/fasta", pattern = ".*\\.gz$", full.names = TRUE)
file.exists(bamfiles)
file.exists(reference)
submit_mpileup(reference = reference, output = output, bamfiles = bamfiles)
```

```{r remove unneeded files}
r1     <- paste0(merged_parent_bams$mrgedparents)
bamin1          <- paste(mpilepoolbams$V1)
bamin2          <- paste(pooldt_bam$V1)
bai <- grep(paste("*.bai"), list.files(path="."), invert=FALSE, value=TRUE)

#file.remove(r1, bamin1, bamin2, bai)
```

***make sure the correct russell is used for the parent***

```{r index}
cmd          <- "samtools index 'ER10522_Russell_Russell_V2a.chromosomes.and.unassigned.fsa.bam'"
Log          <- "i273" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
system(bsub_cmd)
```


```{r sort}
cmd       <- glue::glue("samtools sort -o 's_ER10522_Russell_Russell_V2a.chromosomes.and.unassigned.fsa.bam' 'ER10522_Russell_Russell_V2a.chromosomes.and.unassigned.fsa.bam'")
Log       <- "s280";Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd  <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
system(bsub_cmd)
```

```{r index}
cmd       <- glue::glue("samtools index 's_ER10522_Russell_Russell_V2a.chromosomes.and.unassigned.fsa.bam'")
Log       <- "i287";Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
bsub_cmd  <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
system(bsub_cmd)
```

```{r run bcftools mpileup for chr25 for QTLseqR}
submit_mpileup <- function(reference = "*.fa.gz", output = c(), bamfiles = c(), job.init = "asub", job.opts = "-q normal") {
  cmd          <- "bcftools mpileup -B -I -Q 20 -f {reference} --excl-flags 'UNMAP,SECONDARY,QCFAIL,DUP' -a {mpileuptags} {bamfiles} -Ou | bcftools call  -f {bcfcalltags} -p {pvalThres} -m -v -A -Oz -o {output} && bcftools index -f {output}" 
  tmpf         <- tempfile(pattern = "submit_mileup.", fileext = ".sh")
  writeLines(text = c(glue(cmd)), con = tmpf)
  submit       <- "{job.init} {job.opts} {tmpf}"
  system(glue(submit), intern = TRUE)
}
mpileuptags    <- "DP,AD,SP,INFO/AD"
bcfcalltags    <- "GQ,GP"
pvalThres      <- 0.99
bamfiles       <- 's_ER10522_Russell_Russell_V2a.chromosomes.and.unassigned.fsa.bam'
output         <- 'Russell_for_malepools.vcf.gz'
reference      <- dir("/workspace/hrards/projects/hrtcdf/genome/fasta", pattern = ".*\\.gz$", full.names = TRUE)
file.exists(bamfiles)
file.exists(reference)
submit_mpileup(reference = reference, output = output, bamfiles = bamfiles)
```
