---
title: "qtlseqr"
author: "Casey Flay"
date: "12/08/2021"
output:
  html_document: default
  word_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
library(RLinuxModules)
library(data.table)
library(glue)
library(tidyverse)
library(here)
library(QTLseqr)
setDTthreads(1)
module("load openlava asub/2.2 bwa/0.7.12 samtools/1.9 bcftools/1.10.2 perlbrew/0.76")
knitr::opts_chunk$set(echo = FALSE, include=FALSE)
```

##files for this work were prepared in QTLseqR
```{r bamfiles made in PSApopoolationfiles.Rmd, eval=FALSE}
bcfmpile_parents     <- as.data.table(grep(paste("bcfmpile_parents"), list.files(path="."), invert=FALSE, value=TRUE))
bcfmpile_samples     <- as.data.table(grep("bcfmpile_samples", list.files(path="."), invert=FALSE, value=TRUE))
bcfmpile_mrgdparents <- as.data.table(grep("mrgdparents_sort", list.files(path="."), invert=FALSE, value=TRUE))
abfiles              <- rbindlist(list(bcfmpile_parents,bcfmpile_samples,bcfmpile_mrgdparents))

abfiles <- abfiles[, V2 := str_replace_all(V1, ".vcf$" , "&.vcf")
                 ][, V2 := str_replace_all(V2, "bcfmpile_parents_" , "bcfmpile_parents_&")
                 ][, V2 := str_replace_all(V2, "bcfmpile_samples_" , "bcfmpile_samples_&")
                 ][, V2 := str_replace_all(V2, "_chr25" , "&_chr25")
                 ][, c("V3", "x2", "x3", "x4") := tstrsplit(V2, "&", fixed=TRUE)]
abfiles <- abfiles[x3=="_chr25",]
#abfiles <- abfiles[V3=="Russell", V1 := "Russell_mrgdparents.bam"]
abfiles <-  abfiles[x2=="A1", "malepercentage" := "A1m90"][x2=="B2", "malepercentage" := "B2m75"][x2=="C3", "malepercentage" := "C3m60"][x2=="D6", "malepercentage" := "D6m44"][x2=="E11", "malepercentage" := "E11m80"][x2=="F12", "malepercentage" := "F12m50"][x2=="G13", "malepercentage" := "G13m45"][x2=="H14", "malepercentage" := "H14m68"][x2=="I15", "malepercentage" := "I15m21"][x2=="J16", "malepercentage" := "J16m11"]
abfiles <-  abfiles[, "dir" := "/powerplant/workspace/hrtcdf/github/FSTs/QTLseqR/"]
mps1 <-     abfiles[malepercentage %in% na.omit(abfiles$malepercentage),]
mps2 <-     abfiles[V1=="bcfmpile_samples_Russell_chr25.vcf", ][,malepercentage :="Russell"]
mp  <-     rbind(mps1,mps2)
rm(abfiles, bcfmpile_parents,bcfmpile_mrgdparents, bcfmpile_samples, mps1, mps2)
```


```{r, eval=FALSE, makes a data file for each chromosome for each file}
vcfin   <- mp[,V1]
chrs   <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chr23","chr24","chr25","chr26","chr27","chr28","chr29")
for (i in vcfin){
  for(x in chrs){
  cmd          <- glue::glue("bcftools query {i} -r '{x}' -e 'INFO/DP<20 || INFO/DP>200' -f '[[%CHROM]]\t%POS\t%REF\t%ALT\t%AD{{0}}\t%AD{{1}}\t%DP[[\t%GQ\t%PL]] \n' -o '{i}_{x}.data'")
  Log          <- "q47b" ;Ol<-paste0(here('QTLseqR/logD/'),Log,".out");Oe<-paste0(here('QTLseqR/logD/'),Log,".err")
  bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
  system(bsub_cmd)
  }
}
```

##Calculate AF over all pools
```{r, eval=FALSE copied from newer version for individual files}
datain <- as.data.table(grep ("[.]data", list.files(path="."), invert=FALSE, value=TRUE))
datain <-  datain[, V2 := str_replace_all(V1, "_chr25.vcf_" , "&_chr25.vcf_&")
                 ][, V2 := str_replace_all(V2, ".data" , "&.data")
                 ][, V2 := str_replace_all(V2, "bcfmpile_samples_" , "bcfmpile_samples_&")
                 ][, c("V3", "V4", "V5", "V6", "V7") := tstrsplit(V2, "&", fixed=TRUE)]
datain <-  unique(datain[, V4])



for (x in datain){
dt <-  data.table('s1CHROM'=character(),'s1POS'=integer(),'s1REF'=character(),'s1ALT'=character(),'s1AD0'=integer(),'s1AD1'=integer(),'s1DP'=integer(),'s1GQ'=character(),'s1PL'=character(),'p1CHROM'=character(),'p1POS'=integer(),'p1REF'=character(),'p1ALT'=character(),'p1AD0'=integer(),'p1AD1'=integer(),'p1DP'=integer(),'p1GQ'=character(),'p1PL'=character(),'Cpos'==integer())  
  chrs   <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr20","chr21","chr22","chr23","chr24","chr25","chr26","chr27","chr28","chr29")

  for (i in chrs){
  s1<- fread(paste0("bcfmpile_samples_",glue({x}), "_chr25.vcf_", glue({i}),".data"), col.names = c('s1CHROM','s1POS','s1REF','s1ALT','s1AD0','s1AD1','s1DP','s1GQ','s1PL'))
  s1 <- s1[, "Cpos" := str_c(glue({i}), s1POS, sep=".")]
  setkey(s1, Cpos)
  r1 <- fread(paste0("bcfmpile_samples_Russell_chr25.vcf_", glue({i}),".data"), col.names = c('r1CHROM','r1POS','r1REF','r1ALT','r1AD0','r1AD1','r1DP','r1GQ','r1PL'))
  r1 <- r1[, "Cpos" := str_c(glue({i}), r1POS, sep=".")]
  setkey(r1, Cpos)
  dt1 <- s1[r1]
  rm(r1,s1)
  dt1 <- dt1[!is.na(s1CHROM) & !is.na(r1CHROM),]
  dt1 <- dt1[s1ALT!="."]
  dt <- rbind(dt,dt1,fill=TRUE )
  rm(dt1)
  }
  #SNP-indexper bulk = Alternate allele depth divided by Total read depth
  dt <-dt[,'s1SNPi':=s1AD1/s1DP][,'r1SNPi':=r1AD1/r1DP]

  #Reference allele frequency = Ref allele depth HighBulk + Ref allele depth LowBulk divided by the Total read depth for both bulks
  #in this case       REF_FRQ = ref allele depth parent + Ref allele depth xMalePool/  DPrussell + DPxMalePool
  dt <-dt[,'REF_FRQ':=((s1AD0+r1AD0)/(s1DP+r1DP))]

  #delta(SNP-index) = SNP-indexHighBulk minus SNP-indexLowBulk
  dt<-dt[,'deltaSNP':=s1SNPi-r1SNPi]
  #sapply(dt,class)
  fwrite(dt, file=paste0(glue({x}), "_malepoolSNPS_dt.csv"), append=FALSE)
  rm(dt)
}
```

```{r, eval=FALSE older version for single chromosome delete when the loop above works.}
a <- fread("bcfmpile_samples_A1_chr25.vcfdata", col.names = c('aCHROM','POS','aREF','aALT','aAD0','aAD1','aDP','aGQ','aPL'))
b <- fread("bcfmpile_samples_B2_chr25.vcfdata", col.names = c('bCHROM','POS','bREF','bALT','bAD0','bAD1','bDP','bGQ','bPL'))
c <- fread("bcfmpile_samples_C3_chr25.vcfdata", col.names = c('cCHROM','POS','cREF','cALT','cAD0','cAD1','cDP','cGQ','cPL'))
d <- fread("bcfmpile_samples_D6_chr25.vcfdata", col.names = c('dCHROM','POS','dREF','dALT','dAD0','dAD1','dDP','dGQ','dPL'))
e <- fread("bcfmpile_samples_E11_chr25.vcfdata", col.names = c('eCHROM','POS','eREF','eALT','eAD0','eAD1','eDP','eGQ','ePL'))
f <- fread("bcfmpile_samples_F12_chr25.vcfdata", col.names = c('fCHROM','POS','fREF','fALT','fAD0','fAD1','fDP','fGQ','fPL'))
g <- fread("bcfmpile_samples_G13_chr25.vcfdata", col.names = c('gCHROM','POS','gREF','gALT','gAD0','gAD1','gDP','gGQ','gPL'))
h <- fread("bcfmpile_samples_H14_chr25.vcfdata", col.names = c('hCHROM','POS','hREF','hALT','hAD0','hAD1','hDP','hGQ','hPL'))
i <- fread("bcfmpile_samples_I15_chr25.vcfdata", col.names = c('iCHROM','POS','iREF','iALT','iAD0','iAD1','iDP','iGQ','iPL'))
j <- fread("bcfmpile_samples_J16_chr25.vcfdata", col.names = c('jCHROM','POS','jREF','jALT','jAD0','jAD1','jDP','jGQ','jPL'))
r <- fread("bcfmpile_samples_Russell_chr25.vcfdata", col.names = c('rCHROM','POS','rREF','rALT','rAD0','rAD1','rDP','rGQ','rPL'))
a <- a[, "Cpos" := str_c(aCHROM, POS, sep="_")]; setkey(a, Cpos)
b <- b[, "Cpos" := str_c(bCHROM, POS, sep="_")]; setkey(b, Cpos)
c <- c[, "Cpos" := str_c(cCHROM, POS, sep="_")]; setkey(c, Cpos)
d <- d[, "Cpos" := str_c(dCHROM, POS, sep="_")]; setkey(d, Cpos)
e <- e[, "Cpos" := str_c(eCHROM, POS, sep="_")]; setkey(e, Cpos)
f <- f[, "Cpos" := str_c(fCHROM, POS, sep="_")]; setkey(f, Cpos)
g <- g[, "Cpos" := str_c(gCHROM, POS, sep="_")]; setkey(g, Cpos)
h <- h[, "Cpos" := str_c(hCHROM, POS, sep="_")]; setkey(h, Cpos)
i <- i[, "Cpos" := str_c(iCHROM, POS, sep="_")]; setkey(i, Cpos)
j <- j[, "Cpos" := str_c(jCHROM, POS, sep="_")]; setkey(j, Cpos)
r <- r[, "Cpos" := str_c(rCHROM, POS, sep="_")]; setkey(r, Cpos)


mp <- a[b,][c,][d,][e,][f,][g,][h,][i,][j,][r,]
pools <- as.data.table(c("a","b","c","d","e","f","g","h","i","j","r"))
rm(b,c,d,e,f,g,h,i,j,r)
##filter sites with no alt allele.
mp<- mp[!is.na(aCHROM) &!is.na(bCHROM)& !is.na(cCHROM)& !is.na(dCHROM) & !is.na(eCHROM)& !is.na(fCHROM)& !is.na(gCHROM)& !is.na(hCHROM) & !is.na(iCHROM) & !is.na(rCHROM),]
mp<- mp[aALT!="." |bALT!="." |cALT!="." |dALT!="." |eALT!="." |fALT!="." |gALT!="." |hALT!="." |iALT!="." |jALT!=".",]
sapply(mp ,class)
#SNP-indexper bulk = Alternate allele depth divided by Total read depth
for (i in pools){copy_my_contents_for_below <- paste("mp<-mp",paste(glue("[,'{i}SNPi':={i}AD1/{i}DP]"), collapse = ""))}
copy_my_contents_for_below#copy cmd script and paste below. it works....
mp<-mp [,'aSNPi':=aAD1/aDP][,'bSNPi':=bAD1/bDP][,'cSNPi':=cAD1/cDP][,'dSNPi':=dAD1/dDP][,'eSNPi':=eAD1/eDP][,'fSNPi':=fAD1/fDP][,'gSNPi':=gAD1/gDP][,'hSNPi':=hAD1/hDP][,'iSNPi':=iAD1/iDP][,'jSNPi':=jAD1/jDP][,'rSNPi':=rAD1/rDP]

#Reference allele frequency = Ref allele depth HighBulk + Ref allele depth LowBulk divided by the Total read depth for both bulks
#in this case       REF_FRQ = ref allele depth Russell + Ref allele depth xMalePool/  DPrussell + DPxMalePool
for (i in pools){RAFcopy_my_contents_for_below <- paste("mp<-mp",paste(glue("[,'{i}REF_FRQ':=(rAD0+{i}AD0)/(rDP+{i}DP)]"), collapse = ""))}
RAFcopy_my_contents_for_below#copy script and paste below.
mp<-mp [,'aREF_FRQ':=((rAD0+aAD0)/(rDP+aDP))][,'bREF_FRQ':=((rAD0+bAD0)/(rDP+bDP))][,'cREF_FRQ':=((rAD0+cAD0)/(rDP+cDP))][,'dREF_FRQ':=((rAD0+dAD0)/(rDP+dDP))][,'eREF_FRQ':=((rAD0+eAD0)/(rDP+eDP))][,'fREF_FRQ':=((rAD0+fAD0)/(rDP+fDP))][,'gREF_FRQ':=((rAD0+gAD0)/(rDP+gDP))][,'hREF_FRQ':=((rAD0+hAD0)/(rDP+hDP))][,'iREF_FRQ':=((rAD0+iAD0)/(rDP+iDP))][,'jREF_FRQ':=((rAD0+jAD0)/(rDP+jDP))]

#delta(SNP-index) = SNP-indexHighBulk âˆ’ SNP-indexLowBulk
for (i in pools){deltaSNPcopy_my_contents_for_below <- paste("mp<-mp",paste(glue("[,'{i}deltaSNP':=rSNPi/{i}SNPi]"), collapse = ""))}
deltaSNPcopy_my_contents_for_below#copy cmd script and paste below. it works....
mp<-mp [,'adeltaSNP':=aSNPi-rSNPi][,'bdeltaSNP':=bSNPi-rSNPi][,'cdeltaSNP':=cSNPi-rSNPi][,'ddeltaSNP':=dSNPi-rSNPi][,'edeltaSNP':=eSNPi-rSNPi][,'fdeltaSNP':=fSNPi-rSNPi][,'gdeltaSNP':=gSNPi-rSNPi][,'hdeltaSNP':=hSNPi-rSNPi][,'ideltaSNP':=iSNPi-rSNPi][,'jdeltaSNP':=jSNPi-rSNPi]
#sapply(mp,class)

fwrite(mp, file="malepool_dt.csv", append=FALSE)
rm(a,mp)

```

```{r group for analysis rename for QTLseqR}
mp <- fread('malepool_dt.csv')
pools <- as.data.table(c("a","b","c","d","e","f","g","h","i","j","r"))
for (i in pools){deltaSNPcopy_my_contents_for_below <- glue("[,.(rCHROM,rREF,rALT,rAD0,rAD1,rDP,rGQ,rPL,rSNPi,POS,{i}REF,{i}ALT,{i}AD0,{i}AD1,{i}DP,{i}GQ,{i}PL,{i}SNPi,{i}deltaSNP),{i}REF_FRQ]")}
#deltaSNPcopy_my_contents_for_below#copy cmd script and paste below. it works....

a_df<-mp[,.(rCHROM,rREF,rALT,rAD0,rAD1,rDP,rGQ,rPL,rSNPi,POS,aREF,aALT,aAD0,aAD1,aDP,aGQ,aPL,aSNPi,adeltaSNP,aREF_FRQ)][aALT!=".",]
b_df<-mp[,.(rCHROM,rREF,rALT,rAD0,rAD1,rDP,rGQ,rPL,rSNPi,POS,bREF,bALT,bAD0,bAD1,bDP,bGQ,bPL,bSNPi,bdeltaSNP,bREF_FRQ)][bALT!=".",]
c_df<-mp[,.(rCHROM,rREF,rALT,rAD0,rAD1,rDP,rGQ,rPL,rSNPi,POS,cREF,cALT,cAD0,cAD1,cDP,cGQ,cPL,cSNPi,cdeltaSNP,cREF_FRQ)][cALT!=".",]
d_df<-mp[,.(rCHROM,rREF,rALT,rAD0,rAD1,rDP,rGQ,rPL,rSNPi,POS,dREF,dALT,dAD0,dAD1,dDP,dGQ,dPL,dSNPi,ddeltaSNP,dREF_FRQ)][dALT!=".",]
e_df<-mp[,.(rCHROM,rREF,rALT,rAD0,rAD1,rDP,rGQ,rPL,rSNPi,POS,eREF,eALT,eAD0,eAD1,eDP,eGQ,ePL,eSNPi,edeltaSNP,eREF_FRQ)][eALT!=".",]
f_df<-mp[,.(rCHROM,rREF,rALT,rAD0,rAD1,rDP,rGQ,rPL,rSNPi,POS,fREF,fALT,fAD0,fAD1,fDP,fGQ,fPL,fSNPi,fdeltaSNP,fREF_FRQ)][fALT!=".",]
g_df<-mp[,.(rCHROM,rREF,rALT,rAD0,rAD1,rDP,rGQ,rPL,rSNPi,POS,gREF,gALT,gAD0,gAD1,gDP,gGQ,gPL,gSNPi,gdeltaSNP,gREF_FRQ)][gALT!=".",]
h_df<-mp[,.(rCHROM,rREF,rALT,rAD0,rAD1,rDP,rGQ,rPL,rSNPi,POS,hREF,hALT,hAD0,hAD1,hDP,hGQ,hPL,hSNPi,hdeltaSNP,hREF_FRQ)][hALT!=".",]
i_df<-mp[,.(rCHROM,rREF,rALT,rAD0,rAD1,rDP,rGQ,rPL,rSNPi,POS,iREF,iALT,iAD0,iAD1,iDP,iGQ,iPL,iSNPi,ideltaSNP,iREF_FRQ)][iALT!=".",]
j_df<-mp[,.(rCHROM,rREF,rALT,rAD0,rAD1,rDP,rGQ,rPL,rSNPi,POS,jREF,jALT,jAD0,jAD1,jDP,jGQ,jPL,jSNPi,jdeltaSNP,jREF_FRQ)][jALT!=".",]

##low =Russell,  high=sample

for (i in pools){copy_my_contents_for_below <-paste(glue("{i}_df<-"), glue("setnames({i}_df, c('{i}CHROM','{i}POS','{i}REF','{i}ALT','rAD0','rAD1','rDP','rGQ','rPL','rSNPi','{i}AD0','{i}AD1','{i}DP','{i}GQ','{i}PL','{i}SNPi','{i}REF_FRQ','{i}deltaSNP'), c('CHROM','POS','REF','ALT','AD_REF.LOW','AD_ALT.LOW','DP.LOW','GQ.LOW','PL.LOW','SNPindex.LOW','AD_REF.HIGH','AD_ALT.HIGH','DP.HIGH','GQ.HIGH','PL.HIGH','SNPindex.HIGH','REF_FRQ','deltaSNP'))"),sep="")
}
copy_my_contents_for_below#copy cmd script and paste below.



```
```{r}
#sapply(a_df_nofilt ,class)
```

```{r filter}
a_df <- filterSNPs(SNPset = a_df_nofilt, refAlleleFreq = 0.10, minTotalDepth = 20, maxTotalDepth = 130, depthDifference = 100, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
b_df <- filterSNPs(SNPset = b_df_nofilt, refAlleleFreq = 0.10, minTotalDepth = 20, maxTotalDepth = 130, depthDifference = 100, minSampleDepth = 10, minGQ = 100, verbose = FALSE)
c_df <- filterSNPs(SNPset = c_df_nofilt, refAlleleFreq = 0.10, minTotalDepth = 20, maxTotalDepth = 130, depthDifference = 100, minSampleDepth = 10, minGQ = 100, verbose = TRUE)
d_df <- filterSNPs(SNPset = d_df_nofilt, refAlleleFreq = 0.10, minTotalDepth = 20, maxTotalDepth = 130, depthDifference = 100, minSampleDepth = 10, minGQ = 100, verbose = FALSE)
e_df <- filterSNPs(SNPset = e_df_nofilt, refAlleleFreq = 0.10, minTotalDepth = 20, maxTotalDepth = 130, depthDifference = 100, minSampleDepth = 10, minGQ = 100, verbose = FALSE)
f_df <- filterSNPs(SNPset = f_df_nofilt, refAlleleFreq = 0.10, minTotalDepth = 20, maxTotalDepth = 130, depthDifference = 100, minSampleDepth = 10, minGQ = 100, verbose = FALSE)
g_df <- filterSNPs(SNPset = g_df_nofilt, refAlleleFreq = 0.10, minTotalDepth = 20, maxTotalDepth = 130, depthDifference = 100, minSampleDepth = 10, minGQ = 100, verbose = FALSE)
h_df <- filterSNPs(SNPset = h_df_nofilt, refAlleleFreq = 0.10, minTotalDepth = 20, maxTotalDepth = 130, depthDifference = 100, minSampleDepth = 10, minGQ = 100, verbose = FALSE)
i_df <- filterSNPs(SNPset = i_df_nofilt, refAlleleFreq = 0.10, minTotalDepth = 20, maxTotalDepth = 130, depthDifference = 100, minSampleDepth = 10, minGQ = 100, verbose = FALSE)
j_df <- filterSNPs(SNPset = j_df_nofilt, refAlleleFreq = 0.10, minTotalDepth = 20, maxTotalDepth = 130, depthDifference = 100, minSampleDepth = 10, minGQ = 100, verbose = FALSE)
```

```{r depth plot, include=TRUE, fig.show="hold", fig.width=3, fig.height=3}
ggplot(data = a_df) +geom_histogram(aes(x = DP.HIGH + DP.LOW)) +ggtitle("a90%m") #+xlim(0,1000)
ggplot(data = b_df) +geom_histogram(aes(x = DP.HIGH + DP.LOW)) +ggtitle("b75%m")
ggplot(data = c_df) +geom_histogram(aes(x = DP.HIGH + DP.LOW)) +ggtitle("c60%m")
ggplot(data = d_df) +geom_histogram(aes(x = DP.HIGH + DP.LOW)) +ggtitle("d44%m")
ggplot(data = e_df) +geom_histogram(aes(x = DP.HIGH + DP.LOW)) +ggtitle("e80%m")
ggplot(data = f_df) +geom_histogram(aes(x = DP.HIGH + DP.LOW)) +ggtitle("f50%m")
ggplot(data = g_df) +geom_histogram(aes(x = DP.HIGH + DP.LOW)) +ggtitle("g45%m")
ggplot(data = h_df) +geom_histogram(aes(x = DP.HIGH + DP.LOW)) +ggtitle("h30_68%m")
ggplot(data = i_df) +geom_histogram(aes(x = DP.HIGH + DP.LOW)) +ggtitle("i21%m")
ggplot(data = j_df) +geom_histogram(aes(x = DP.HIGH + DP.LOW)) +ggtitle("j11%m")
```

```{r ref frequency, include=TRUE, fig.show="hold", fig.width= 3, fig.height=3}
ggplot(data = a_df) +geom_histogram(aes(x = REF_FRQ))+ggtitle("a90%m")
ggplot(data = b_df) +geom_histogram(aes(x = REF_FRQ))+ggtitle("b75%m")
ggplot(data = c_df) +geom_histogram(aes(x = REF_FRQ))+ggtitle("c60%m")
ggplot(data = d_df) +geom_histogram(aes(x = REF_FRQ))+ggtitle("d44%m")
ggplot(data = e_df) +geom_histogram(aes(x = REF_FRQ))+ggtitle("e80%m")
ggplot(data = f_df) +geom_histogram(aes(x = REF_FRQ))+ggtitle("f50%m")
ggplot(data = g_df) +geom_histogram(aes(x = REF_FRQ))+ggtitle("g45%m")
ggplot(data = h_df) +geom_histogram(aes(x = REF_FRQ))+ggtitle("h30_68%m")
ggplot(data = i_df) +geom_histogram(aes(x = REF_FRQ))+ggtitle("i21%m")
ggplot(data = j_df) +geom_histogram(aes(x = REF_FRQ))+ggtitle("j11%m")
```


```{r QTLseqAnalysis}
a_df_qtl <- runQTLseqAnalysis(a_df,windowSize = 2e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
b_df_qtl <- runQTLseqAnalysis(b_df,windowSize = 1e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
c_df_qtl <- runQTLseqAnalysis(c_df,windowSize = 1e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
d_df_qtl <- runQTLseqAnalysis(d_df,windowSize = 1e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
e_df_qtl <- runQTLseqAnalysis(e_df,windowSize = 1e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
f_df_qtl <- runQTLseqAnalysis(f_df,windowSize = 2e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
g_df_qtl <- runQTLseqAnalysis(g_df,windowSize = 1e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
h_df_qtl <- runQTLseqAnalysis(h_df,windowSize = 1e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
i_df_qtl <- runQTLseqAnalysis(i_df,windowSize = 1e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
j_df_qtl <- runQTLseqAnalysis(j_df,windowSize = 2e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99))
```

```{r GprimeAnalysis}
a_df_qtl <- runGprimeAnalysis(a_df_qtl, windowSize = 2e6, outlierFilter = "deltaSNP", filterThreshold = 0.1)
b_df_qtl <- runGprimeAnalysis(b_df_qtl, windowSize = 1e6, outlierFilter = "deltaSNP", filterThreshold = 0.1)
c_df_qtl <- runGprimeAnalysis(c_df_qtl, windowSize = 1e6, outlierFilter = "deltaSNP", filterThreshold = 0.1)
d_df_qtl <- runGprimeAnalysis(d_df_qtl, windowSize = 1e6, outlierFilter = "deltaSNP", filterThreshold = 0.1)
e_df_qtl <- runGprimeAnalysis(e_df_qtl, windowSize = 1e6, outlierFilter = "deltaSNP", filterThreshold = 0.1)
f_df_qtl <- runGprimeAnalysis(f_df_qtl, windowSize = 2e6, outlierFilter = "deltaSNP", filterThreshold = 0.1)
g_df_qtl <- runGprimeAnalysis(g_df_qtl, windowSize = 1e6, outlierFilter = "deltaSNP", filterThreshold = 0.1)
h_df_qtl <- runGprimeAnalysis(h_df_qtl, windowSize = 1e6, outlierFilter = "deltaSNP", filterThreshold = 0.1)
i_df_qtl <- runGprimeAnalysis(i_df_qtl, windowSize = 1e6, outlierFilter = "deltaSNP", filterThreshold = 0.1)
j_df_qtl <- runGprimeAnalysis(j_df_qtl, windowSize = 2e6, outlierFilter = "deltaSNP", filterThreshold = 0.1)
```
```{r}
sapply(a_df_qtl,class)
```

```{r plotGprimeDist, include=TRUE, fig.show="hold", fig.width= 5, fig.height=3}
a90_Hampel_filt <- plotGprimeDist(SNPset = a_df_qtl, outlierFilter = "Hampel")+ggtitle("a90%m")
b75_Hampel_filt <- plotGprimeDist(SNPset = b_df_qtl, outlierFilter = "Hampel")+ggtitle("b75%m")
c60_Hampel_filt <- plotGprimeDist(SNPset = c_df_qtl, outlierFilter = "Hampel")+ggtitle("c60%m")
d44_Hampel_filt <- plotGprimeDist(SNPset = d_df_qtl, outlierFilter = "Hampel")+ggtitle("d44%m")
e80_Hampel_filt <- plotGprimeDist(SNPset = e_df_qtl, outlierFilter = "Hampel")+ggtitle("e80%m")
f50_Hampel_filt <- plotGprimeDist(SNPset = f_df_qtl, outlierFilter = "Hampel")+ggtitle("f50%m")
g45_Hampel_filt <- plotGprimeDist(SNPset = g_df_qtl, outlierFilter = "Hampel")+ggtitle("g45%m")
h68_Hampel_filt <- plotGprimeDist(SNPset = h_df_qtl, outlierFilter = "Hampel")+ggtitle("h30_68%m")
i21_Hampel_filt <- plotGprimeDist(SNPset = i_df_qtl, outlierFilter = "Hampel")+ggtitle("i21%m")
j11_Hampel_filt <- plotGprimeDist(SNPset = j_df_qtl, outlierFilter = "Hampel")+ggtitle("j11%m")

a90_Hampel_filt
e80_Hampel_filt
b75_Hampel_filt
h68_Hampel_filt
c60_Hampel_filt
f50_Hampel_filt
g45_Hampel_filt
d44_Hampel_filt
i21_Hampel_filt
j11_Hampel_filt
```

```{r plotGprimeDist deltaSNP outlier, eval=FALSE, fig.show="hold", fig.width= 5, fig.height=3}
a90_deltaSNP_filt <- plotGprimeDist(SNPset = a_df_qtl, outlierFilter = "deltaSNP", filterThreshold = 0.1) +ggtitle("a90%m")
b75_deltaSNP_filt <- plotGprimeDist(SNPset = b_df_qtl, outlierFilter = "deltaSNP", filterThreshold = 0.1) +ggtitle("b75%m")
c60_deltaSNP_filt <- plotGprimeDist(SNPset = c_df_qtl, outlierFilter = "deltaSNP", filterThreshold = 0.1) +ggtitle("c60%m")
d44_deltaSNP_filt <- plotGprimeDist(SNPset = d_df_qtl, outlierFilter = "deltaSNP", filterThreshold = 0.1) +ggtitle("d44%m")
e80_deltaSNP_filt <- plotGprimeDist(SNPset = e_df_qtl, outlierFilter = "deltaSNP", filterThreshold = 0.1) +ggtitle("e80%m")
f50_deltaSNP_filt <- plotGprimeDist(SNPset = f_df_qtl, outlierFilter = "deltaSNP", filterThreshold = 0.1) +ggtitle("f50%m")
g45_deltaSNP_filt <- plotGprimeDist(SNPset = g_df_qtl, outlierFilter = "deltaSNP", filterThreshold = 0.1) +ggtitle("g45%m")
h68_deltaSNP_filt <- plotGprimeDist(SNPset = h_df_qtl, outlierFilter = "deltaSNP", filterThreshold = 0.1) +ggtitle("h30_68%m")
i21_deltaSNP_filt <- plotGprimeDist(SNPset = i_df_qtl, outlierFilter = "deltaSNP", filterThreshold = 0.1) +ggtitle("i21%m")
j11_deltaSNP_filt <- plotGprimeDist(SNPset = j_df_qtl, outlierFilter = "deltaSNP", filterThreshold = 0.1) +ggtitle("j11%m")

a90_deltaSNP_filt
e80_deltaSNP_filt
b75_deltaSNP_filt
h68_deltaSNP_filt
c60_deltaSNP_filt
f50_deltaSNP_filt
g45_deltaSNP_filt
d44_deltaSNP_filt
i21_deltaSNP_filt
j11_deltaSNP_filt
```

```{r plot QTLstats nSNPs, include=TRUE, fig.show="hold", fig.width= 3, fig.height=3}
a90_plotQTLS <- plotQTLStats(SNPset = a_df_qtl, var = "nSNPs")+ggtitle("a90%m")+ylim(0,25000)
b75_plotQTLS <- plotQTLStats(SNPset = b_df_qtl, var = "nSNPs")+ggtitle("b75%m")+ylim(0,25000)
c60_plotQTLS <- plotQTLStats(SNPset = c_df_qtl, var = "nSNPs")+ggtitle("c60%m")+ylim(0,25000)
d44_plotQTLS <- plotQTLStats(SNPset = d_df_qtl, var = "nSNPs")+ggtitle("d44%m")+ylim(0,25000)
e80_plotQTLS <- plotQTLStats(SNPset = e_df_qtl, var = "nSNPs")+ggtitle("e80%m")+ylim(0,25000)
f50_plotQTLS <- plotQTLStats(SNPset = f_df_qtl, var = "nSNPs")+ggtitle("f50%m")+ylim(0,25000)
g45_plotQTLS <- plotQTLStats(SNPset = g_df_qtl, var = "nSNPs")+ggtitle("g45%m")+ylim(0,25000)
h68_plotQTLS <- plotQTLStats(SNPset = h_df_qtl, var = "nSNPs")+ggtitle("h30_68%m")+ylim(0,25000)
i21_plotQTLS <- plotQTLStats(SNPset = i_df_qtl, var = "nSNPs")+ggtitle("i21%m")+ylim(0,25000)
j11_plotQTLS <- plotQTLStats(SNPset = j_df_qtl, var = "nSNPs")+ggtitle("j11%m")+ylim(0,25000)

a90_plotQTLS
e80_plotQTLS
b75_plotQTLS
h68_plotQTLS
c60_plotQTLS
f50_plotQTLS
g45_plotQTLS
d44_plotQTLS
i21_plotQTLS
j11_plotQTLS
```


```{r plot QTLstats deltaSNP, include=TRUE, fig.show="hold", fig.width= 3, fig.height=3}
a90_plotQTLSdelta <- plotQTLStats(SNPset = a_df_qtl, var = "deltaSNP", plotIntervals = TRUE)+ggtitle("a90%m")
b75_plotQTLSdelta <- plotQTLStats(SNPset = b_df_qtl, var = "deltaSNP", plotIntervals = TRUE)+ggtitle("b75%m")
c60_plotQTLSdelta <- plotQTLStats(SNPset = c_df_qtl, var = "deltaSNP", plotIntervals = TRUE)+ggtitle("c60%m")
d44_plotQTLSdelta <- plotQTLStats(SNPset = d_df_qtl, var = "deltaSNP", plotIntervals = TRUE)+ggtitle("d44%m")
e80_plotQTLSdelta <- plotQTLStats(SNPset = e_df_qtl, var = "deltaSNP", plotIntervals = TRUE)+ggtitle("e80%m")
f50_plotQTLSdelta <- plotQTLStats(SNPset = f_df_qtl, var = "deltaSNP", plotIntervals = TRUE)+ggtitle("f50%m")
g45_plotQTLSdelta <- plotQTLStats(SNPset = g_df_qtl, var = "deltaSNP", plotIntervals = TRUE)+ggtitle("g45%m")
h68_plotQTLSdelta <- plotQTLStats(SNPset = h_df_qtl, var = "deltaSNP", plotIntervals = TRUE)+ggtitle("h30_68%m")
i21_plotQTLSdelta <- plotQTLStats(SNPset = i_df_qtl, var = "deltaSNP", plotIntervals = TRUE)+ggtitle("i21%m")
j11_plotQTLSdelta <- plotQTLStats(SNPset = j_df_qtl, var = "deltaSNP", plotIntervals = TRUE)+ggtitle("j11%m")

a90_plotQTLSdelta
e80_plotQTLSdelta
b75_plotQTLSdelta
h68_plotQTLSdelta
c60_plotQTLSdelta
f50_plotQTLSdelta
g45_plotQTLSdelta
d44_plotQTLSdelta
i21_plotQTLSdelta
j11_plotQTLSdelta
```

```{r plotQTLstats Gprime, fig.show="hold", include=TRUE, fig.width= 3, fig.height=3}
a90_plotgp <- plotQTLStats(SNPset = a_df_qtl, var = "Gprime", plotThreshold = TRUE, q = 0.01)+ggtitle("a90%m")+ylim(0,7)
b75_plotgp <- plotQTLStats(SNPset = b_df_qtl, var = "Gprime", plotThreshold = TRUE, q = 0.01)+ggtitle("b75%m")+ylim(0,7)
c60_plotgp <- plotQTLStats(SNPset = c_df_qtl, var = "Gprime", plotThreshold = TRUE, q = 0.01)+ggtitle("c60%m")+ylim(0,7)
d44_plotgp <- plotQTLStats(SNPset = d_df_qtl, var = "Gprime", plotThreshold = TRUE, q = 0.01)+ggtitle("d44%m")+ylim(0,7)
e80_plotgp <- plotQTLStats(SNPset = e_df_qtl, var = "Gprime", plotThreshold = TRUE, q = 0.01)+ggtitle("e80%m")+ylim(0,7)
f50_plotgp <- plotQTLStats(SNPset = f_df_qtl, var = "Gprime", plotThreshold = TRUE, q = 0.01)+ggtitle("f50%m")+ylim(0,7)
g45_plotgp <- plotQTLStats(SNPset = g_df_qtl, var = "Gprime", plotThreshold = TRUE, q = 0.01)+ggtitle("g45%m")+ylim(0,7)
h68_plotgp <- plotQTLStats(SNPset = h_df_qtl, var = "Gprime", plotThreshold = TRUE, q = 0.01)+ggtitle("h30_68%m")+ylim(0,7)
i21_plotgp <- plotQTLStats(SNPset = i_df_qtl, var = "Gprime", plotThreshold = TRUE, q = 0.01)+ggtitle("i21%m")+ylim(0,7)
j11_plotgp <- plotQTLStats(SNPset = j_df_qtl, var = "Gprime", plotThreshold = TRUE, q = 0.01)+ggtitle("j11%m")+ylim(0,7)

a90_plotgp
e80_plotgp
b75_plotgp
h68_plotgp
c60_plotgp
f50_plotgp
g45_plotgp
d44_plotgp
i21_plotgp
j11_plotgp
```

```{r}
#aqtl <- getQTLTable(SNPset = a_df_qtl, method = "Gprime", alpha = 0.01, export=FALSE)

#fqtl <- getQTLTable(SNPset = f_df_qtl, method = "Gprime", alpha = 0.01, export=FALSE)

#jqtl <- getQTLTable(SNPset = j_df_qtl, method = "Gprime", alpha = 0.01, export=FALSE)
```

