---
title: "qtlseqr"
author: "Casey Flay"
date: "12/08/2021"
output:
  html_document:
    df_print: paged
  word_document: default
---

```{r setup, include=FALSE}
library(knitr)
library(RLinuxModules)
library(data.table)
library(glue) 
library(tidyverse)
library(here)
library(QTLseqr)
setDTthreads(1)
module("load openlava asub/2.2 bwa/0.7.12 samtools/1.9 bcftools/1.10.2 perlbrew/0.76")
knitr::opts_chunk$set(echo = FALSE, include=FALSE, warning = TRUE)
```

```{r transform filenames to sample, lane and suffix for data.table from symlink}
compare_mpool  <- as.data.table(grep(paste("malepool_compare.csv"), list.files(path="."), invert=FALSE, value=TRUE))
compare_mpool  <- compare_mpool[, V2 := str_replace_all(V1, "_malepool_compare.csv", "&_malepool_compare.csv")
                  ][, V2 := str_replace_all(V2, ".csv" , "&.csv")
                  ][, c("V2", "V3", "V4") := tstrsplit(V2, "&", fixed=TRUE)]
compare_mpool  <- compare_mpool[, "V4" := str_sub(compare_mpool[,V1], 1, 1)] #get column for first letter
compare_mpool  <- compare_mpool[V4=="E"]#V4=="F"|V4=="G"|V4=="H"|V4=="I"#too big to run all at once.  cycle through A,BC,DE,FGHIJ manually
```

```{r group samples against parent pools and rename for QTLseqR and assign columns for QTLseqR}
obj_names <- compare_mpool[,V2]
my_files  <- compare_mpool[,V1]

#70gb RAM load below, only run when reduction filters are inplace
for(i in seq(along=my_files)) {
  assign(obj_names[i], fread(my_files[i]))
}

#relabel samples for qtlseqR
samplesold <- c('s1CHROM','s1POS','s1REF','s1ALT','r1AD0','r1AD1','r1DP','r1GQ','r1PL','r1SNPi','s1AD0','s1AD1','s1DP','s1GQ','s1PL','s1SNPi','REF_FRQ','deltaSNP')
samplesnew <-c('CHROM','POS','REF','ALT','AD_REF.LOW','AD_ALT.LOW','DP.LOW','GQ.LOW','PL.LOW','SNPindex.LOW','AD_REF.HIGH','AD_ALT.HIGH','DP.HIGH','GQ.HIGH','PL.HIGH','SNPindex.HIGH','REF_FRQ','deltaSNP')

for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  dt_name <-assign(obj_names[i], setnames(dt_name, glue("{samplesold}"), glue("{samplesnew}")))
}
```

```{r clean columns and change to chr01 for ggplot to assemble in the right order}

for(i in seq(along=obj_names)) {
  assign(obj_names[i], get(obj_names[i], envir = .GlobalEnv)[, "CHROM" := as.factor(CHROM)])
}
#sapply(F12G13, class)
chrs <- c("chr1", "chr2", "chr3",  "chr4",  "chr5",  "chr6",  "chr7",  "chr8",  "chr9", "chr10", "chr11", "chr12", "chr13", "chr14", "chr15", "chr16", "chr17", "chr18", "chr19",  "chr20", "chr21", "chr22", "chr23", "chr24", "chr25", "chr26", "chr27", "chr28", "chr29")

for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  dt_name <-assign(obj_names[i], dt_name[, CHROM := fct_relevel(dt_name[,CHROM], chrs, after = Inf)])
}
```

```{r filter, echo=TRUE}
for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  dt_name <- assign(obj_names[i], filterSNPs(SNPset = dt_name, refAlleleFreq = 0.05, minTotalDepth = 60, maxTotalDepth = 160, depthDifference = 50, minSampleDepth = 10, minGQ = 100, verbose = FALSE))
}
```


```{r depth plot, include=TRUE, fig.show="hold", fig.width=3, fig.height=3, eval=FALSE}
for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  print(ggplot(data = dt_name) +geom_histogram(aes(x = DP.HIGH + DP.LOW)))+ggtitle(obj_names[i])+ylim(0,120000)
}
```

```{r ref frequency, include=TRUE, fig.show="hold", fig.width= 3, fig.height=3, eval=FALSE}
for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  print(ggplot(data = dt_name) +geom_histogram(aes(x = REF_FRQ))+ggtitle(obj_names[i]))
}
```

```{r QTLseqAnalysis}
for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  dt_name <-assign(obj_names[i], runQTLseqAnalysis(dt_name,windowSize = 1e6, popStruc = "F1", bulkSize = c(10,10), replications = 10000, intervals = c(95, 99)))
}
```

```{r GprimeAnalysis}
for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  dt_name <- assign(obj_names[i], runGprimeAnalysis(dt_name, windowSize = 1e6, outlierFilter = "deltaSNP", filterThreshold = 0.4))
}
```

```{r plotGprimeDist, eval=FALSE, fig.show="hold", fig.width= 5, fig.height=3}
for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  print(plotGprimeDist(SNPset = dt_name, outlierFilter = "Hampel")+ggtitle(obj_names[i]))
}
```

```{r plotGprimeDist deltaSNP outlier, eval=FALSE, fig.show="hold", fig.width= 5, fig.height=3}
for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  print(plotGprimeDist(SNPset = dt_name, outlierFilter = "deltaSNP", filterThreshold = 0.4) +ggtitle(obj_names[i]))
}
```

```{r plot QTLstats nSNPs, eval=FALSE, include=TRUE, fig.width= 20}
for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  print(plotQTLStats(SNPset = dt_name, var = "nSNPs") +ggtitle(obj_names[i]))+ylim(0,25000)
}
```

```{r plot QTLstats deltaSNP, eval=FALSE, include=TRUE, fig.width= 20}
for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  print(plotQTLStats(SNPset = dt_name, var = "deltaSNP", plotIntervals = TRUE) +ggtitle(obj_names[i]))
}
```

```{r males plotQTLstats Gprime, eval=FALSE, warnings=TRUE, include=TRUE, fig.show="hold", fig.width= 20}
#using var= "negLog10Pval" instead of "Gprime" seems to give better results
for(i in seq(along=obj_names)) {
  dt_name <- get(obj_names[i], envir = .GlobalEnv)
  print(plotQTLStats(SNPset = dt_name, var = "negLog10Pval", plotThreshold = TRUE, q = 0.01) + labs(title = obj_names[i]) + theme(plot.title = element_text(size = 18, hjust = 0.5))+ylim(0,8))
#ggsave(paste0(obj_names[i],"_plotgp.png"), plot = last_plot(), width = 35, height = 12, units = c("cm"), dpi = 72)
}
```

```{r reduce column number, eval=TRUE, warnings=FALSE, include=TRUE, fig.show="hold", fig.width= 15}
for(i in seq(along=obj_names)) {
dt_name <- as.data.table(get(obj_names[i], envir = .GlobalEnv))

dt_name <- assign(obj_names[i], dt_name[,"POSmb" := POS/1000000])
quant<- as.numeric(quantile(dt_name[,negLog10Pval], probs = 0.99, na.rm=T))#probs0.999 gives top 0.001% of SNPs
dt_name <- assign(obj_names[i], dt_name[negLog10Pval>quant, "top":="Top 1%"][negLog10Pval<quant, top:="Remaining SNPs"])
threshold <- min(dt_name[qvalue<0.05, negLog10Pval])

if(!is.null(dt_name$AD_REF.HIGH)){
  dt_name <- dt_name[, `:=` ("AD_REF.HIGH"=NULL,"V19"=NULL, "AD_ALT.HIGH"=NULL, "DP.HIGH"=NULL, "GQ.HIGH"=NULL, "PL.HIGH"=NULL, "Cpos"=NULL,  "r1CHROM"=NULL,   "r1POS"=NULL, "r1REF"=NULL, "r1ALT"=NULL, "AD_REF.LOW"=NULL, "AD_ALT.LOW"=NULL, "DP.LOW"=NULL,  "GQ.LOW"=NULL, "PL.LOW"=NULL, "SNPindex.HIGH"=NULL, "SNPindex.LOW"=NULL, "REF_FRQ"=NULL, "deltaSNP"=NULL, "nSNPs"=NULL, "tricubeDeltaSNP"=NULL, "minDP"=NULL, "tricubeDP"=NULL, "CI_95"=NULL, "CI_99"=NULL, "G"=NULL)]
}
}

```

```{r ggplot qtls, eval=FALSE, warnings=FALSE, include=TRUE, fig.show="hold", fig.width= 15}
for(i in seq(along=obj_names)) {
dt_name   <- as.data.table(get(obj_names[i], envir = .GlobalEnv))

dt_name   <- assign(obj_names[i], dt_name[,"POSmb" := POS/1000000])
quant     <- as.numeric(quantile(dt_name[,negLog10Pval], probs = 0.99, na.rm=T))#probs0.999 gives top 0.001% of SNPs
dt_name   <- assign(obj_names[i], dt_name[negLog10Pval>quant, "top":="Top 0.01%"][negLog10Pval<quant, top:="Remaining SNPs"])
threshold <- min(dt_name[qvalue<0.05, negLog10Pval])
threshold[!is.finite(threshold)] <- 0

dt_name   <- dt_name[CHROM=="chr25",]#just chr25

if  (threshold==0){
    #if there is a significant threshold to plot apply hline threshold to graph
p <- ggplot(data=dt_name, aes(x=POSmb, y=negLog10Pval))+
  labs(title = obj_names[i]) +
  scale_fill_discrete(guide=FALSE)+
  geom_point(aes(color=factor(dt_name[,top])),size=0.5, na.rm=T)+
  scale_color_manual(name="top",breaks = c("Top 0.01%","Remaining SNPs"),values=c("chartreuse3","black"))+
  scale_x_continuous(limits=c(0, 20), breaks=c(0, 5, 10, 15))+
  xlab("Nucleotide Position")+  
  ylab("-Log10 p-value")+
  facet_wrap(~CHROM, nrow=3, ncol=10)
  print(p)
print(p)
} else {
  #else apply graph with no threshold line
  p <- ggplot(data=dt_name, aes(x=POSmb, y=negLog10Pval))+
  labs(title = obj_names[i], ) +
  scale_fill_discrete(guide=FALSE)+
  geom_point(aes(color=factor(dt_name[,top])),size=0.5, na.rm=T)+
  geom_hline(yintercept = threshold, linetype='dashed')+  
  scale_color_manual(name="top",breaks = c("Top 0.01%","Remaining SNPs"),values=c("chartreuse3","black"))+
  scale_x_continuous(limits=c(0, 20), breaks=c(0, 5, 10, 15))+
  xlab("Nucleotide Position")+  
  ylab("-Log10 p-value")+
  facet_wrap(~CHROM, nrow=3, ncol=10)
  print(p)
}

ggsave(paste0(obj_names[i],"_chr25_mplotg.png"), plot = last_plot(), width = 10, height = 7, units = c("cm"), dpi = 320)
}
```

```{r write top 1percent snps and sig snps so seperate plots can be made of each for chr25}
for(i in seq(along=obj_names)) {
dt_name <- as.data.table(get(obj_names[i], envir = .GlobalEnv))

dt_name   <- dt_name[,"parent" := obj_names[i]]
dt_name   <- assign(obj_names[i], dt_name[,"POSmb" := POS/1000000])
quant     <- as.numeric(quantile(dt_name[,negLog10Pval], probs = 0.99, na.rm=T))#probs0.999 gives top 0.001% of SNPs
dt_name   <- assign(obj_names[i], dt_name[negLog10Pval>quant, "top":="Top0.01_percent"][negLog10Pval<quant, top:="Remaining_SNPs"])

dt_name   <- dt_name[, "neglog10threshold":= min(dt_name[qvalue<0.05, negLog10Pval])][neglog10threshold==Inf, neglog10threshold:=100]

dt_name   <- dt_name[qvalue>0.05, "sig" := "insignificant"][qvalue<0.05, sig := "significant"]

dt_name   <- dt_name[top=="Top0.01_percent"|sig == "significant",] #filter just top and significant
fwrite(dt_name, file="de_topandsig_maleaf_snps.csv", append=T) #rename this chunk for A,BC,DE,FGHI runs
}
```

```{r ggplot significant and top1%graphs, eval=FALSE, warnings=FALSE, include=TRUE, fig.show="hold", fig.width= 15}
s1    <- fread("a_topandsig_maleaf_snps.csv")
s2    <- fread("bc_topandsig_maleaf_snps.csv")
s3    <- fread("de_topandsig_maleaf_snps.csv")
s4    <- fread("fghi_topandsig_maleaf_snps.csv")

sigs   <- rbindlist(list(s1,s2,s3,s4))

sigs   <- sigs[CHROM=="chr25"&parent!="A1D6"&parent!="B2D6"&parent!="C3D6"&parent!="D6E11"&parent!="D6F12"&parent!="D6G13"&parent!="D6H14"&parent!="D6I15"&parent!="D6J16",]#just chr25, and exclude D

#Add male percentage for colour scale
comparison_percentage <- fread("comparison_percentage.csv")# from\Projects\Fst%20mapping\individuals%20pooled\AlleleFrequencyTestParentage.xlsx
for(i in seq(along=comparison_percentage[,V1])) {
  i <- comparison_percentage[i,]
  sigs <- sigs[parent == i$V1, "male_diff":=  i$V2]
}

qsig <- sigs[sig=="significant",]
lq <- ggplot(data=qsig, aes(x=POSmb, y=negLog10Pval))+
  labs(title = "Significant SNPs on chr25") +
  geom_point(aes(color=male_diff),size=0.01, na.rm=T)+
  #scale_color_manual(name="top",breaks = c("Top 0.01%","Remaining SNPs"),values=c("chartreuse3","black"))+
  scale_x_continuous(limits=c(0, 15), breaks=c(0, 5, 10, 15))+
  xlab("Nucleotide Position")+  
  ylab("-Log10 p-value")+
  facet_wrap(~CHROM, nrow=3, ncol=10)
lq

topperc <- sigs[top=="Top0.01_percent",]
lt <- ggplot(data=topperc, aes(x=POSmb, y=negLog10Pval))+
  labs(title = "Top 1% of SNPs on chr25") +
  geom_point(aes(color=male_diff),size=0.01, na.rm=T)+
  #scale_color_manual(name="top",breaks = c("Top 0.01%","Remaining SNPs"),values=c("chartreuse3","black"))+
  scale_x_continuous(limits=c(0, 15), breaks=c(0, 5, 10, 15))+
  xlab("Nucleotide Position")+  
  ylab("-Log10 p-value")+
  facet_wrap(~CHROM, nrow=3, ncol=10)
lt

qsig <- sigs[sig=="significant",]
gq <- ggplot(data=qsig, aes(x=POSmb, y=Gprime))+
  labs(title = "Significant SNPs on chr25") +
  geom_point(aes(color=male_diff),size=0.01, na.rm=T)+
  #scale_color_manual(name="top",breaks = c("Top 0.01%","Remaining SNPs"),values=c("chartreuse3","black"))+
  scale_x_continuous(limits=c(0, 15), breaks=c(0, 5, 10, 15))+
  xlab("Nucleotide Position")+  
  ylab("-Gprime")+
  facet_wrap(~CHROM, nrow=3, ncol=10)
gq

topperc <- sigs[top=="Top0.01_percent",]
gt <- ggplot(data=topperc, aes(x=POSmb, y=Gprime))+
  labs(title = "Top 1% of SNPs on chr25") +
  geom_point(aes(color=male_diff),size=0.01, na.rm=T)+
  #scale_color_manual(name="top",breaks = c("Top 0.01%","Remaining SNPs"),values=c("chartreuse3","black"))+
  scale_x_continuous(limits=c(0, 15), breaks=c(0, 5, 10, 15))+
  xlab("Nucleotide Position")+  
  ylab("-Gprime")+
  facet_wrap(~CHROM, nrow=3, ncol=10)
gt

ggsave("neglogp_q_Significant_SNPs_on_chr25.png", plot = lq, width = 15, height = 10, units = c("cm"), dpi = 320)
ggsave("neglogp_Top0.01_percent_SNPs_on_chr25.png", plot = lt, width = 15, height = 10, units = c("cm"), dpi = 320)
ggsave("Gprime_q_Significant_SNPs_on_chr25.png", plot = gq, width = 15, height = 10, units = c("cm"), dpi = 320)
ggsave("Gprime_Top1perc_SNPs_on_chr25.png", plot = gt, width = 15, height = 10, units = c("cm"), dpi = 320)
```



