---
title: "Male allele frequency"
author: "Casey"
date: "03/08/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(RLinuxModules)
library(data.table)
library(glue)
library(here)
library(tidyverse)
library(stringr)
module("load openlava asub bwa/0.7.12 samtools/1.9 bcftools/1.10.2")
```

```{r}
mainDir <- "/powerplant/workspace/hrtcdf/github/FSTs/"
subDir <- "bamlink"
royDir <- "/workspace/hrards/projects/hrtcdf/10.alignments/bwa"
outDir <- str_c(mainDir, subDir, sep = "", collapse = TRUE)
outDir

if (dir.exists(mainDir)){
    setwd(file.path(mainDir, subDir))
} else {
    dir.create(file.path(mainDir, subDir))
    setwd(file.path(mainDir, subDir))
}

tpoolfiles <- dir(royDir, pattern = "^HL55WDRXX", full.names = TRUE, include.dirs = TRUE)
file.symlink (tpoolfiles, outDir)
```

```{bash}
cd /powerplant/workspace/hrtcdf/github/FSTs/TestPool/bamlinkindexed
for bam in $(ls H*) ;do 
       input="/powerplant/workspace/hrtcdf/github/FSTs/bamlinkindexed/${bam}"
       logDir="/powerplant/workspace/hrtcdf/github/FSTs/TestPool/bamlinkindexed/logDir"
       outDir="/powerplant/workspace/hrtcdf/github/FSTs/TestPool/bamlinkindexed"
       bsub -n 8 -o $logDir/${bam}.log -e $logDir/${bam}.err "samtools index $bam"
       done
```

```{r transform filenames to sample and directory for data.table}
files <- dir("/workspace/hrards/projects/hrtcdf/10.alignments/bwa", pattern = "^HL55WDRXX", full.names = TRUE)
apfiles <- str_replace(files, "HL55WDRXX_" , "HL55WDRXX_&")

dt <- {data.table(apfiles) %>%
  separate(apfiles, into = c("dir", "sample"), sep = "&", fill = "left") %>%
  mutate("version" = str_extract(dir, regex("/")))
}

colnames(dt) <- c("dir", "sample", "version")
write.table(dt, 'testpool.tsv')
```

```{r transform filenames to sample, lane and suffix for data.table from link}
lindts <- as.data.table(grep(paste(".bam.gz$", collapse = "|"), list.files(path="."), invert=FALSE, value=TRUE))
lindts <- lindts[, V1 := str_replace_all(V1, "_L" , "&_L")]
lindts <- lindts[, V1 := str_replace_all(V1, ".bam" , "&.bam&")]
lindts <- lindts[, c("sample", "lane", "bam", "version") := tstrsplit(V1, "&", fixed=TRUE)]
lindts <- lindts[, V1 := str_replace_all(V1,"&","")]
lindts
fwrite(lindts, 'linkpool.tsv', append=FALSE)
```

```{r run-mileup for chr25}
submit_mpileup <- function(reference = "*.fa.gz", output = c(), bamfiles = c(), job.init = "asub", job.opts = "-q normal") {
  cmd <- "bcftools mpileup -B -f {reference} -r chr25 -a {mpileupTags} {bamfiles} -Ou | bcftools call -f {bcfcallTags} -p {pvalThres} -m -A -Oz -o {output}"
  tmpf <- tempfile(pattern = "submit_mileup.", fileext = ".sh")
  writeLines(text = c(glue(cmd)), con = tmpf)
  submit <- "{job.init} {job.opts} {tmpf}"
  system(glue(submit), intern = TRUE)
}
mpileupTags <- "DP,AD,SP,INFO/AD"
bcfCallTags <- "GQ,GP"
  pvalThres <- 0.99

bamfiles    <- paste0(lindts$V1)
sample      <- paste(lindts$sample,lindts$lane, sep = "")
sample      <- paste(tools::file_path_sans_ext(sample), "c25.vcf.gz", sep = ".")
readgroup   <- paste(tools::file_path_sans_ext(lindts$sample))

reference   <- dir("/workspace/hrards/projects/hrtcdf/genome/fasta", pattern = ".*\\.gz$", full.names = TRUE)
reference
file.exists(bamfiles)
file.exists(reference)

submit_mpileup(reference = reference, output = sample, bamfiles = bamfiles)
```


```{bash}
zcat CK02_01_L001.bam.gz | grep -v "##" | cut -f1,2,3,4,5,6,7,8,9,10 | less
```



##change file locations and work out the minor allele frequency for each of the pools.
##minor allele frequency, average depth, and fraction_missing will need to be adjusted for each pool.
## dont include INFO/F_MISSING as there is no hedder for it. added to chr25 set
## there was no allele frequency from mpileup.... it wasn't added as a tag. go back down the snake to add.

```{bash #filter}
cd "/powerplant/workspace/hrtcdf/github/FSTs/TestPool"
if  [ ! -d "15_filter" ]; then mkdir "15_filter" ; fi

t=8
#set=M.01_d5_ld100_q10_f0.8_g10
set=f99q5
for bam in $(ls CK02_01_L001*) ;do ## just leave the CK* star for full run.
       input="/powerplant/workspace/hrtcdf/FSTs/TestPool/${bam}"
       logDir="/powerplant/workspace/hrtcdf/FSTs/TestPool/15_filter/logDir"
       outDir="/powerplant/workspace/hrtcdf/FSTs/TestPool/15_filter"
       if  [ ! -d "15_filter/logDir" ]; then mkdir "15_filter/logDir" ; fi

#bsub -n $t -o $logDir/${bam}.log -e $logDir/${bam}.err "bcftools stats $input > $input.stats"

bsub -n $t -o $logDir/${set}${bam}.log -e $logDir/${set}${bam}.err "bcftools view -i 'AVG(INFO/DP)>5 && AVG(INFO/DP)<99 && QUAL>=5' $bam | bcftools filter -g10 -G10 -Oz -o $outDir/${set}${bam} && bcftools stats $outDir/${set}${bam} > $outDir/${set}${bam}.stats"

done
```

```{r data from vcfs to data table}
#ALT{0} retrieves the first alternate allele while AD{0} retrieves the depth for the reference with AD{1} the depth for the first alternate. column names adjusted to make ALT1 the same as AD1.
hc  <- "bcftools view  'Hort16A.vcf.gz' -Ou | bcftools norm --multiallelics +any -Ou | bcftools query  -f'[%SAMPLE]\t%POS\t%REF\t%ALT\t%ALT{0}\t%ALT{1}\t%ALT{2}\t%DP\t%AD\t%AD{0}\t%AD{1}\t%AD{2}\t%AD{3}\n'"
h <- fread(cmd=hc, col.names = c('hSAMPLE','POS','hREF','hALT','hALT1','hALT2','hALT3','hDP','hAD','hREFD','hAD1','hAD2','hAD3'), key = 'POS')

oc  <- "bcftools view  'CK01_01_01_01.vcf.gz' -Ou | bcftools norm --multiallelics +any -Ou | bcftools query  -f'[%SAMPLE]\t%POS\t%REF\t%ALT\t%ALT{0}\t%ALT{1}\t%ALT{2}\t%DP\t%AD\t%AD{0}\t%AD{1}\t%AD{2}\t%AD{3}\n'"
o <- fread(cmd=oc, col.names = c('oSAMPLE','POS','oREF','oALT','oALT1','oALT2','oALT3','oDP','oAD','oREFD','oAD1','oAD2','oAD3'), key = 'POS')

fc  <- "bcftools view  'T94.30-03-10f.vcf.gz' -Ou | bcftools norm --multiallelics +any -Ou | bcftools query  -f'[%SAMPLE]\t%POS\t%REF\t%ALT\t%ALT{0}\t%ALT{1}\t%ALT{2}\t%DP\t%AD\t%AD{0}\t%AD{1}\t%AD{2}\t%AD{3}\n'"
f <- fread(cmd=fc, col.names = c('fSAMPLE','POS','fREF','fALT','fALT1','fALT2','fALT3','fDP','fAD','fREFD','fAD1','fAD2','fAD3'), key = 'POS')

cc  <- "bcftools view  'T94.30-04-08c.vcf.gz' -Ou | bcftools norm --multiallelics +any -Ou | bcftools query  -f'[%SAMPLE]\t%POS\t%REF\t%ALT\t%ALT{0}\t%ALT{1}\t%ALT{2}\t%DP\t%AD\t%AD{0}\t%AD{1}\t%AD{2}\t%AD{3}\n'"
c <- fread(cmd=cc, col.names = c('cSAMPLE','POS','cREF','cALT','cALT1','cALT2','cALT3','cDP','cAD','cREFD','cAD1','cAD2','cAD3'), key = 'POS')

bc  <- "bcftools view  'T94.30-04-08b.vcf.gz' -Ou | bcftools norm --multiallelics +any -Ou | bcftools query  -f'[%SAMPLE]\t%POS\t%REF\t%ALT\t%ALT{0}\t%ALT{1}\t%ALT{2}\t%DP\t%AD\t%AD{0}\t%AD{1}\t%AD{2}\t%AD{3}\n'"
b <- fread(cmd=bc, col.names = c('bSAMPLE','POS','bREF','bALT','bALT1','bALT2','bALT3','bDP','bAD','bREFD','bAD1','bAD2','bAD3'), key = 'POS')

tolmatrix  <- merge.data.table(h,o, all=TRUE,  by ='POS')
tolmatrix  <- merge.data.table(tolmatrix,f, all=TRUE,  by ='POS')
tolmatrix  <- merge.data.table(tolmatrix,c, all=TRUE,  by ='POS')
tolmatrix  <- merge.data.table(tolmatrix,b, all=TRUE,  by ='POS')
rm(h,o,f,c,b)
fwrite(tolmatrix, file="tolmatrix.csv", append=FALSE)
#tm0s2 <- tolmatrix[POS %between% c(6185769, 6188976),]
```

##Add RG info https://github.com/Actinidia/FlavourGenetics/blob/afe4f2334ca7a3a95b1738e16e4c63e988f9aef6/DryMatter/ExpRequestor10619/2019-07-29-ER10619-Align-QC.ipynb
## helpfull-- https://www.poftut.com/linux-cut-command-examples/
```{bash}
inDir="/powerplant/workspace/hrtcdf/FSTs/TestPool/"

for SAMP in $inDir/$(ls CK*) ;do
name=`basename $SAMP .bam`
lane=$(echo `basename $SAMP .bam.gz` | cut -c11-)
SM=$(echo `basename $SAMP .bam.gz` | cut -c-12)
RG=${SM}_$lane
line=$(echo -e "-r 'ID:${RG}' -r 'LB:1' -r 'SM:${SM}'")
echo $name
echo $SM
echo $RG
echo $lane
echo $line

#echo "samtools addreplacerg ${line} -o $OUTPUT/${name}.RG.bam $SAMP"
done #| asub -j asgsdfhsd
```


```{bash}
pwd
##Use these in terminal
## zcat CK02_01_L001.bcf.gz | grep -v "##" | cut -f1,2,3,4,5,6,7,8,9,10 | less -n 20
## zcat bwa_subset_filtered.vcf | grep '>' | more
```




