---
title: Principal Component Analysis applied directly to Sequence Matrix from Tomokazu
  Konishi, QTL applied to all individuals with WGS
author: "Casey Flay"
date: "27/12/2021"
output:
  word_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, include=FALSE, fig.width=12, fig.height=8, fig.path='Figs/')
library(RLinuxModules)
library(glue)
library(here)
library(data.table)
library(tidyverse)
library(ggplot2)
library(magrittr)
library(GenomicRanges)
library(IRanges)
library(RColorBrewer)
library(rgl)
library(tools)
library(rlist)
library(qpcR)
setDTthreads(1)
module("load openlava asub/2.2 bwa/0.7.12 samtools/1.9 bcftools/1.10.2 vcftools")
```

##get parents from all files

```{r parents NCBI aligned bamfiles from male pool and testpools, eval=FALSE}
Rualn_bamDir <- "/output/genomic/plant/Actinidia/chinensis/Resequencing/Alignments/Russell_V2a.chromosomes.and.unassigned"
Rualn_bams   <- as.data.table(grep(paste('.bam$', collapse = "|"), list.files(path=Rualn_bamDir), invert=FALSE, value=TRUE))
Rualn_bams   <- Rualn_bams[, "V2" := str_replace_all(V1, "_Russell_V2a", "&_Russell_V2a&")]
Rualn_bams   <- Rualn_bams[, c("PedigreeItem","reference",  "ext") := tstrsplit(V2, "&", fixed=TRUE)]
Rualn_bams   <- Rualn_bams[, PedigreeItem := str_replace_all(PedigreeItem, c("ER\\d{5}_|79-1-1"), "")]
Rualn_bams   <- Rualn_bams[,"remove" := str_detect(PedigreeItem, c("SRA_|UNK_|K12.|M97|165_02_03|K15.|M15|M10|M08|K13.|DA|HH|ME|RE|T94.30-04-08d|EA01_02"), negate = FALSE)]
Rualn_bams   <- Rualn_bams[remove!="TRUE",]
Rualn_bams   <- Rualn_bams[,remove := str_detect(V1, c("ER10645|ER10646|Zara|ER10454|ER10439|ER10522_CK51_06|ER10522_CK51_09|KuimiSport"), negate = FALSE)][remove!="TRUE",]
Rualn_bams   <- Rualn_bams[, "pathandfile" := paste(Rualn_bamDir, V1, sep="/")]
#Rualn_bams   <- Rualn_bams[PedigreeItem=="Red19", PedigreeItem := ("Red 19")][PedigreeItem=="Gold3", PedigreeItem := ("Gold 3")][PedigreeItem=="Gold9", PedigreeItem := ("Gold 9")]

Rualn_bams   <- Rualn_bams[, "isfile" := (file.exists(pathandfile))][isfile != FALSE,]

setkeyv(Rualn_bams,"PedigreeItem")
PID_avaliable_PSAphen <- data.table(list("CK01_01_01_01",	"CK02_01",	"CK06_01",	"CK09_11",	"CK10_02",	"CK17_02",	"CK17_03",	"CK18_01",	"CK19_03",	"CK20_01",	"CK22_03",	"CK22_04", "CK23_08", "CK51_01", "CK51_06", "CK51_09", "CK51_11", "EA01_01", "Gold3", "Gold9", "Hayward", "Hort16A", "Hort22d", "Jinfeng", "Kuimi", "M33", "Red19", "Russell", "T94.30-03-10f", "T94.30-04-08b", "T94.30-04-08c")); setnames(PID_avaliable_PSAphen,c("V1"),c("PedigreeItem"))

pheneotype_5sus_1tol <- data.table(list("5","5","5","4","3","1","2","4","3","5","1","4","1","3","5","3","4","4","2","2","1","4","4","1","1","1","2","2","3","2","1")); setnames(pheneotype_5sus_1tol,c("V1"),c("PSA_1tol_5sus"))
Phenotype_dt <- cbind(PID_avaliable_PSAphen,pheneotype_5sus_1tol)

Phenotype_dt <- Phenotype_dt[, "PedigreeItem2" := str_replace_all(PedigreeItem, c("CK02_01"="P1",	"CK19_03"="P2",	"CK23_08"="P3",	"CK51_09"="P4",	"CK09_11"="P5",	"CK17_03"="P6",	"CK18_01"="P7",	"Russell"="P8", "CK10_02"="P9",	"CK20_01"="P10",	"CK22_03"="P11",	"CK51_01"="P12",	"CK01_01_01_01"="P13",	"CK02_04"="P14",	"CK06_01"="P15",	"CK10_05"="P16",	"CK13_02"="P17",	"CK15_03"="P18",	"CK15_04"="P19",	"CK16_01"="P20",	"CK17_02"="P21",	"CK26_01"="P22"))]

Phenotype_dt <- Phenotype_dt[, "PedigreeItem3" := str_replace_all(PedigreeItem2, c("CK22_04"="GP1", "CK51_06"="GP2", "CK51_11"="GP3", "Gold3"="GP4", "Gold9"="GP5", "Hort16A"="GP6",  "Hort22d"="GP7", "Jinfeng"="GP8", "Kuimi"="GP9", "M33"="GP10",  "Red19"="GP11", "T94.30-03-10f"="GP12",  "T94.30-04-08b"="GP13", "T94.30-04-08c"="GP14"))]

Phenotype_dt <- Phenotype_dt[, `:=` (PedigreeItem=as.character(PedigreeItem), PSA_1tol_5sus=as.numeric(PSA_1tol_5sus))]
setkey(Phenotype_dt,"PedigreeItem")
Rualn_bams_phe <- Rualn_bams[Phenotype_dt]
fwrite(Rualn_bams[Phenotype_dt], "Rualn_bams_phe.txt", append = F)

#file.symlink(Rualn_bams$pathandfile, here('QTLseqR/21.psaQTLcheck'))

```

## index bams

```{r index, eval=FALSE}
bamin          <- paste0(here('QTLseqR/21.psaQTLcheck/'),Rualn_bams$V1)
file.exists(bamin)
for (i in bamin){
cmd        <- glue::glue("samtools index {i}")
Log          <- "idx" ;Ol<-paste0(here('QTLseqR/21.psaQTLcheck/LogD/'),Log,".out");Oe<-paste0(here('QTLseqR/21.psaQTLcheck/LogD/'),Log,".err")
bsub_cmd     <- glue("bsub -n 8 -o {Ol} -e {Oe} {cmd}")
#system(bsub_cmd)
}
```

## generate vcfs of parent files which are sole parents 

```{r run bcftools mpileup, eval=FALSE}
submit_mpileup <- function(reference = "*.fa.gz", output = c(), bamfiles = c(), job.init = "asub", job.opts = "-q normal") {
  cmd          <- "bcftools mpileup -B -I -Q 20 -f {reference} --excl-flags 'UNMAP,SECONDARY,QCFAIL,DUP' -a {mpileuptags} {bamfiles} -Ou  |bcftools call -f {bcfcalltags} -p {pvalThres} -m -A -Oz -o {output} && bcftools index {output}" 
  tmpf         <- tempfile(pattern = "submit_mileup.", fileext = ".sh")
  writeLines(text = c(glue(cmd)), con = tmpf)
  submit       <- "{job.init} {job.opts} {tmpf}"
  system(glue(submit), intern = TRUE)
}
mpileuptags    <- "DP,AD,SP,INFO/AD"
bcfcalltags    <- "GQ,GP"
pvalThres      <- 0.99
PedigreeItem   <- Rualn_bams$PedigreeItem
bamfiles       <- paste0(here('QTLseqR/21.psaQTLcheck/'),Rualn_bams$V1)
output         <- paste(PedigreeItem,"_alone", ".vcf", sep = "")
reference      <- dir("/workspace/hrards/projects/hrtcdf/genome/fasta", pattern = ".*\\.gz$", full.names = TRUE)
file.exists(bamfiles)
file.exists(reference)
#submit_mpileup(reference = reference, output = output, bamfiles = bamfiles)
```

## get bed regions from sole parents using pool parent region bed files for each pool and extract data from those regions.
```{r data from individual qtl by parentpool}
vcf_files  <- grep(pattern="*_alone.vcf$", list.files(path=here("QTLseqR/21.psaQTLcheck"), full.names=T), invert=F, value=T)

#just get those with a window of 5000 snps either side of the peak using bedfiles with the 5000bp*2 regions.
list_chrbeds_parent  <- grep(pattern="BSAregions_G5000_", list.files(path=here("QTLseqR"), full.names=T), invert=F, value=T)
names_chrbeds_parent <- str_replace_all(list_chrbeds_parent, c("/powerplant/workspace/hrtcdf/github/FSTs/QTLseqR/BSAregions_"="", "_bsa.bed"="", "G5000_"=""))

BSAallsnps_fromvcf   <- data.table("SAMPLE"=character(), "CHROM"=numeric(), "POS"=numeric(), "REF"=character(), "ALT0"=character(), "ALT1"=character(), "ALT2"=character(), "AD"=character(), "AD0"=numeric(), "AD1"=numeric(), "poolparent"=character(), "poolparent_chrpos"=character(),"qtl_name"=character(),"CHROMn"=numeric())#,"PPn"=numeric()

#for each parent get data from vcf sample files to a single data table.
for (b in seq(along=list_chrbeds_parent)){
  for (x in seq(along=vcf_files)){
  Bft <- fread(cmd=paste("bcftools query -f'[%SAMPLE \t%CHROM \t%POS \t%REF \t%ALT{0} \t%ALT{1} \t%ALT{2} \t%AD \t%AD{0} \t%AD{1} \n]'",paste0("-R '",list_chrbeds_parent[b],"'"), vcf_files[x]), fill=T)
  Bft <- Bft[, "poolparent" := names_chrbeds_parent[b]]
  
  #Bft <- Bft[, V3 := as.integer(str_pad(Bft$V3,8, side = c("left"), pad="0"))]
  Bft <- Bft[, "poolparent_chrpos" := paste(poolparent, V2, V3, sep=".")]  
  Bft <- setnames(Bft, c("V1","V2","V3","V4","V5","V6","V7","V8","V9","V10","poolparent", "poolparent_chrpos"), c("SAMPLE","CHROM", "POS", "REF", "ALT0", "ALT1", "ALT2","AD", "AD0", "AD1", "poolparent","poolparent_chrpos"))
  
  bed0            <- fread(list_chrbeds_parent[b])#get bedfile info for qtl_name
  #attach names from bed range qtl to data.table
  vcfGR  <- GRanges(seqnames = Bft$CHROM, ranges=IRanges(start = Bft$POS, end = Bft$POS))
  bedGR  <- GRanges(seqnames = bed0$V1, ranges=IRanges(start = bed0$V2, end = bed0$V3, names = bed0$V4))
  coords <- findOverlaps(vcfGR, bedGR)
  vcfGR[queryHits(coords)]
  Bft_vcf_hits <- Bft[queryHits(coords),]
  bed0[subjectHits(coords),]
  Bft_vcf_hits$qtl_name <- bed0[subjectHits(coords), V4]
  Bft_vcf_hits <- as.data.table(Bft_vcf_hits)
  Bft_vcf_hits <- Bft_vcf_hits[, "CHROMn" := as.numeric(str_replace_all(CHROM, "chr", ""))]
   #Bft_vcf_hits <- Bft_vcf_hits[, "PPn" := as.factor(as.numeric(str_replace_all(poolparent, "P", "")))]
  BSAallsnps_fromvcf  <- rbind(Bft_vcf_hits, BSAallsnps_fromvcf, use.names=FALSE)
  }
}
BSAallsnps_fromvcf  <- BSAallsnps_fromvcf[, "ChrPos" :=  as.numeric(paste(str_replace_all(CHROM, "chr",""), POS, sep="."))]

#attach base calls from samplepools (sp) 
BSAallsnps          <- fread(paste0(here('QTLseqR/'), "BSA0.5snps.csv"))
BSAallsnps          <- BSAallsnps[, "ChrPos" :=  as.numeric(paste(str_replace_all(CHROM, "chr",""), POS, sep="."))]
BSAallsnps          <- BSAallsnps[, "poolparent_chrpos" :=  paste(str_replace_all(parent, "_bsa",""),CHROM, POS, sep=".")]
BSAallsnps          <- BSAallsnps[order(Chromosome, POS),]
BSAallsnps_spChrPos <- BSAallsnps[,poolparent_chrpos]
BSAallsnps_ALT      <- BSAallsnps[,r1ALT]#changed from ALT as line 53 FSTs/WTLseqR/QTLseqR_ALL_pools.Rmd uses the r for reference which is the pool.
BSAallsnps_REF      <- BSAallsnps[,r1REF]

#takes a while to run: search for each ChrPos in BSAallsnps_fromvcf and apply the alternate to the data.table.
BSAallsnps_fromvcf_wALT   <- data.table("SAMPLE"=character(), "CHROM"=numeric(), "POS"=numeric(), "REF"=character(), "ALT0"=character(), "ALT1"=character(), "ALT2"=character(), "AD"=character(), "AD0"=numeric(), "AD1"=numeric(), "ChrPos"=numeric(), "bsaqtl_alt"=character(), "poolparent"=character(),"poolparent_chrpos"=character())
 for (cp in seq(along=BSAallsnps_spChrPos)){
 BSAallsnps_fromvcf_wALT   <- BSAallsnps_fromvcf[poolparent_chrpos==BSAallsnps_spChrPos[cp], `:=` ("bsaqtl_ref"= BSAallsnps_REF[cp],"bsaqtl_alt"= BSAallsnps_ALT[cp])]
}

BSAallsnps_fromvcf_wALT <- BSAallsnps_fromvcf_wALT[str_length(REF)>1, REF := "-"] #replace long reference allele strings
BSAallsnps_fromvcf_wALT <- BSAallsnps_fromvcf_wALT[, SAMPLE := str_replace_all(SAMPLE, c("ER10522_" = "", "SRA_" = "", "ER10400_" = "","ER10508_"=""))]#clean pedigreeID

fwrite(BSAallsnps_fromvcf_wALT,file=paste0("BSAallsnps_fromvcf_wALT.csv"), append=F)

for (b in seq(along=list_chrbeds_parent)){
BSAallsnps_fromvcf_wALTp <- BSAallsnps_fromvcf_wALT[poolparent==names_chrbeds_parent[b],]
BSAallsnps_fromvcf_wALTp <- unique(BSAallsnps_fromvcf_wALTp, by=c("poolparent","ChrPos","SAMPLE"))
sapply(BSAallsnps_fromvcf_wALTp,class)
fwrite(BSAallsnps_fromvcf_wALTp, file=paste0("prediction_", "G10000_", names_chrbeds_parent[b], ".csv"), append=F)
}

```

## get one bed at a time for the samples in each pool.

```{r samples in each pool}
poolparents <- c('P1','P2','P3','P4','P9','P10','P11','P12')
#poolparents <- c('P9')#subset P12 also gets i right when testing individual lines in loop.

if (file.exists("G10000_vcf_hits.csv")){
file.remove("G10000_vcf_hits.csv")#remove file to avoid appending to this file in loop
}

for (i in poolparents){
pred <-fread(paste0("/powerplant/workspace/hrtcdf/github/FSTs/QTLseqR/21.psaQTLcheck/prediction_G10000_",{i},".csv"))

#rename samples to match the pool parent
pred <- pred[, SAMPLE := str_replace_all(SAMPLE, c("ER10522_" = "", "SRA_" = "", "ER10400_" = "","ER10508_"=""))]#clean pedigreeID

pred <- pred[, SAMPLE := str_replace_all(SAMPLE, c("CK02_01"="P1",	"CK19_03"="P2",	"CK23_08"="P3",	"CK51_09"="P4",	"CK09_11"="P5",	"CK17_03"="P6",	"CK18_01"="P7",	"Russell"="P8", "CK10_02"="P9",	"CK20_01"="P10",	"CK22_03"="P11",	"CK51_01"="P12",	"CK01_01_01_01"="P13",	"CK02_04"="P14",	"CK06_01"="P15",	"CK10_05"="P16",	"CK13_02"="P17",	"CK15_03"="P18",	"CK15_04"="P19",	"CK16_01"="P20",	"CK17_02"="P21",	"CK26_01"="P22", "CK22_04"="GP1", "CK51_06"="GP2", "CK51_11"="GP3", "Gold3"="GP4", "Gold9"="GP5", "Hort16A"="GP6",  "Hort22d"="GP7", "Jinfeng"="GP8", "Kuimi"="GP9", "M33"="GP10",  "Red19"="GP11", "T94.30-03-10f"="GP12",  "T94.30-04-08b"="GP13", "T94.30-04-08c"="GP14"))]

 fwrite(pred, paste0(i, "G10000_vcf_hits.csv"))#for next chunk individual pools
 fwrite(pred, "G10000_vcf_hits.csv", append=T)#for  all pools
}
```

## transform data to fit PCA script format this is altered to fit 2 homologues(strands) from each parent contributing to the pool. as the PCA will include one of the parents, its closest match will be found on the PCA plot and doesn't need to be calculated as a distance between the parents and pool.
## Run all in a big loop with analysis and plotting for each pool

```{r transform and write figures, fig.show='hold'}
# options(digits=10),
pool  <- c('P1','P2','P3','P4','P9','P10','P11','P12')
#pool  <- c('P11') #subset
for (pooli in pool){
pool <- pooli

write(pool, file="tmp_pool") #to be used in the next plotting chunk based on this one.

fileP <- paste0("/powerplant/workspace/hrtcdf/github/FSTs/QTLseqR/21.psaQTLcheck/",pool,"G10000_vcf_hits.csv")
sites <- fread(fileP)
sites <- sites[,"ChrPos" := paste(CHROMn,POS,sep=".")] #an issue with fread requires these to be re-writen.

## addpool parent as a sample Pxx
samplepoolsites  <- sites[SAMPLE==pool] #use the pool parent to select chromosome and position sites for Pxx
samplepoolsites  <- samplepoolsites[bsaqtl_ref=="", `:=` (bsaqtl_ref=REF,bsaqtl_alt="-")] #fill missing ref sites. #bsaqtl_alt=ALT0
samplepoolsites  <- samplepoolsites[, `:=` (SAMPLE = paste0(poolparent,"sp"), REF=bsaqtl_ref,ALT0=bsaqtl_alt)]#change sample name to pxx, and place sample pool calls to ref and alt0.
sites            <- rbind(sites,samplepoolsites); rm(samplepoolsites)#attach sample pool sites to parent sites.
px               <- unique(sites$SAMPLE) #get sample names

sites  <- sites[AD0=="0", `:=` (REF="-")][AD==".", `:=` (REF="-",ALT0="-",ALT1="-")][REF=="", `:=` (REF="-")][ALT0==".", `:=` (ALT0="-")][ALT1==".", `:=` (ALT1="-")][str_length(REF)>1, `:=` (REF="-")][str_length(ALT0)>1, `:=` (ALT0="-")][str_length(ALT1)>1, `:=` (ALT1="-")][bsaqtl_ref=="", bsaqtl_ref := "-"][bsaqtl_alt=="", bsaqtl_alt := "-"]

###stack reference and alt sourced homologue columns
for (i in seq(along=px)){
###REF give new column for ref and change name of REF to bases(sp) so it can match alts
assign(paste0("REFsites",px[i]), sites[SAMPLE==px[i],  .(poolparent, SAMPLE,  qtl_name, REF,ChrPos)][,"R0A1" :=0])
assign(paste0("REFsites",px[i]), setnames(get(paste0("REFsites",px[i]), envir = .GlobalEnv), c("REF"), paste0("Bases",px[i])))

###alts give new column for alt and change name of REF to bases(sp) so it can match ref
assign(paste0("ALTsites",px[i]), sites[SAMPLE==px[i],  .(poolparent, SAMPLE,  qtl_name, ALT0,ChrPos)][,"R0A1" :=1])
assign(paste0("ALTsites",px[i]), setnames(get(paste0("ALTsites",px[i]), envir = .GlobalEnv), c("ALT0"), paste0("Bases",px[i])))

##rbind ref and alts to make a column stacking both.
assign(paste0("stackedsites",px[i]), rbind.DataTable(get(paste0("REFsites",px[i]), envir = .GlobalEnv), get(paste0("ALTsites",px[i]), envir = .GlobalEnv)))

##Give new names to stackedsites columns to include all data into the base column name
old_names  <- names(get(paste0("REFsites",px[i]), envir = .GlobalEnv))
addpp      <- paste0("pp",get(paste0("stackedsites",px[i]), envir = .GlobalEnv)[1,c(1)])
new_names  <- c(paste0("poolparent_",px[i]), paste0("SAMPLE_",px[i]), paste0("qtl_name_",px[i]), paste0("Bases_",px[i],addpp), paste0("ChrPos"), paste0("R0A1"))

assign(paste0("stackedsites",px[i]), setnames(get(paste0("stackedsites",px[i]), envir = .GlobalEnv),c(old_names),c(new_names)))
}

Rualn_bams_phe <- fread("Rualn_bams_phe.txt", select = c("PedigreeItem3"))
PedigreeItems_v <- paste0("stackedsites", Rualn_bams_phe[,PedigreeItem3])

mrgd <- stackedsitesP1[, .(poolparent_P1, SAMPLE_P1, qtl_name_P1)]
mrgd <- mrgd[, `:=` ("Bases"="N", "ChrPos"=stackedsitesP1$ChrPos, "R0A1"=stackedsitesP1$R0A1)]

setnames(mrgd, names(mrgd), c("poolparent", "SAMPLE", "qtl_name", "BasesPxx", "ChrPos", "R0A1"))
setkey(mrgd, ChrPos, R0A1)
assign(paste0("stack",pool), mrgd)

assign(paste0("stack",pool), setkey(get(paste0("stackedsites",pool), envir = .GlobalEnv), ChrPos, R0A1))

assign(paste0("stack",pool), merge.data.table(get(paste0("stackedsites",pool), envir = .GlobalEnv),mrgd, by = c("ChrPos","R0A1"), all=F))
for (PItem in PedigreeItems_v){
assign(paste0("stack",pool), merge.data.table(get(paste0("stack",pool), envir = .GlobalEnv),get(PItem, envir = .GlobalEnv), by = c("ChrPos","R0A1"), all=F))
}

#attach the SAMP_spqtlname_R0A1 name to sequence title instead of e.g bases.x use multiples of inserted for each data.table merged for the subset_title and mrgdt.
#transpose to wide for each qtl_name over all merged parent columns

qtl_names       <- unlist(na.omit(unique(get(paste0("stack",pool), envir = .GlobalEnv)[,c(5)])), use.names = F)
mrgd_names      <- names(get(paste0("stack",pool), envir = .GlobalEnv))[5]

for (i in qtl_names){
  assign(i, get(paste0("stack",pool), envir = .GlobalEnv)[get(mrgd_names)==i]) #subset for each QTL
  subset_title   <- as.character(get(i)[1, c(5,9,13,17,21,25,29,33,37,41, 45,49,53,57,61,65,69,73,77,81,85,89, 93,97,101,105,109,113,117,121, 125,129,133)]) # get names of each QTL, attach sample name for bases column

  subset_Bases         <- str_subset(names(get(i)), "Bases*")                                      # get old column names
  nameswith_sp         <- str_replace_all(str_subset(names(get(i)), "qtl_name*"), "qtl_name_","")  # get the sample name from bases
  subset_title_qtlname <- str_c(nameswith_sp,subset_title, sep="_")  # new row names with qtl_name and correct sample and sample pool name sp..
  assign(i, setnames(get(i), subset_Bases, subset_title_qtlname))    #setnames qtl_names to bases column
  assign(i, t(get(i)[, subset_title_qtlname, with = FALSE]))         #transpose and subset
  irownames<- as.data.table(rownames(get(i)),get(i))                 #get the rownames and paste as a column
  assign(i, cbind(irownames, apply(get(i), 1, paste, collapse="")))
  assign(i, get(i)[-2,])
  write.table(get(i),file=paste0("wide",i,".tsv"),append = FALSE,col.names=F,row.names=F, sep="\t")
  }
#}
#```

##reading the aligned sequence data: the data have to be formatted in tab-separated text with two colums, (name of sequence) \t (aligned sequence)

#```{r, running this chunk is dependant on running the above chunk for each pool, eval=TRUE}

if(file.exists(paste0(pool,"sPC_sample_QTLs.txt"))) {file.remove(paste0(pool,"sPC_sample_QTLs.txt"))}

for (ii in qtl_names){
#sites <- fread(file=paste0("/powerplant/workspace/hrtcdf/github/FSTs/QTLseqR/21.psaQTLcheck/","wide",{ii}), header=F)
sites <- get(ii)
#sites <- read.table(file="/powerplant/workspace/hrtcdf/github/Stats/human_aligned.txt", header=F, sep="\t") #Example data
sites<-as.matrix(sites)
dim(sites)

### finding the size of data
n_sample <- dim(sites)[1]
n_seq <- nchar(sites[2,2])

#check that each is the same length by changing the vector e.g. sites[x,2]
tapply(sites[,2], sites[,1], nchar)


### translation of the sequence to boolean vectors 

	bool <- array(0, dim=c(n_sample, 5*n_seq))

	colnames(bool) <- c(paste("A_", 1:n_seq, sep=""),paste("T_", 1:n_seq, sep=""),paste("G_", 1:n_seq, sep=""),paste("C_", 1:n_seq, sep=""),paste("N_", 1:n_seq, sep=""))
	rownames(bool) <- sites[ ,1]


  for (s in 1:n_sample){
       se <- sites[s, 2]
       se <- tolower(se)

	for (le in  1:n_seq){
	 base <- substr(se, le,le)
 
	if(base =="a") {
	bool[s, le] <-1
		} else {

	if(base =="t") {
	bool[s, le+n_seq] <-1
		} else {

	if(base =="g") {
	bool[s, le+n_seq*2] <-1
		} else {

	if(base =="c") {
	bool[s, le+n_seq*3] <-1
		} else {

	bool[s, le+n_seq*4] <-1
	}}}}
	}}

	apply(bool, 1, sum)  # here you can verify the translation
		# they should show identical values same as n_seq


############ PCA 

## centering : the center can be replaced to certain group
	center<- apply(bool, 2, mean)
	diffs<-sweep(bool, 2, center) #applying the mean to bool(the matrix) to move the center closer to zero for diffs
	diffs <- diffs/(2^0.5) # compensating the doubled counts in Euclidean distance metrics

	# checking distribution of the Euclidean distances
#    		dist<- (apply(diffs^2, 1, sum))^0.5
#         	qqnorm(dist)
# 		hist(dist)
		
### PCA core
	res_svd <- svd(diffs)  #
	str(res_svd)
			Left <- res_svd$u		# the left singular vector
			Right <- res_svd$v		# the right singular vector
			sqL <- diag(res_svd$d)		# diagonal matrix of the singular values

 	### calculating pc's 
	sPC_nuc  	<-	 Right %*% sqL / (n_sample^0.5)
	sPC_sample	 <-	 Left %*% sqL/ (n_seq^0.5)

	rownames(sPC_nuc)<- colnames(bool) 
 	rownames(sPC_sample)<- rownames(bool) 

#### output to text files
	write.table(sPC_sample, file="sPC_sample.txt", sep="\t")
	write.table(sPC_nuc, file="sPC_nuc.txt", sep="\t")

#### append to another file for each qtl analysed.
sPC_sample <- data.table(row.names(sPC_sample), sPC_sample)
sPC_sample <- sPC_sample[, "qtl" := ii]
fwrite(sPC_sample, file=paste0(pool,"sPC_sample_QTLs.txt"), sep="\t", append=T,col.names=F)
}
#}
#```

## plot all qtl for each pool

#```{r plot qtl for pool, eval=TRUE}
#attach phenotype
#PCA_of_poolQTL  <- fread("P11sPC_sample_QTLs.txt")
PCA_of_poolQTL  <- fread(paste0(pool,"sPC_sample_QTLs.txt"))

Phenotype <- fread("Rualn_bams_phe.txt", select = c("PedigreeItem3","PSA_1tol_5sus")); setkey(Phenotype,PedigreeItem3)

PCA_of_poolQTL  <- PCA_of_poolQTL[, c("PedigreeItem", "s2","s3","s4") := tstrsplit(V1, "_", fixed=TRUE)]
PCA_of_poolQTL  <- PCA_of_poolQTL[, PedigreeItem := str_replace_all(PCA_of_poolQTL$PedigreeItem,c(".x"="poolsample",".y"=""))]
setnames(PCA_of_poolQTL, names(PCA_of_poolQTL), str_replace_all(names(PCA_of_poolQTL),c(V34="QTL_name",V2="PC1",V3="PC2")))
setkey(PCA_of_poolQTL,PedigreeItem)

PCA_of_poolQTL  <- Phenotype[PCA_of_poolQTL] #merge phenotypes 
#alter phenotype names

PCA_of_poolQTL  <- PCA_of_poolQTL[, PSA_1tol_5sus := as.character(PCA_of_poolQTL$PSA_1tol_5sus)]
samplepoolname  <- paste0("Sample_pool (",pool,")")

PCA_of_poolQTL  <- PCA_of_poolQTL[is.na(PSA_1tol_5sus), PSA_1tol_5sus := glue(samplepoolname)] # paste0("Sample pool (",pool,")")
#PCA_of_poolQTL  <- PCA_of_poolQTL[, PSA_1tol_5sus := str_replace_all(PCA_of_poolQTL$PSA_1tol_5sus, "0", paste0("Sample pool (",pool,")"))] #
PCA_of_poolQTL  <- PCA_of_poolQTL[, PSA_1tol_5sus := str_replace_all(PCA_of_poolQTL$PSA_1tol_5sus, c( "^1"="Tolerant plant", "^2"="Tolerant plant", "^3"="Susceptible plant", "^4"="Susceptible plant", "^5"="Susceptible plant"))] #"0"=samplepoolname,
unique(PCA_of_poolQTL$PSA_1tol_5sus)

pool                <- names(fread("tmp_pool"))
PC1_contribution    <- round((res_svd$d/sum(res_svd$d)*100)[1], digits=2) #all contributions(res_svd$d/sum(res_svd$d)*100)[1:6]
PC2_contribution    <- round((res_svd$d/sum(res_svd$d)*100)[2], digits=2)

PCA_of_poolQTLp <- function(A) {
    ggplot(PCA_of_poolQTL, aes(x=PC2, y=PC1, colour=QTL_name, fill=QTL_name,shape=as.factor(PSA_1tol_5sus)))+ #,shape=PSA_1tol_5sus
    geom_point(size= 1, alpha = 7/10)+
    scale_shape_manual(values=c(8,16,17)) +
    scale_x_continuous(name=paste0("PC2 (",PC2_contribution," %)")) + #
    scale_y_continuous(name=paste0("PC1 (",PC1_contribution," %)")) + #, breaks=seq(-0.2,0.2,0.04)
    theme(plot.title   = element_text(hjust = 0.5, vjust = 0.5, size=12),
          axis.title.y = element_text(hjust = 0.5, vjust = 0.5, size=8),
          axis.title.x = element_text(hjust = 0.5, vjust = 0,   size=8),
          axis.text.x  = element_text(angle = 0,   vjust = 0.5, size=8),
          axis.text.y  = element_text(vjust = 0,   hjust = 0.5, size=8),
          legend.title = element_text(size=9),
          legend.text  = element_text(size=8))+
    guides(shape = guide_legend(str_wrap(paste0("PSA phenotype of parental individuals with the sample pool QTL were derived from"), 23), order=1), colour = guide_legend("QTL name", order=2), fill="none")
    }
p <- PCA_of_poolQTLp()
ggsave(paste0("PCA_poolparents_vs_pool_",pool,".png"), plot = p, height = 12, width=15, units = c("cm"), dpi = 320)
#p
}
```
